{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/apua/.local/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:14: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.\n",
      "  from pandas.core.index import Index as PandasIndex\n",
      "/home/apua/.local/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:34: UserWarning: pandas >= 1.0 is not supported.\n",
      "  warnings.warn('pandas >= 1.0 is not supported.')\n"
     ]
    }
   ],
   "source": [
    "# Leiden and igraph\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.neighbors import kneighbors_graph\n",
    "import igraph as ig\n",
    "import leidenalg\n",
    "import re\n",
    "import scipy.stats as stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "\n",
    "# Ensembl ID to Entrez ID conversion\n",
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "\n",
    "\n",
    "# GO analysis\n",
    "from goatools.base import download_go_basic_obo\n",
    "from goatools.base import download_ncbi_associations\n",
    "from goatools.obo_parser import GODag\n",
    "from goatools.anno.genetogo_reader import Gene2GoReader\n",
    "from goatools.goea.go_enrichment_ns import GOEnrichmentStudyNS\n",
    "\n",
    "\n",
    "# Word cloud\n",
    "import wordcloud\n",
    "import math\n",
    "import random\n",
    "\n",
    "# GTF parser for Ensembl ID to gene symbol conversion\n",
    "from gtfparse import read_gtf\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def latent_variables_to_leiden(df, n_neighbors, directed, seed):\n",
    "    \n",
    "    adjacency = kneighbors_graph(df, n_neighbors, mode='connectivity', include_self=False)\n",
    "    \n",
    "    distance = kneighbors_graph(df, n_neighbors, mode='distance', include_self=False)\n",
    "    \n",
    "    sources, targets = adjacency.nonzero()\n",
    "\n",
    "    # weight is inverse distance squared\n",
    "\n",
    "    weights = 1 /  distance[sources, targets].A1 ** 2\n",
    "\n",
    "    \n",
    "    g = ig.Graph(directed=directed)\n",
    "    g.add_vertices(adjacency.shape[0])  # this adds adjacency.shape[0] vertices\n",
    "    g.add_edges(list(zip(sources, targets)))\n",
    "    g.es['weight'] = weights\n",
    "    \n",
    "    part = leidenalg.find_partition(g, leidenalg.ModularityVertexPartition, weights='weight', seed=seed)\n",
    "    \n",
    "    \n",
    "    membership_df = pd.DataFrame({'cluster_id':part.membership}, index=df.index)\n",
    "    \n",
    "    return (membership_df, part)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensembl to Entrez\n",
    "\n",
    "class EnsemblIDToEntrezIDConverter():\n",
    "    def __init__(self):\n",
    "        with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "            df_a = ro.conversion.rpy2py(ro.r('''\n",
    "                                             library(org.Hs.eg.db)\n",
    "                                             as.data.frame(org.Hs.egENSEMBL)\n",
    "                                            '''))\n",
    "            self.entrez_ensembl_df = df_a.groupby(['ensembl_id']).first()\n",
    "    \n",
    "    def convert(self, gene_list_or_set):\n",
    "        \n",
    "        ensembl_id_df = pd.DataFrame(index = { re.sub(\"\\..*$\",\"\",x) for x in gene_list_or_set })\n",
    "        entrez_ids = set(ensembl_id_df.merge(self.entrez_ensembl_df, left_index=True, right_index=True)['gene_id'].astype(int))\n",
    "        return entrez_ids\n",
    "        \n",
    "    \n",
    "    \n",
    "    \n",
    "# Gene Ontology\n",
    "    \n",
    "def get_entrez_ensembl_df():\n",
    "    with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "        df_a = ro.conversion.rpy2py(ro.r('''\n",
    "                                         library(org.Hs.eg.db)\n",
    "                                         as.data.frame(org.Hs.egENSEMBL)\n",
    "                                        '''))\n",
    "        entrez_ensembl_df = df_a.groupby(['ensembl_id']).first()\n",
    "        return entrez_ensembl_df\n",
    "\n",
    "\n",
    "class MyGeneOntologyAnalysis():\n",
    "    def __init__(self):\n",
    "        \n",
    "\n",
    "        obo_fname = download_go_basic_obo()\n",
    "        file_gene2go = download_ncbi_associations()\n",
    "        obodag = GODag(\"go-basic.obo\")\n",
    "\n",
    "        # Read NCBI's gene2go. Store annotations in a list of namedtuples\n",
    "        objanno = Gene2GoReader(file_gene2go, taxids=[9606])\n",
    "\n",
    "        # Get associations for each branch of the GO DAG (BP, MF, CC)\n",
    "        ns2assoc = objanno.get_ns2assc()\n",
    "\n",
    "        for nspc, id2gos in ns2assoc.items():\n",
    "            print(\"{NS} {N:,} annotated human genes\".format(NS=nspc, N=len(id2gos)))\n",
    "          \n",
    "        genes_with_annotation = set.union(*(set(x.keys()) for x in ns2assoc.values()))\n",
    "\n",
    "        \n",
    "        self.alpha = 0.05\n",
    "        \n",
    "        self.goeaobj = GOEnrichmentStudyNS(\n",
    "            genes_with_annotation, # List of human genes\n",
    "            ns2assoc, # geneid/GO associations\n",
    "            obodag, # Ontologies\n",
    "            propagate_counts = False,\n",
    "            alpha = self.alpha, # default significance cut-off\n",
    "            methods = ['fdr_bh']) # defult multipletest correction method\n",
    "        \n",
    "        \n",
    "\n",
    "    def goea_results_all(self, gene_set):\n",
    "        return self.goeaobj.run_study(gene_set)\n",
    "    \n",
    "    \n",
    "    def goea_results_significant(self, gene_set):\n",
    "        all_results = self.goea_results_all(gene_set)\n",
    "        return [r for r in all_results if r.p_fdr_bh < self.alpha]\n",
    "    \n",
    "    \n",
    "\n",
    "# Word cloud\n",
    "    \n",
    "class GeneOntologyWordCloud():\n",
    "    def __init__(self):\n",
    "        self.wc = wordcloud.WordCloud(colormap='rainbow', \n",
    "                             stopwords=['integral', 'component', 'of', 'process', 'activity', 'to'],\n",
    "                             collocations = True,\n",
    "                             ranks_only=True,\n",
    "                         )\n",
    "        \n",
    "    def gen_random_text(self):\n",
    "        # Generate a spacer between gene ontology terms\n",
    "        return ' '.join(''.join((random.choice('abcdefghijklmnopqrstuvwxyz') for _ in range(150))) for _ in range(3))\n",
    "        \n",
    "    def goea_to_text(self, goea_results):\n",
    "        # Generate a text concatenating GO term names, each name being repeated proportionally to the minus log of its pvalue\n",
    "        return ' '.join((''.join( ' ' + self.gen_random_text() + ' ' + x.name for _ in range(int((-10) * math.log2(x.p_uncorrected))))\n",
    "                for x in goea_results  if x.enrichment == 'e'))\n",
    "    \n",
    "    def generate_image(self, goea_results):\n",
    "        return self.wc.generate(self.goea_to_text(goea_results)).to_image()\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gtf_genes_df():\n",
    "    try:\n",
    "        gtf_df_genes = pd.read_csv(\"gtf_df_genes.csv\")\n",
    "    except:\n",
    "        gtf_df = read_gtf(\"/ceph/genome/human/gencode25/gtf.CHR/_m/gencode.v25.annotation.gtf\")\n",
    "        gtf_df_genes = gtf_df[gtf_df[\"feature\"] == \"gene\"][['gene_id', 'gene_name']]\n",
    "        gtf_df_genes.to_csv(\"gtf_df_genes.csv\", index=None)\n",
    "        \n",
    "    return gtf_df_genes\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ClusterGOWordCloud():\n",
    "    def __init__(self, n_neighbors, directed, seed):\n",
    "        \n",
    "        self.n_neighbors = n_neighbors\n",
    "        expression_df = pd.read_csv('../../_m/latent_variables.csv', index_col=0)\n",
    "        mucols = [x for x in expression_df.columns if 'mu' in x]\n",
    "        self.expression_df = expression_df[mucols]\n",
    "        \n",
    "        \n",
    "        \n",
    "        (self.mdf, self.part) =  latent_variables_to_leiden(self.expression_df, n_neighbors, directed, seed)\n",
    "        \n",
    "        \n",
    "        self.gtf_df_genes = get_gtf_genes_df()\n",
    "\n",
    "        self.e2e = EnsemblIDToEntrezIDConverter()\n",
    "        self.mygoa = MyGeneOntologyAnalysis()\n",
    "        self.gowc = GeneOntologyWordCloud()\n",
    "    \n",
    "    \n",
    "    def cluster_df(self, cluster_id):\n",
    "        \n",
    "        return self.expression_df.iloc[self.part[cluster_id]][[]]\\\n",
    "        .merge(cgowc.gtf_df_genes, left_index=True, right_on='gene_id', how='left')\\\n",
    "        .set_index('gene_id')\n",
    "        \n",
    "    \n",
    "    def pipeline(self, cluster_id, filename_prefix):\n",
    "        \n",
    "        #with open(\"%s_neighbors.txt\" % filename_prefix, \"wt\") as f:\n",
    "        #    for x in nn:\n",
    "        #        print(x, file=f)\n",
    "\n",
    "        #self.gn.neighbors_df(gene_id, self.n_neighbors)\\\n",
    "        #.to_csv(\"%s_neighbors.csv\" % filename_prefix)\n",
    "        \n",
    "        #nn = self.gn.neighbors(gene_id, self.n_neighbors)\n",
    "        \n",
    "        self.cluster_df(cluster_id).to_csv(\"%s_genes.csv\" % filename_prefix)\n",
    "        \n",
    "        nn = set((self.expression_df.index[x] for x in self.part[cluster_id]))\n",
    "                \n",
    "        go_r = self.mygoa.goea_results_significant(self.e2e.convert(nn))\n",
    "        self.mygoa.goeaobj.wr_tsv(\"%s_go_enrichment.tsv\" % filename_prefix, go_r)\n",
    "        \n",
    "        if len(go_r) > 0:\n",
    "            p = self.gowc.generate_image(go_r)\n",
    "            p.save(\"%s_go_wordcloud.png\" % filename_prefix)\n",
    "        \n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: AnnotationDbi\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: stats4\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: parallel\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from 'package:base':\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,\n",
      "    tapply, union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: Biobase\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: IRanges\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following object is masked from 'package:base':\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2020-04-23) 47,239 GO Terms\n",
      "HMS:0:00:03.522364 330,518 annotations, 20,562 genes, 18,403 GOs, 1 taxids READ: gene2go \n",
      "MF 17,537 annotated human genes\n",
      "BP 18,499 annotated human genes\n",
      "CC 19,334 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 90% 18,499 of 20,517 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 94% 19,334 of 20,517 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 85% 17,537 of 20,517 population items found in association\n"
     ]
    }
   ],
   "source": [
    "cgowc = ClusterGOWordCloud(8, False, 1)\n",
    "\n",
    "#1,5,6\n",
    "#8\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cgowc.part)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "D2 junction 5-6 is in cluster 0\n",
      "D2 junction 5-7 is in cluster 6\n",
      "SETD1A is in cluster 3\n"
     ]
    }
   ],
   "source": [
    "a = [x for x in range(len(cgowc.part)) if 1 in cgowc.part[x]][0]\n",
    "print(\"D2 junction 5-6 is in module\", a)\n",
    "\n",
    "a = [x for x in range(len(cgowc.part)) if 0 in cgowc.part[x]][0]\n",
    "print(\"D2 junction 5-7 is in module\", a)\n",
    "\n",
    "d = {z[0]:z[1]  for z in zip(cgowc.gtf_df_genes['gene_name'], cgowc.gtf_df_genes['gene_id'])}\n",
    "a = cgowc.mdf.loc[d['SETD1A'], 'cluster_id']\n",
    "print(\"SETD1A is in module\", a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GWAS, TWAS and DE enrichment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2699"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de_genes = set(pd.read_csv('/ceph/projects/v3_phase3_paper/analysis/differential_expression/_m/genes/diffExpr_szVctl_FDR05.txt',\n",
    "                      sep='\\t', usecols=[0], index_col=0).index)\n",
    "len(de_genes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "389"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "twas_genes = set(pd.read_csv('/ceph/users/apua/projects/caudate_twas_reader/genes/_m/twas_significant_genes.csv')['gene_id'])\n",
    "len(twas_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2000"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_genes = set(pd.read_csv('/ceph/projects/v3_phase3_paper/inputs/gwas/PGC2_CLOZUK/table_s3/hg38/genes/_m/gwas_genes.csv')['gene_id'])\n",
    "len(gwas_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "383"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mhc_genes = set(pd.read_csv('/ceph/projects/v3_phase3_paper/inputs/gwas/PGC2_CLOZUK/table_s3/hg38/mhc_region_genes/_m/mhc_genes.csv')['gene_id'])\n",
    "len(mhc_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fet(a, b, u):\n",
    "    # a, b, u are sets\n",
    "    # u is the universe\n",
    "    \n",
    "    yes_a = u.intersection(a)\n",
    "    yes_b = u.intersection(b)\n",
    "    no_a = u - a\n",
    "    no_b = u - b\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    m = [[len(yes_a.intersection(yes_b)), len(no_a.intersection(yes_b)) ], \n",
    "                               [len(yes_a.intersection(no_b)), len(no_a.intersection(no_b))]]\n",
    "    return stats.fisher_exact(m) #, m, len(yes_b)/len(u), len(yes_a)/len(u), len(yes_a.intersection(yes_b)) * len(no_a.intersection(no_b)) / (len(yes_a.intersection(no_b)) * len(no_a.intersection(yes_b)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>gwas_or</th>\n",
       "      <th>gwas_p</th>\n",
       "      <th>twas_or</th>\n",
       "      <th>twas_p</th>\n",
       "      <th>de_or</th>\n",
       "      <th>de_p</th>\n",
       "      <th>twas_fdr_bh</th>\n",
       "      <th>gwas_fdr_bh</th>\n",
       "      <th>de_fdr_bh</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4917</td>\n",
       "      <td>1.266284</td>\n",
       "      <td>0.001428</td>\n",
       "      <td>1.391645</td>\n",
       "      <td>0.008004</td>\n",
       "      <td>2.569408</td>\n",
       "      <td>6.466662e-92</td>\n",
       "      <td>0.026679</td>\n",
       "      <td>0.028564</td>\n",
       "      <td>6.466662e-91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2877</td>\n",
       "      <td>0.778369</td>\n",
       "      <td>0.017285</td>\n",
       "      <td>0.592902</td>\n",
       "      <td>0.007999</td>\n",
       "      <td>0.246380</td>\n",
       "      <td>5.581596e-53</td>\n",
       "      <td>0.026679</td>\n",
       "      <td>0.057615</td>\n",
       "      <td>3.721064e-52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2288</td>\n",
       "      <td>0.899150</td>\n",
       "      <td>0.360731</td>\n",
       "      <td>1.449815</td>\n",
       "      <td>0.021982</td>\n",
       "      <td>0.859366</td>\n",
       "      <td>4.910687e-02</td>\n",
       "      <td>0.055053</td>\n",
       "      <td>0.515329</td>\n",
       "      <td>5.777278e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1949</td>\n",
       "      <td>1.189048</td>\n",
       "      <td>0.117726</td>\n",
       "      <td>1.688713</td>\n",
       "      <td>0.002093</td>\n",
       "      <td>1.784156</td>\n",
       "      <td>5.959732e-17</td>\n",
       "      <td>0.010467</td>\n",
       "      <td>0.196210</td>\n",
       "      <td>1.191946e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1891</td>\n",
       "      <td>1.216091</td>\n",
       "      <td>0.077965</td>\n",
       "      <td>1.745942</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0.910389</td>\n",
       "      <td>2.657709e-01</td>\n",
       "      <td>0.008424</td>\n",
       "      <td>0.141754</td>\n",
       "      <td>2.797588e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1869</td>\n",
       "      <td>1.320908</td>\n",
       "      <td>0.011018</td>\n",
       "      <td>0.837988</td>\n",
       "      <td>0.481794</td>\n",
       "      <td>0.548429</td>\n",
       "      <td>1.043910e-10</td>\n",
       "      <td>0.759806</td>\n",
       "      <td>0.044072</td>\n",
       "      <td>1.739850e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1285</td>\n",
       "      <td>0.720704</td>\n",
       "      <td>0.041116</td>\n",
       "      <td>1.138840</td>\n",
       "      <td>0.548444</td>\n",
       "      <td>2.171390</td>\n",
       "      <td>4.011151e-22</td>\n",
       "      <td>0.783491</td>\n",
       "      <td>0.082233</td>\n",
       "      <td>1.337050e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1095</td>\n",
       "      <td>1.039440</td>\n",
       "      <td>0.761506</td>\n",
       "      <td>0.681277</td>\n",
       "      <td>0.244801</td>\n",
       "      <td>0.202418</td>\n",
       "      <td>2.688646e-24</td>\n",
       "      <td>0.544002</td>\n",
       "      <td>0.801585</td>\n",
       "      <td>1.075458e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1024</td>\n",
       "      <td>0.682996</td>\n",
       "      <td>0.040966</td>\n",
       "      <td>0.193611</td>\n",
       "      <td>0.000414</td>\n",
       "      <td>0.318141</td>\n",
       "      <td>1.880132e-15</td>\n",
       "      <td>0.004141</td>\n",
       "      <td>0.082233</td>\n",
       "      <td>3.418421e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1020</td>\n",
       "      <td>0.662454</td>\n",
       "      <td>0.027436</td>\n",
       "      <td>0.194404</td>\n",
       "      <td>0.000409</td>\n",
       "      <td>0.182406</td>\n",
       "      <td>1.948498e-24</td>\n",
       "      <td>0.004141</td>\n",
       "      <td>0.078390</td>\n",
       "      <td>9.742489e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>978</td>\n",
       "      <td>0.891954</td>\n",
       "      <td>0.574324</td>\n",
       "      <td>0.767274</td>\n",
       "      <td>0.493874</td>\n",
       "      <td>0.247440</td>\n",
       "      <td>7.138112e-19</td>\n",
       "      <td>0.759806</td>\n",
       "      <td>0.675675</td>\n",
       "      <td>1.784528e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>887</td>\n",
       "      <td>0.581306</td>\n",
       "      <td>0.008912</td>\n",
       "      <td>0.690150</td>\n",
       "      <td>0.318071</td>\n",
       "      <td>0.232534</td>\n",
       "      <td>4.200609e-18</td>\n",
       "      <td>0.636143</td>\n",
       "      <td>0.044072</td>\n",
       "      <td>9.334686e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>884</td>\n",
       "      <td>1.109010</td>\n",
       "      <td>0.500733</td>\n",
       "      <td>1.016012</td>\n",
       "      <td>0.886132</td>\n",
       "      <td>0.439310</td>\n",
       "      <td>9.457375e-09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.625916</td>\n",
       "      <td>1.351054e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>795</td>\n",
       "      <td>0.534681</td>\n",
       "      <td>0.004478</td>\n",
       "      <td>0.337148</td>\n",
       "      <td>0.022021</td>\n",
       "      <td>1.122408</td>\n",
       "      <td>3.080122e-01</td>\n",
       "      <td>0.055053</td>\n",
       "      <td>0.038343</td>\n",
       "      <td>3.080122e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>721</td>\n",
       "      <td>1.134354</td>\n",
       "      <td>0.457119</td>\n",
       "      <td>0.955760</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>7.223223</td>\n",
       "      <td>3.437122e-116</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.609492</td>\n",
       "      <td>6.874244e-115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>689</td>\n",
       "      <td>1.570363</td>\n",
       "      <td>0.005752</td>\n",
       "      <td>0.794982</td>\n",
       "      <td>0.629024</td>\n",
       "      <td>0.115921</td>\n",
       "      <td>6.615419e-21</td>\n",
       "      <td>0.838698</td>\n",
       "      <td>0.038343</td>\n",
       "      <td>1.890120e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>520</td>\n",
       "      <td>0.563061</td>\n",
       "      <td>0.037868</td>\n",
       "      <td>0.656070</td>\n",
       "      <td>0.457572</td>\n",
       "      <td>0.486857</td>\n",
       "      <td>8.455929e-05</td>\n",
       "      <td>0.759806</td>\n",
       "      <td>0.082233</td>\n",
       "      <td>1.056991e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>476</td>\n",
       "      <td>0.868569</td>\n",
       "      <td>0.649069</td>\n",
       "      <td>0.718591</td>\n",
       "      <td>0.695604</td>\n",
       "      <td>0.248598</td>\n",
       "      <td>1.050619e-09</td>\n",
       "      <td>0.869505</td>\n",
       "      <td>0.721188</td>\n",
       "      <td>1.616337e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>329</td>\n",
       "      <td>1.363791</td>\n",
       "      <td>0.216466</td>\n",
       "      <td>0.835660</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.366532</td>\n",
       "      <td>6.086230e-05</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.333024</td>\n",
       "      <td>8.114973e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.499663e-01</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.777403e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            n_genes   gwas_or    gwas_p   twas_or    twas_p     de_or  \\\n",
       "cluster_id                                                              \n",
       "0              4917  1.266284  0.001428  1.391645  0.008004  2.569408   \n",
       "1              2877  0.778369  0.017285  0.592902  0.007999  0.246380   \n",
       "2              2288  0.899150  0.360731  1.449815  0.021982  0.859366   \n",
       "3              1949  1.189048  0.117726  1.688713  0.002093  1.784156   \n",
       "4              1891  1.216091  0.077965  1.745942  0.001264  0.910389   \n",
       "5              1869  1.320908  0.011018  0.837988  0.481794  0.548429   \n",
       "6              1285  0.720704  0.041116  1.138840  0.548444  2.171390   \n",
       "7              1095  1.039440  0.761506  0.681277  0.244801  0.202418   \n",
       "8              1024  0.682996  0.040966  0.193611  0.000414  0.318141   \n",
       "9              1020  0.662454  0.027436  0.194404  0.000409  0.182406   \n",
       "10              978  0.891954  0.574324  0.767274  0.493874  0.247440   \n",
       "11              887  0.581306  0.008912  0.690150  0.318071  0.232534   \n",
       "12              884  1.109010  0.500733  1.016012  0.886132  0.439310   \n",
       "13              795  0.534681  0.004478  0.337148  0.022021  1.122408   \n",
       "14              721  1.134354  0.457119  0.955760  1.000000  7.223223   \n",
       "15              689  1.570363  0.005752  0.794982  0.629024  0.115921   \n",
       "16              520  0.563061  0.037868  0.656070  0.457572  0.486857   \n",
       "17              476  0.868569  0.649069  0.718591  0.695604  0.248598   \n",
       "18              329  1.363791  0.216466  0.835660  1.000000  0.366532   \n",
       "19               19  0.000000  1.000000  0.000000  1.000000  0.000000   \n",
       "\n",
       "                     de_p  twas_fdr_bh  gwas_fdr_bh      de_fdr_bh  \n",
       "cluster_id                                                          \n",
       "0            6.466662e-92     0.026679     0.028564   6.466662e-91  \n",
       "1            5.581596e-53     0.026679     0.057615   3.721064e-52  \n",
       "2            4.910687e-02     0.055053     0.515329   5.777278e-02  \n",
       "3            5.959732e-17     0.010467     0.196210   1.191946e-16  \n",
       "4            2.657709e-01     0.008424     0.141754   2.797588e-01  \n",
       "5            1.043910e-10     0.759806     0.044072   1.739850e-10  \n",
       "6            4.011151e-22     0.783491     0.082233   1.337050e-21  \n",
       "7            2.688646e-24     0.544002     0.801585   1.075458e-23  \n",
       "8            1.880132e-15     0.004141     0.082233   3.418421e-15  \n",
       "9            1.948498e-24     0.004141     0.078390   9.742489e-24  \n",
       "10           7.138112e-19     0.759806     0.675675   1.784528e-18  \n",
       "11           4.200609e-18     0.636143     0.044072   9.334686e-18  \n",
       "12           9.457375e-09     1.000000     0.625916   1.351054e-08  \n",
       "13           3.080122e-01     0.055053     0.038343   3.080122e-01  \n",
       "14          3.437122e-116     1.000000     0.609492  6.874244e-115  \n",
       "15           6.615419e-21     0.838698     0.038343   1.890120e-20  \n",
       "16           8.455929e-05     0.759806     0.082233   1.056991e-04  \n",
       "17           1.050619e-09     0.869505     0.721188   1.616337e-09  \n",
       "18           6.086230e-05     1.000000     0.333024   8.114973e-05  \n",
       "19           2.499663e-01     1.000000     1.000000   2.777403e-01  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def enrichment_rows():\n",
    "    part = cgowc.part\n",
    "    df = cgowc.expression_df\n",
    "    u = set(cgowc.expression_df.index[3:])\n",
    "    for ii in range(len(part)):\n",
    "        a = set((df.index[x] for x in part[ii]))\n",
    "        yield (ii,\n",
    "               len(part[ii]),\n",
    "               *fet(a, gwas_genes, u),\n",
    "               *fet(a, twas_genes, u),\n",
    "               *fet(a, de_genes, u),\n",
    "               )\n",
    "        \n",
    "    #print(ii, 0 in part[ii], 1 in part[ii], fet(a, gwas_genes, u), fet(a, twas_genes, u), fet(a, de_genes, u), len(part[ii]))\n",
    "    \n",
    "edf1 = pd.DataFrame.from_records(enrichment_rows(), \n",
    "                                 columns=['module_id', 'n_genes', 'gwas_or', 'gwas_p', 'twas_or', 'twas_p', 'de_or', 'de_p'],\n",
    "                                 index='module_id')\n",
    "edf1['twas_fdr_bh'] = multipletests(edf1['twas_p'], method='fdr_bh')[1]\n",
    "edf1['gwas_fdr_bh'] = multipletests(edf1['gwas_p'], method='fdr_bh')[1]\n",
    "edf1['de_fdr_bh'] = multipletests(edf1['de_p'], method='fdr_bh')[1]\n",
    "\n",
    "edf1[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', 'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']].to_csv('module_enrichment.csv')\n",
    "\n",
    "edf1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>gwas_or</th>\n",
       "      <th>gwas_p</th>\n",
       "      <th>twas_or</th>\n",
       "      <th>twas_p</th>\n",
       "      <th>de_or</th>\n",
       "      <th>de_p</th>\n",
       "      <th>twas_fdr_bh</th>\n",
       "      <th>gwas_fdr_bh</th>\n",
       "      <th>de_fdr_bh</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4917</td>\n",
       "      <td>1.324797</td>\n",
       "      <td>0.000445</td>\n",
       "      <td>1.444210</td>\n",
       "      <td>0.004375</td>\n",
       "      <td>2.551013</td>\n",
       "      <td>7.972072e-90</td>\n",
       "      <td>0.017390</td>\n",
       "      <td>0.008896</td>\n",
       "      <td>7.972072e-89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2877</td>\n",
       "      <td>0.870493</td>\n",
       "      <td>0.225448</td>\n",
       "      <td>0.627400</td>\n",
       "      <td>0.024133</td>\n",
       "      <td>0.246658</td>\n",
       "      <td>7.371443e-53</td>\n",
       "      <td>0.057630</td>\n",
       "      <td>0.375747</td>\n",
       "      <td>4.914295e-52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2288</td>\n",
       "      <td>1.047981</td>\n",
       "      <td>0.682872</td>\n",
       "      <td>1.609621</td>\n",
       "      <td>0.005217</td>\n",
       "      <td>0.859728</td>\n",
       "      <td>4.902329e-02</td>\n",
       "      <td>0.017390</td>\n",
       "      <td>0.951527</td>\n",
       "      <td>5.767446e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1949</td>\n",
       "      <td>1.012280</td>\n",
       "      <td>0.899541</td>\n",
       "      <td>1.802268</td>\n",
       "      <td>0.000816</td>\n",
       "      <td>1.791460</td>\n",
       "      <td>5.286182e-17</td>\n",
       "      <td>0.005545</td>\n",
       "      <td>0.999490</td>\n",
       "      <td>1.057236e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1891</td>\n",
       "      <td>1.273429</td>\n",
       "      <td>0.041003</td>\n",
       "      <td>1.797865</td>\n",
       "      <td>0.001047</td>\n",
       "      <td>0.910112</td>\n",
       "      <td>2.644619e-01</td>\n",
       "      <td>0.005545</td>\n",
       "      <td>0.117152</td>\n",
       "      <td>2.783810e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1869</td>\n",
       "      <td>1.048663</td>\n",
       "      <td>0.698717</td>\n",
       "      <td>0.554064</td>\n",
       "      <td>0.025933</td>\n",
       "      <td>0.559047</td>\n",
       "      <td>4.705380e-10</td>\n",
       "      <td>0.057630</td>\n",
       "      <td>0.951527</td>\n",
       "      <td>7.842299e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1285</td>\n",
       "      <td>0.646181</td>\n",
       "      <td>0.017297</td>\n",
       "      <td>0.944156</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.179576</td>\n",
       "      <td>3.004132e-22</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.082011</td>\n",
       "      <td>1.001377e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1095</td>\n",
       "      <td>1.034995</td>\n",
       "      <td>0.804356</td>\n",
       "      <td>0.682027</td>\n",
       "      <td>0.278153</td>\n",
       "      <td>0.203759</td>\n",
       "      <td>3.620710e-24</td>\n",
       "      <td>0.505732</td>\n",
       "      <td>0.999490</td>\n",
       "      <td>1.448284e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1024</td>\n",
       "      <td>0.627987</td>\n",
       "      <td>0.026124</td>\n",
       "      <td>0.213462</td>\n",
       "      <td>0.001107</td>\n",
       "      <td>0.320125</td>\n",
       "      <td>3.543820e-15</td>\n",
       "      <td>0.005545</td>\n",
       "      <td>0.087081</td>\n",
       "      <td>6.443309e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1020</td>\n",
       "      <td>0.738109</td>\n",
       "      <td>0.125045</td>\n",
       "      <td>0.213244</td>\n",
       "      <td>0.001109</td>\n",
       "      <td>0.182561</td>\n",
       "      <td>2.849244e-24</td>\n",
       "      <td>0.005545</td>\n",
       "      <td>0.303429</td>\n",
       "      <td>1.424622e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>978</td>\n",
       "      <td>0.952002</td>\n",
       "      <td>0.861848</td>\n",
       "      <td>0.845377</td>\n",
       "      <td>0.774114</td>\n",
       "      <td>0.248228</td>\n",
       "      <td>9.899234e-19</td>\n",
       "      <td>0.977790</td>\n",
       "      <td>0.999490</td>\n",
       "      <td>2.474808e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>887</td>\n",
       "      <td>0.571363</td>\n",
       "      <td>0.013425</td>\n",
       "      <td>0.760318</td>\n",
       "      <td>0.547618</td>\n",
       "      <td>0.233378</td>\n",
       "      <td>5.870878e-18</td>\n",
       "      <td>0.782311</td>\n",
       "      <td>0.082011</td>\n",
       "      <td>1.304640e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>884</td>\n",
       "      <td>1.067069</td>\n",
       "      <td>0.713645</td>\n",
       "      <td>1.035031</td>\n",
       "      <td>0.880011</td>\n",
       "      <td>0.443173</td>\n",
       "      <td>1.207681e-08</td>\n",
       "      <td>0.977790</td>\n",
       "      <td>0.951527</td>\n",
       "      <td>1.725259e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>795</td>\n",
       "      <td>0.570028</td>\n",
       "      <td>0.020503</td>\n",
       "      <td>0.276447</td>\n",
       "      <td>0.011106</td>\n",
       "      <td>1.124579</td>\n",
       "      <td>3.072798e-01</td>\n",
       "      <td>0.031733</td>\n",
       "      <td>0.082011</td>\n",
       "      <td>3.072798e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>721</td>\n",
       "      <td>1.277914</td>\n",
       "      <td>0.188717</td>\n",
       "      <td>1.051337</td>\n",
       "      <td>0.867762</td>\n",
       "      <td>7.211527</td>\n",
       "      <td>1.182565e-115</td>\n",
       "      <td>0.977790</td>\n",
       "      <td>0.343122</td>\n",
       "      <td>2.365130e-114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>689</td>\n",
       "      <td>1.278180</td>\n",
       "      <td>0.177376</td>\n",
       "      <td>0.661768</td>\n",
       "      <td>0.394268</td>\n",
       "      <td>0.118292</td>\n",
       "      <td>2.013550e-20</td>\n",
       "      <td>0.606566</td>\n",
       "      <td>0.343122</td>\n",
       "      <td>5.753000e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>520</td>\n",
       "      <td>0.463029</td>\n",
       "      <td>0.017031</td>\n",
       "      <td>0.430399</td>\n",
       "      <td>0.171004</td>\n",
       "      <td>0.471073</td>\n",
       "      <td>4.337957e-05</td>\n",
       "      <td>0.342008</td>\n",
       "      <td>0.082011</td>\n",
       "      <td>5.783943e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>476</td>\n",
       "      <td>0.978302</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.788842</td>\n",
       "      <td>0.838136</td>\n",
       "      <td>0.248875</td>\n",
       "      <td>1.041236e-09</td>\n",
       "      <td>0.977790</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.601902e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>329</td>\n",
       "      <td>1.453951</td>\n",
       "      <td>0.136543</td>\n",
       "      <td>0.454941</td>\n",
       "      <td>0.335096</td>\n",
       "      <td>0.368512</td>\n",
       "      <td>6.055869e-05</td>\n",
       "      <td>0.558494</td>\n",
       "      <td>0.303429</td>\n",
       "      <td>7.569837e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.499970e-01</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.777744e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            n_genes   gwas_or    gwas_p   twas_or    twas_p     de_or  \\\n",
       "cluster_id                                                              \n",
       "0              4917  1.324797  0.000445  1.444210  0.004375  2.551013   \n",
       "1              2877  0.870493  0.225448  0.627400  0.024133  0.246658   \n",
       "2              2288  1.047981  0.682872  1.609621  0.005217  0.859728   \n",
       "3              1949  1.012280  0.899541  1.802268  0.000816  1.791460   \n",
       "4              1891  1.273429  0.041003  1.797865  0.001047  0.910112   \n",
       "5              1869  1.048663  0.698717  0.554064  0.025933  0.559047   \n",
       "6              1285  0.646181  0.017297  0.944156  1.000000  2.179576   \n",
       "7              1095  1.034995  0.804356  0.682027  0.278153  0.203759   \n",
       "8              1024  0.627987  0.026124  0.213462  0.001107  0.320125   \n",
       "9              1020  0.738109  0.125045  0.213244  0.001109  0.182561   \n",
       "10              978  0.952002  0.861848  0.845377  0.774114  0.248228   \n",
       "11              887  0.571363  0.013425  0.760318  0.547618  0.233378   \n",
       "12              884  1.067069  0.713645  1.035031  0.880011  0.443173   \n",
       "13              795  0.570028  0.020503  0.276447  0.011106  1.124579   \n",
       "14              721  1.277914  0.188717  1.051337  0.867762  7.211527   \n",
       "15              689  1.278180  0.177376  0.661768  0.394268  0.118292   \n",
       "16              520  0.463029  0.017031  0.430399  0.171004  0.471073   \n",
       "17              476  0.978302  1.000000  0.788842  0.838136  0.248875   \n",
       "18              329  1.453951  0.136543  0.454941  0.335096  0.368512   \n",
       "19               19  0.000000  1.000000  0.000000  1.000000  0.000000   \n",
       "\n",
       "                     de_p  twas_fdr_bh  gwas_fdr_bh      de_fdr_bh  \n",
       "cluster_id                                                          \n",
       "0            7.972072e-90     0.017390     0.008896   7.972072e-89  \n",
       "1            7.371443e-53     0.057630     0.375747   4.914295e-52  \n",
       "2            4.902329e-02     0.017390     0.951527   5.767446e-02  \n",
       "3            5.286182e-17     0.005545     0.999490   1.057236e-16  \n",
       "4            2.644619e-01     0.005545     0.117152   2.783810e-01  \n",
       "5            4.705380e-10     0.057630     0.951527   7.842299e-10  \n",
       "6            3.004132e-22     1.000000     0.082011   1.001377e-21  \n",
       "7            3.620710e-24     0.505732     0.999490   1.448284e-23  \n",
       "8            3.543820e-15     0.005545     0.087081   6.443309e-15  \n",
       "9            2.849244e-24     0.005545     0.303429   1.424622e-23  \n",
       "10           9.899234e-19     0.977790     0.999490   2.474808e-18  \n",
       "11           5.870878e-18     0.782311     0.082011   1.304640e-17  \n",
       "12           1.207681e-08     0.977790     0.951527   1.725259e-08  \n",
       "13           3.072798e-01     0.031733     0.082011   3.072798e-01  \n",
       "14          1.182565e-115     0.977790     0.343122  2.365130e-114  \n",
       "15           2.013550e-20     0.606566     0.343122   5.753000e-20  \n",
       "16           4.337957e-05     0.342008     0.082011   5.783943e-05  \n",
       "17           1.041236e-09     0.977790     1.000000   1.601902e-09  \n",
       "18           6.055869e-05     0.558494     0.303429   7.569837e-05  \n",
       "19           2.499970e-01     1.000000     1.000000   2.777744e-01  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def enrichment_rows_nomhc():\n",
    "    part = cgowc.part\n",
    "    df = cgowc.expression_df\n",
    "    u = set(cgowc.expression_df.index[3:]) - mhc_genes\n",
    "    for ii in range(len(part)):\n",
    "        a = set((df.index[x] for x in part[ii])) - mhc_genes\n",
    "        yield (ii,\n",
    "               len(part[ii]),\n",
    "               *fet(a, gwas_genes - mhc_genes, u),\n",
    "               *fet(a, twas_genes - mhc_genes, u),\n",
    "               *fet(a, de_genes - mhc_genes, u),\n",
    "              )\n",
    "        \n",
    "    #print(ii, 0 in part[ii], 1 in part[ii], fet(a, gwas_genes, u), fet(a, twas_genes, u), fet(a, de_genes, u), len(part[ii]))\n",
    "    \n",
    "edf2 = pd.DataFrame.from_records(enrichment_rows_nomhc(), \n",
    "                                columns=['module_id', 'n_genes', 'gwas_or', 'gwas_p', 'twas_or', 'twas_p', 'de_or', 'de_p'],\n",
    "                                index='module_id')\n",
    "edf2['twas_fdr_bh'] = multipletests(edf2['twas_p'], method='fdr_bh')[1]\n",
    "edf2['gwas_fdr_bh'] = multipletests(edf2['gwas_p'], method='fdr_bh')[1]\n",
    "edf2['de_fdr_bh'] = multipletests(edf2['de_p'], method='fdr_bh')[1]\n",
    "\n",
    "edf2[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', 'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']].to_csv('module_enrichment_excluding_mhc_region.csv')\n",
    "\n",
    "edf2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GO enrichment for each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 4688 IDs ...\n",
      " 94%  4,269 of  4,560 study items found in association\n",
      " 97%  4,560 of  4,688 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   7,554 GO terms are associated with  4,269 of  4,688 study items\n",
      "  METHOD fdr_bh:\n",
      "     385 GO terms found significant (< 0.05=alpha) (351 enriched +  34 purified): statsmodels fdr_bh\n",
      "   3,145 study items associated with significant GO IDs (enriched)\n",
      "     631 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 4688 IDs ...\n",
      " 97%  4,444 of  4,560 study items found in association\n",
      " 97%  4,560 of  4,688 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "   1,279 GO terms are associated with  4,444 of  4,688 study items\n",
      "  METHOD fdr_bh:\n",
      "     280 GO terms found significant (< 0.05=alpha) (265 enriched +  15 purified): statsmodels fdr_bh\n",
      "   4,142 study items associated with significant GO IDs (enriched)\n",
      "     640 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 4688 IDs ...\n",
      " 92%  4,202 of  4,560 study items found in association\n",
      " 97%  4,560 of  4,688 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "   2,267 GO terms are associated with  4,202 of  4,688 study items\n",
      "  METHOD fdr_bh:\n",
      "     139 GO terms found significant (< 0.05=alpha) (119 enriched +  20 purified): statsmodels fdr_bh\n",
      "   3,644 study items associated with significant GO IDs (enriched)\n",
      "     497 study items associated with significant GO IDs (purified)\n",
      "    804 items WROTE: cluster0_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 836 IDs ...\n",
      " 89%    395 of    444 study items found in association\n",
      " 53%    444 of    836 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "     793 GO terms are associated with    395 of    836 study items\n",
      "  METHOD fdr_bh:\n",
      "       7 GO terms found significant (< 0.05=alpha) (  6 enriched +   1 purified): statsmodels fdr_bh\n",
      "     200 study items associated with significant GO IDs (enriched)\n",
      "       3 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 836 IDs ...\n",
      " 82%    365 of    444 study items found in association\n",
      " 53%    444 of    836 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     211 GO terms are associated with    365 of    836 study items\n",
      "  METHOD fdr_bh:\n",
      "      17 GO terms found significant (< 0.05=alpha) (  2 enriched +  15 purified): statsmodels fdr_bh\n",
      "     114 study items associated with significant GO IDs (enriched)\n",
      "     196 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 836 IDs ...\n",
      " 54%    239 of    444 study items found in association\n",
      " 53%    444 of    836 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     285 GO terms are associated with    239 of    836 study items\n",
      "  METHOD fdr_bh:\n",
      "       6 GO terms found significant (< 0.05=alpha) (  3 enriched +   3 purified): statsmodels fdr_bh\n",
      "      45 study items associated with significant GO IDs (enriched)\n",
      "      84 study items associated with significant GO IDs (purified)\n",
      "     30 items WROTE: cluster1_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 2139 IDs ...\n",
      " 92%  1,876 of  2,040 study items found in association\n",
      " 95%  2,040 of  2,139 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   4,343 GO terms are associated with  1,876 of  2,139 study items\n",
      "  METHOD fdr_bh:\n",
      "      49 GO terms found significant (< 0.05=alpha) ( 24 enriched +  25 purified): statsmodels fdr_bh\n",
      "     725 study items associated with significant GO IDs (enriched)\n",
      "     138 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 2139 IDs ...\n",
      " 96%  1,950 of  2,040 study items found in association\n",
      " 95%  2,040 of  2,139 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     848 GO terms are associated with  1,950 of  2,139 study items\n",
      "  METHOD fdr_bh:\n",
      "      52 GO terms found significant (< 0.05=alpha) ( 30 enriched +  22 purified): statsmodels fdr_bh\n",
      "   1,581 study items associated with significant GO IDs (enriched)\n",
      "     623 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 2139 IDs ...\n",
      " 93%  1,890 of  2,040 study items found in association\n",
      " 95%  2,040 of  2,139 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "   1,382 GO terms are associated with  1,890 of  2,139 study items\n",
      "  METHOD fdr_bh:\n",
      "      30 GO terms found significant (< 0.05=alpha) ( 16 enriched +  14 purified): statsmodels fdr_bh\n",
      "   1,590 study items associated with significant GO IDs (enriched)\n",
      "      53 study items associated with significant GO IDs (purified)\n",
      "    131 items WROTE: cluster2_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 1810 IDs ...\n",
      " 92%  1,584 of  1,728 study items found in association\n",
      " 95%  1,728 of  1,810 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   4,247 GO terms are associated with  1,584 of  1,810 study items\n",
      "  METHOD fdr_bh:\n",
      "      32 GO terms found significant (< 0.05=alpha) ( 19 enriched +  13 purified): statsmodels fdr_bh\n",
      "     411 study items associated with significant GO IDs (enriched)\n",
      "      69 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 1810 IDs ...\n",
      " 96%  1,655 of  1,728 study items found in association\n",
      " 95%  1,728 of  1,810 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     781 GO terms are associated with  1,655 of  1,810 study items\n",
      "  METHOD fdr_bh:\n",
      "      37 GO terms found significant (< 0.05=alpha) ( 24 enriched +  13 purified): statsmodels fdr_bh\n",
      "   1,393 study items associated with significant GO IDs (enriched)\n",
      "     590 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 1810 IDs ...\n",
      " 92%  1,589 of  1,728 study items found in association\n",
      " 95%  1,728 of  1,810 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "   1,283 GO terms are associated with  1,589 of  1,810 study items\n",
      "  METHOD fdr_bh:\n",
      "      19 GO terms found significant (< 0.05=alpha) ( 14 enriched +   5 purified): statsmodels fdr_bh\n",
      "   1,273 study items associated with significant GO IDs (enriched)\n",
      "       9 study items associated with significant GO IDs (purified)\n",
      "     88 items WROTE: cluster3_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 1585 IDs ...\n",
      " 89%  1,259 of  1,408 study items found in association\n",
      " 89%  1,408 of  1,585 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   3,039 GO terms are associated with  1,259 of  1,585 study items\n",
      "  METHOD fdr_bh:\n",
      "      18 GO terms found significant (< 0.05=alpha) (  7 enriched +  11 purified): statsmodels fdr_bh\n",
      "     216 study items associated with significant GO IDs (enriched)\n",
      "      40 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 1585 IDs ...\n",
      " 94%  1,330 of  1,408 study items found in association\n",
      " 89%  1,408 of  1,585 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     639 GO terms are associated with  1,330 of  1,585 study items\n",
      "  METHOD fdr_bh:\n",
      "      20 GO terms found significant (< 0.05=alpha) (  6 enriched +  14 purified): statsmodels fdr_bh\n",
      "     737 study items associated with significant GO IDs (enriched)\n",
      "     471 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 1585 IDs ...\n",
      " 91%  1,286 of  1,408 study items found in association\n",
      " 89%  1,408 of  1,585 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "   1,075 GO terms are associated with  1,286 of  1,585 study items\n",
      "  METHOD fdr_bh:\n",
      "       9 GO terms found significant (< 0.05=alpha) (  5 enriched +   4 purified): statsmodels fdr_bh\n",
      "     406 study items associated with significant GO IDs (enriched)\n",
      "      10 study items associated with significant GO IDs (purified)\n",
      "     47 items WROTE: cluster4_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 1609 IDs ...\n",
      " 94%  1,407 of  1,503 study items found in association\n",
      " 93%  1,503 of  1,609 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   5,005 GO terms are associated with  1,407 of  1,609 study items\n",
      "  METHOD fdr_bh:\n",
      "     334 GO terms found significant (< 0.05=alpha) (317 enriched +  17 purified): statsmodels fdr_bh\n",
      "   1,030 study items associated with significant GO IDs (enriched)\n",
      "      55 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 1609 IDs ...\n",
      " 98%  1,470 of  1,503 study items found in association\n",
      " 93%  1,503 of  1,609 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     618 GO terms are associated with  1,470 of  1,609 study items\n",
      "  METHOD fdr_bh:\n",
      "      57 GO terms found significant (< 0.05=alpha) ( 54 enriched +   3 purified): statsmodels fdr_bh\n",
      "   1,260 study items associated with significant GO IDs (enriched)\n",
      "     249 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 1609 IDs ...\n",
      " 92%  1,389 of  1,503 study items found in association\n",
      " 93%  1,503 of  1,609 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "   1,274 GO terms are associated with  1,389 of  1,609 study items\n",
      "  METHOD fdr_bh:\n",
      "      52 GO terms found significant (< 0.05=alpha) ( 48 enriched +   4 purified): statsmodels fdr_bh\n",
      "   1,045 study items associated with significant GO IDs (enriched)\n",
      "      56 study items associated with significant GO IDs (purified)\n",
      "    443 items WROTE: cluster5_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 803 IDs ...\n",
      " 88%    564 of    639 study items found in association\n",
      " 80%    639 of    803 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   2,236 GO terms are associated with    564 of    803 study items\n",
      "  METHOD fdr_bh:\n",
      "       4 GO terms found significant (< 0.05=alpha) (  4 enriched +   0 purified): statsmodels fdr_bh\n",
      "      69 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 803 IDs ...\n",
      " 95%    605 of    639 study items found in association\n",
      " 80%    639 of    803 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     435 GO terms are associated with    605 of    803 study items\n",
      "  METHOD fdr_bh:\n",
      "      13 GO terms found significant (< 0.05=alpha) ( 10 enriched +   3 purified): statsmodels fdr_bh\n",
      "     304 study items associated with significant GO IDs (enriched)\n",
      "     179 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 803 IDs ...\n",
      " 84%    535 of    639 study items found in association\n",
      " 80%    639 of    803 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     701 GO terms are associated with    535 of    803 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  0 enriched +   1 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "     256 study items associated with significant GO IDs (purified)\n",
      "     18 items WROTE: cluster6_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 364 IDs ...\n",
      " 88%    189 of    216 study items found in association\n",
      " 59%    216 of    364 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "     878 GO terms are associated with    189 of    364 study items\n",
      "  METHOD fdr_bh:\n",
      "       3 GO terms found significant (< 0.05=alpha) (  3 enriched +   0 purified): statsmodels fdr_bh\n",
      "      25 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 364 IDs ...\n",
      " 88%    191 of    216 study items found in association\n",
      " 59%    216 of    364 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     212 GO terms are associated with    191 of    364 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 364 IDs ...\n",
      " 79%    170 of    216 study items found in association\n",
      " 59%    216 of    364 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     271 GO terms are associated with    170 of    364 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      3 items WROTE: cluster7_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 449 IDs ...\n",
      " 85%    268 of    314 study items found in association\n",
      " 70%    314 of    449 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   1,453 GO terms are associated with    268 of    449 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 449 IDs ...\n",
      " 93%    293 of    314 study items found in association\n",
      " 70%    314 of    449 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     254 GO terms are associated with    293 of    449 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 449 IDs ...\n",
      " 85%    266 of    314 study items found in association\n",
      " 70%    314 of    449 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     379 GO terms are associated with    266 of    449 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      0 items. NOT WRITING cluster8_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 238 IDs ...\n",
      " 92%     94 of    102 study items found in association\n",
      " 43%    102 of    238 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "     371 GO terms are associated with     94 of    238 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  1 enriched +   0 purified): statsmodels fdr_bh\n",
      "      29 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 238 IDs ...\n",
      " 78%     80 of    102 study items found in association\n",
      " 43%    102 of    238 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "      88 GO terms are associated with     80 of    238 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 238 IDs ...\n",
      " 68%     69 of    102 study items found in association\n",
      " 43%    102 of    238 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     116 GO terms are associated with     69 of    238 study items\n",
      "  METHOD fdr_bh:\n",
      "       2 GO terms found significant (< 0.05=alpha) (  1 enriched +   1 purified): statsmodels fdr_bh\n",
      "      15 study items associated with significant GO IDs (enriched)\n",
      "      20 study items associated with significant GO IDs (purified)\n",
      "      3 items WROTE: cluster9_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 410 IDs ...\n",
      " 88%    205 of    232 study items found in association\n",
      " 57%    232 of    410 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "     853 GO terms are associated with    205 of    410 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 410 IDs ...\n",
      " 91%    212 of    232 study items found in association\n",
      " 57%    232 of    410 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     221 GO terms are associated with    212 of    410 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 410 IDs ...\n",
      " 82%    191 of    232 study items found in association\n",
      " 57%    232 of    410 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     316 GO terms are associated with    190 of    410 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      0 items. NOT WRITING cluster10_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 461 IDs ...\n",
      " 88%    251 of    285 study items found in association\n",
      " 62%    285 of    461 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   1,343 GO terms are associated with    251 of    461 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 461 IDs ...\n",
      " 94%    269 of    285 study items found in association\n",
      " 62%    285 of    461 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     226 GO terms are associated with    269 of    461 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  0 enriched +   1 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       3 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 461 IDs ...\n",
      " 89%    253 of    285 study items found in association\n",
      " 62%    285 of    461 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     352 GO terms are associated with    253 of    461 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  1 enriched +   0 purified): statsmodels fdr_bh\n",
      "      12 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      2 items WROTE: cluster11_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 517 IDs ...\n",
      " 86%    320 of    370 study items found in association\n",
      " 72%    370 of    517 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   1,427 GO terms are associated with    320 of    517 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  1 enriched +   0 purified): statsmodels fdr_bh\n",
      "       7 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 517 IDs ...\n",
      " 94%    348 of    370 study items found in association\n",
      " 72%    370 of    517 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     283 GO terms are associated with    348 of    517 study items\n",
      "  METHOD fdr_bh:\n",
      "       2 GO terms found significant (< 0.05=alpha) (  1 enriched +   1 purified): statsmodels fdr_bh\n",
      "       4 study items associated with significant GO IDs (enriched)\n",
      "      47 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 517 IDs ...\n",
      " 90%    334 of    370 study items found in association\n",
      " 72%    370 of    517 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     466 GO terms are associated with    334 of    517 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      3 items WROTE: cluster12_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 397 IDs ...\n",
      " 88%    236 of    268 study items found in association\n",
      " 68%    268 of    397 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   1,143 GO terms are associated with    236 of    397 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  1 enriched +   0 purified): statsmodels fdr_bh\n",
      "       5 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 397 IDs ...\n",
      " 95%    254 of    268 study items found in association\n",
      " 68%    268 of    397 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     247 GO terms are associated with    254 of    397 study items\n",
      "  METHOD fdr_bh:\n",
      "       2 GO terms found significant (< 0.05=alpha) (  1 enriched +   1 purified): statsmodels fdr_bh\n",
      "      50 study items associated with significant GO IDs (enriched)\n",
      "      21 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 397 IDs ...\n",
      " 88%    235 of    268 study items found in association\n",
      " 68%    268 of    397 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     414 GO terms are associated with    235 of    397 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      3 items WROTE: cluster13_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 556 IDs ...\n",
      " 91%    441 of    482 study items found in association\n",
      " 87%    482 of    556 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   2,135 GO terms are associated with    441 of    556 study items\n",
      "  METHOD fdr_bh:\n",
      "       9 GO terms found significant (< 0.05=alpha) (  9 enriched +   0 purified): statsmodels fdr_bh\n",
      "      73 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 556 IDs ...\n",
      " 96%    463 of    482 study items found in association\n",
      " 87%    482 of    556 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     415 GO terms are associated with    463 of    556 study items\n",
      "  METHOD fdr_bh:\n",
      "      26 GO terms found significant (< 0.05=alpha) ( 23 enriched +   3 purified): statsmodels fdr_bh\n",
      "     321 study items associated with significant GO IDs (enriched)\n",
      "     112 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 556 IDs ...\n",
      " 91%    438 of    482 study items found in association\n",
      " 87%    482 of    556 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     620 GO terms are associated with    438 of    556 study items\n",
      "  METHOD fdr_bh:\n",
      "       6 GO terms found significant (< 0.05=alpha) (  5 enriched +   1 purified): statsmodels fdr_bh\n",
      "      32 study items associated with significant GO IDs (enriched)\n",
      "       9 study items associated with significant GO IDs (purified)\n",
      "     41 items WROTE: cluster14_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 262 IDs ...\n",
      " 88%    150 of    171 study items found in association\n",
      " 65%    171 of    262 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "     921 GO terms are associated with    150 of    262 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 262 IDs ...\n",
      " 89%    153 of    171 study items found in association\n",
      " 65%    171 of    262 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     189 GO terms are associated with    153 of    262 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 262 IDs ...\n",
      " 80%    136 of    171 study items found in association\n",
      " 65%    171 of    262 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     230 GO terms are associated with    136 of    262 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      0 items. NOT WRITING cluster15_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 411 IDs ...\n",
      " 80%    260 of    324 study items found in association\n",
      " 79%    324 of    411 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "   1,286 GO terms are associated with    260 of    411 study items\n",
      "  METHOD fdr_bh:\n",
      "      15 GO terms found significant (< 0.05=alpha) ( 15 enriched +   0 purified): statsmodels fdr_bh\n",
      "      55 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 411 IDs ...\n",
      " 93%    300 of    324 study items found in association\n",
      " 79%    324 of    411 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     256 GO terms are associated with    300 of    411 study items\n",
      "  METHOD fdr_bh:\n",
      "      15 GO terms found significant (< 0.05=alpha) ( 11 enriched +   4 purified): statsmodels fdr_bh\n",
      "      73 study items associated with significant GO IDs (enriched)\n",
      "     101 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 411 IDs ...\n",
      " 82%    266 of    324 study items found in association\n",
      " 79%    324 of    411 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     383 GO terms are associated with    266 of    411 study items\n",
      "  METHOD fdr_bh:\n",
      "       5 GO terms found significant (< 0.05=alpha) (  4 enriched +   1 purified): statsmodels fdr_bh\n",
      "      13 study items associated with significant GO IDs (enriched)\n",
      "       2 study items associated with significant GO IDs (purified)\n",
      "     35 items WROTE: cluster16_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 138 IDs ...\n",
      " 79%     59 of     75 study items found in association\n",
      " 54%     75 of    138 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "     298 GO terms are associated with     59 of    138 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 138 IDs ...\n",
      " 92%     69 of     75 study items found in association\n",
      " 54%     75 of    138 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "      84 GO terms are associated with     69 of    138 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 138 IDs ...\n",
      " 76%     57 of     75 study items found in association\n",
      " 54%     75 of    138 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "      87 GO terms are associated with     57 of    138 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      0 items. NOT WRITING cluster17_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 133 IDs ...\n",
      " 89%     81 of     91 study items found in association\n",
      " 68%     91 of    133 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "     404 GO terms are associated with     81 of    133 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 133 IDs ...\n",
      " 92%     84 of     91 study items found in association\n",
      " 68%     91 of    133 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "     101 GO terms are associated with     84 of    133 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 133 IDs ...\n",
      " 78%     71 of     91 study items found in association\n",
      " 68%     91 of    133 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "     158 GO terms are associated with     71 of    133 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      0 items. NOT WRITING cluster18_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 14 IDs ...\n",
      "100%     14 of     14 study items found in association\n",
      "100%     14 of     14 study items found in population(20517)\n",
      "Calculating 12,372 uncorrected p-values using fisher_scipy_stats\n",
      "  12,372 GO terms are associated with 18,499 of 20,517 population items\n",
      "      49 GO terms are associated with     14 of     14 study items\n",
      "  METHOD fdr_bh:\n",
      "       5 GO terms found significant (< 0.05=alpha) (  5 enriched +   0 purified): statsmodels fdr_bh\n",
      "      10 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 14 IDs ...\n",
      "100%     14 of     14 study items found in association\n",
      "100%     14 of     14 study items found in population(20517)\n",
      "Calculating 1,756 uncorrected p-values using fisher_scipy_stats\n",
      "   1,756 GO terms are associated with 19,334 of 20,517 population items\n",
      "      23 GO terms are associated with     14 of     14 study items\n",
      "  METHOD fdr_bh:\n",
      "       6 GO terms found significant (< 0.05=alpha) (  6 enriched +   0 purified): statsmodels fdr_bh\n",
      "      12 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 14 IDs ...\n",
      " 86%     12 of     14 study items found in association\n",
      "100%     14 of     14 study items found in population(20517)\n",
      "Calculating 4,207 uncorrected p-values using fisher_scipy_stats\n",
      "   4,207 GO terms are associated with 17,536 of 20,517 population items\n",
      "      15 GO terms are associated with     12 of     14 study items\n",
      "  METHOD fdr_bh:\n",
      "       3 GO terms found significant (< 0.05=alpha) (  3 enriched +   0 purified): statsmodels fdr_bh\n",
      "       9 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "     14 items WROTE: cluster19_go_enrichment.tsv\n"
     ]
    }
   ],
   "source": [
    "# Run pipeline for each cluster\n",
    "for cluster_id in range(len(cgowc.part)):\n",
    "    cgowc.pipeline(cluster_id, 'module%d' % cluster_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/ceph/users/apua/projects/v2_caudate_gene_vae_394/disvae/model/embedding/leiden/_h'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
