{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################\n",
    "### libraries ####\n",
    "##################\n",
    "\n",
    "suppressMessages({library(data.table)\n",
    "                  library(MatrixEQTL)\n",
    "                  library(reticulate)\n",
    "                  library(dplyr)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read parquet file using python\n",
    "use_python(\"/usr/bin/python\")\n",
    "pandas <- import(\"pandas\")\n",
    "\n",
    "read_parquet <- function(path, columns = NULL){\n",
    "    path = path.expand(path)\n",
    "    path = normalizePath(path)\n",
    "    if(!is.null(columns)) columns = as.list(columns)\n",
    "    xdf = pandas$read_parquet(path, engine='fastparquet', columns=columns)\n",
    "    xdf = tibble::rownames_to_column(xdf, var='Name')\n",
    "    return(tibble::as_tibble(xdf))\n",
    "}\n",
    "\n",
    "## getRPKM based on recount @ https://github.com/leekgroup/recount\n",
    "## Example include using Summary Object:jRp10m = recount::getRPKM(rse_jxn, 'Length')\n",
    "getRPKM <- function(counts, length_var) {\n",
    "    mapped <- colSums(counts)\n",
    "    bg <- matrix(mapped, ncol = ncol(counts), nrow = nrow(counts), byrow = TRUE)\n",
    "    len <- length_var\n",
    "    wid <- matrix(len, nrow = nrow(counts), ncol = ncol(counts), byrow = FALSE)\n",
    "    counts / (wid/1000) / (bg/1e6)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################\n",
    "# statistical model ##\n",
    "######################\n",
    "\n",
    "cov_file = paste0(\"/ceph/projects/v3_phase3_paper/inputs/gtex_v8/covariates/\",\n",
    "                  \"_m/GTEx_Analysis_v8_eQTL_covariates/\",\n",
    "                  \"Brain_Caudate_basal_ganglia.v8.covariates.txt\")\n",
    "cov <- fread(cov_file, data.table=FALSE)\n",
    "rownames(cov) <- cov[, 1]\n",
    "cov <- cov[, -1]\n",
    "cov <- as.data.frame(t(cov))\n",
    "\n",
    "dropList = c('GTEX-11UD1', 'GTEX-12ZZW', 'GTEX-139TS', 'GTEX-13PLJ', 'GTEX-13PVQ',\n",
    "             'GTEX-1477Z', 'GTEX-17MF6', 'GTEX-1EH9U', 'GTEX-1F75B', 'GTEX-1RNSC',\n",
    "             'GTEX-QDT8', 'GTEX-X261', 'GTEX-ZV68')\n",
    "cov = cov[!(rownames(cov) %in% dropList), ]\n",
    "\n",
    "mod = model.matrix(~pcr + platform + sex + PC1 + PC2 + PC3 + PC4 + PC5, data=cov) \n",
    "\n",
    "#######################\n",
    "#### extract PCA ######\n",
    "#######################\n",
    "\n",
    "PCs = cov[, grep(\"InferredCov*\", colnames(cov))]\n",
    "\n",
    "##################\n",
    "### Covariates ###\n",
    "##################\n",
    "\n",
    "covs = t(cbind(mod[,-1],PCs))\n",
    "save(covs, file=\"covariates.rda\")\n",
    "\n",
    "covs = SlicedData$new(covs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################\n",
    "### load pheno ###\n",
    "##################\n",
    "\n",
    "pheno_file <- paste0(\"/ceph/projects/v3_phase3_paper/inputs/\",\n",
    "                     \"gtex_v8/phenotypes/_m/gtex_v8_sample_data.txt\")\n",
    "pd <- fread(pheno_file, data.table=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "######################\n",
    "### load expression ##\n",
    "######################\n",
    "\n",
    "basename = \"/ceph/projects/v3_phase3_paper/inputs/gtex_v8/expression\"\n",
    "gene_file = paste0(basename,\"/_m/genes_gtex_v8_counts.txt.gz\")\n",
    "tx_file = paste0(basename,\"/_m/transcripts_gtex_v8_tpm.txt.gz\")\n",
    "jxn_file = paste0(basename,\"/_m/junctions_gtex_v8_count.txt.gz\")\n",
    "exon_file = \"/ceph/tmp/tmp_gtex/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_exon_reads.parquet\"\n",
    "\n",
    "## read files\n",
    "rse_gene = fread(gene_file, data.table=FALSE)\n",
    "rse_jxn = fread(jxn_file, data.table=FALSE)\n",
    "rse_tx = fread(tx_file, data.table=FALSE)\n",
    "rse_exon = read_parquet(exon_file)\n",
    "\n",
    "gene_annot = rse_gene[, 1:2]\n",
    "exon_annot = rse_exon[, 1:2]\n",
    "jxn_annot = rse_jxn[, 1:2]\n",
    "tx_annot = rse_tx[, 1:2]\n",
    "\n",
    "## Subset and keep caudate samples\n",
    "rse_gene = rse_gene[, (colnames(rse_gene) %in% c('Name',pd$SAMPID))]\n",
    "rse_exon = rse_exon[, (colnames(rse_exon) %in% c('Name', pd$SAMPID))]\n",
    "rse_jxn = rse_jxn[, (colnames(rse_jxn) %in% c('Name', pd$SAMPID))]\n",
    "rse_tx = rse_tx[, (colnames(rse_tx) %in% c('transcript_id', pd$SAMPID))]\n",
    "\n",
    "## Remove duplicates\n",
    "rse_gene = rse_gene[!(duplicated(rse_gene$Name)), ]\n",
    "rse_exon = rse_exon[!(duplicated(rse_exon$Name)), ]\n",
    "rse_jxn = rse_jxn[!(duplicated(rse_jxn$Name)), ]\n",
    "rse_tx = rse_tx[!(duplicated(rse_tx$transcript_id)), ]\n",
    "\n",
    "## Add rownames\n",
    "if(tibble::has_rownames(rse_gene)){\n",
    "    rse_gene = tibble::remove_rownames(rse_gene)\n",
    "} \n",
    "rse_gene = tibble::column_to_rownames(rse_gene, var=\"Name\")\n",
    "if(tibble::has_rownames(rse_exon)){\n",
    "    rse_exon = tibble::remove_rownames(rse_exon)\n",
    "}\n",
    "rse_exon = tibble::column_to_rownames(rse_exon, var=\"Name\")\n",
    "if(tibble::has_rownames(rse_jxn)){\n",
    "    rse_jxn = tibble::remove_rownames(rse_jxn)\n",
    "}\n",
    "rse_jxn = tibble::column_to_rownames(rse_jxn, var=\"Name\")\n",
    "if(tibble::has_rownames(rse_tx)){\n",
    "    rse_tx = tibble::remove_rownames(rse_tx)\n",
    "}\n",
    "rse_tx = tibble::column_to_rownames(rse_tx, var=\"transcript_id\")\n",
    "\n",
    "## filter, based on brainseq phase2\n",
    "rse_gene = rse_gene[rowMeans(x=rse_gene, na.rm = TRUE) > 0.2,]\n",
    "rse_exon = rse_exon[rowMeans(x=rse_exon, na.rm = TRUE) > 0.2,]\n",
    "rse_jxn = rse_jxn[rowMeans(x=rse_jxn, na.rm = TRUE) > 0.4,]\n",
    "rse_tx = rse_tx[rowMeans(x=rse_tx, na.rm = TRUE) > 0.4,]\n",
    "\n",
    "## Feature Length\n",
    "gene_len = fread(paste0('/ceph/projects/v3_phase3_paper/inputs/gtex_v8/',\n",
    "                        'annotation/feature_lengths/_m/gene_annotation.txt'), \n",
    "                 data.table=F)\n",
    "exon_len = fread(paste0(\"/ceph/projects/v3_phase3_paper/inputs/gtex_v8/\", \n",
    "                        \"annotation/feature_lengths/_m/exon_annotation.txt\"), \n",
    "                 data.table=F)\n",
    "gene_len = gene_len[(gene_len$gene_id %in% rownames(rse_gene)), ]\n",
    "exon_len = exon_len[(exon_len$exon_id %in% rownames(rse_exon)), ]\n",
    "\n",
    "## extract rpkms\n",
    "geneRpkm = getRPKM(rse_gene, gene_len$length)\n",
    "exonRpkm = getRPKM(rse_exon, exon_len$length)\n",
    "jxnRp10m = getRPKM(rse_jxn, rep(100, dim(rse_jxn)[1]))\n",
    "txTpm = rse_tx #transcripts are in TPM\n",
    "\n",
    "## Normalize expression\n",
    "geneExpression = log2(geneRpkm+1)\n",
    "exonExpression = log2(exonRpkm+1)\n",
    "jxnExpression = log2(jxnRp10m+1)\n",
    "txExpression = log2(txTpm+1)\n",
    "\n",
    "## Rename sample by SUBJID\n",
    "colnames(geneExpression) <- pd$SUBJID[match(colnames(geneExpression), pd$SAMPID)]\n",
    "colnames(exonExpression) <- pd$SUBJID[match(colnames(exonExpression), pd$SAMPID)]\n",
    "colnames(jxnExpression) <- pd$SUBJID[match(colnames(jxnExpression), pd$SAMPID)]\n",
    "colnames(txExpression) <- pd$SUBJID[match(colnames(txExpression), pd$SAMPID)]\n",
    "\n",
    "## Match sample ids with covs\n",
    "geneExpression = geneExpression[,(colnames(geneExpression) %in% colnames(covs))]\n",
    "exonExpression = exonExpression[,(colnames(exonExpression) %in% colnames(covs))]\n",
    "jxnExpression = jxnExpression[,(colnames(jxnExpression) %in% colnames(covs))]\n",
    "txExpression = txExpression[,(colnames(txExpression) %in% colnames(covs))]\n",
    "\n",
    "## Check colnames match\n",
    "all((colnames(geneExpression) == colnames(covs)) & \n",
    "    (colnames(exonExpression) == colnames(covs)) & \n",
    "    (colnames(jxnExpression) == colnames(covs)) & \n",
    "    (colnames(txExpression) == colnames(covs)))\n",
    "\n",
    "## save expression\n",
    "save(geneExpression, exonExpression, jxnExpression, txExpression, file=\"expression.rda\")\n",
    "#load(\"expression.rda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "######################\n",
    "### snp data ####\n",
    "######################\n",
    "\n",
    "snp_file <- paste0(\"/ceph/projects/v3_phase3_paper/inputs/gtex_v8/\",\n",
    "                   \"genotypes_chr11/subset_caudate/a_transpose/\",\n",
    "                   \"_m/gtex_v8_chr11.traw\")\n",
    "snp <- fread(snp_file, data.table=FALSE)\n",
    "rownames(snp) <- snp$SNP\n",
    "colnames(snp) <- gsub('_.*', '', colnames(snp))\n",
    "\n",
    "snpMap <- snp[,c(\"SNP\", \"CHR\", \"POS\")]\n",
    "snp_annot <- snp[, 1:6]\n",
    "snp <- snp[, -1:-6]\n",
    "\n",
    "## Match samples\n",
    "snp = snp[, (colnames(snp) %in% colnames(covs))]\n",
    "all(colnames(snp) == colnames(covs))\n",
    "\n",
    "######################\n",
    "# create SNP objects #\n",
    "######################\n",
    "\n",
    "theSnps = SlicedData$new(as.matrix(snp))\n",
    "theSnps$ResliceCombined(sliceSize = 50000)\n",
    "\n",
    "snpspos = snpMap[,c(\"SNP\",\"CHR\",\"POS\")]\n",
    "colnames(snpspos) = c(\"name\",\"chr\",\"pos\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "basename = \"/ceph/projects/v3_phase3_paper/inputs/gtex_v8/expression\"\n",
    "gene_file = paste0(basename,\"/_m/genes_gtex_v8_counts.txt.gz\")\n",
    "tx_file = paste0(basename,\"/_m/transcripts_gtex_v8_tpm.txt.gz\")\n",
    "jxn_file = paste0(basename,\"/_m/junctions_gtex_v8_count.txt.gz\")\n",
    "exon_file = \"/ceph/tmp/tmp_gtex/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_exon_reads.parquet\"\n",
    "\n",
    "## read files\n",
    "rse_gene = fread(gene_file, data.table=FALSE)\n",
    "rse_jxn = fread(jxn_file, data.table=FALSE)\n",
    "rse_tx = fread(tx_file, data.table=FALSE)\n",
    "rse_exon = read_parquet(exon_file)\n",
    "\n",
    "gene_annot = rse_gene[, 1:2]\n",
    "exon_annot = rse_exon[, 1:2]\n",
    "jxn_annot = rse_jxn[, 1:2]\n",
    "tx_annot = rse_tx[, 1:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "###################\n",
    "### Subset DRD2 ###\n",
    "###################\n",
    "\n",
    "drd2_gene = subset(gene_annot, Description=='DRD2')\n",
    "drd2_exon = subset(exon_annot, Description=='DRD2')\n",
    "drd2_jxn = subset(jxn_annot, Description=='ENSG00000149295.13')\n",
    "drd2_tx = subset(tx_annot, gene_id=='ENSG00000149295.13')\n",
    "\n",
    "##### DRD2 region\n",
    "drd2_upstream = gene_len[grepl('DRD2', gene_len$gene_name), ]$start - .5e6\n",
    "drd2_downstream = gene_len[grepl('DRD2', gene_len$gene_name), ]$end + .5e6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################\n",
    "### feature annotation ###\n",
    "##########################\n",
    "\n",
    "##### gene level\n",
    "posGene = gene_len[gene_len$chr == 'chr11', c('gene_id', 'chr', 'start', 'end')]\n",
    "posGene$chr = as.integer(gsub(\"chr\", \"\", posGene$chr))\n",
    "\n",
    "##### transcript level\n",
    "tx_len = fread(paste0('/ceph/projects/v3_phase3_paper/inputs/gtex_v8/',\n",
    "                      'annotation/feature_lengths/_m/transcript_annotation.txt'),\n",
    "               data.table=FALSE)\n",
    "posTx = tx_len[tx_len$chr == 'chr11', c('transcript_id', 'chr', 'start', 'end')]\n",
    "posTx$chr = as.integer(gsub(\"chr\", \"\", posTx$chr))\n",
    "\n",
    "##### exon level\n",
    "posExon = exon_len[exon_len$chr == 'chr11', c('exon_id', 'chr', 'start', 'end')]\n",
    "posExon$chr = as.integer(gsub(\"chr\", \"\", posExon$chr))\n",
    "\n",
    "##### junction level \n",
    "jxnPos = as.data.frame(matrix(unlist(strsplit(rownames(jxnExpression), \"_\")), \n",
    "                              ncol=3, byrow=TRUE))\n",
    "names(jxnPos) = c(\"chr\", \"start\",\"end\")\n",
    "jxnPos$name = rownames(jxnExpression)\n",
    "jxnPos$start = as.integer(as.character(jxnPos$start))\n",
    "jxnPos$end = as.integer(as.character(jxnPos$end))\n",
    "posJxn = jxnPos[jxnPos$chr == 'chr11', ]\n",
    "posJxn = posJxn[, c(4,1:3)]\n",
    "posJxn = subset(posJxn, !(chr %in% c('chrX', 'chrY', 'chrM')))\n",
    "posJxn$chr = as.integer(as.character(gsub(\"chr\", \"\", posJxn$chr)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################\n",
    "### sliced expression data ##\n",
    "geneSlice = SlicedData$new(as.matrix(geneExpression))\n",
    "exonSlice = SlicedData$new(as.matrix(exonExpression))\n",
    "jxnSlice = SlicedData$new(as.matrix(jxnExpression))\n",
    "txSlice = SlicedData$new(as.matrix(txExpression))\n",
    "\n",
    "geneSlice$ResliceCombined(sliceSize = 5000)\n",
    "exonSlice$ResliceCombined(sliceSize = 5000)\n",
    "jxnSlice$ResliceCombined(sliceSize = 5000)\n",
    "txSlice$ResliceCombined(sliceSize = 5000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Begin eQTL analysis\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Matching data files and location files\n",
      "\n",
      "1 of 37987 genes matched\n",
      "\n",
      "3128903 of 3128903 SNPs matched\n",
      "\n",
      "\n",
      "Task finished in 1.52 seconds\n",
      "\n",
      "Reordering genes\n",
      "\n",
      "Task finished in 3.672 seconds\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.025 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.435 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Task finished in 0.039 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "82.73% done, 36 cis-eQTLs\n",
      "\n",
      "84.32% done, 95 cis-eQTLs\n",
      "\n",
      "85.91% done, 95 cis-eQTLs\n",
      "\n",
      "87.50% done, 95 cis-eQTLs\n",
      "\n",
      "89.08% done, 95 cis-eQTLs\n",
      "\n",
      "90.67% done, 95 cis-eQTLs\n",
      "\n",
      "92.26% done, 95 cis-eQTLs\n",
      "\n",
      "93.84% done, 95 cis-eQTLs\n",
      "\n",
      "95.43% done, 95 cis-eQTLs\n",
      "\n",
      "97.02% done, 95 cis-eQTLs\n",
      "\n",
      "98.61% done, 95 cis-eQTLs\n",
      "\n",
      "Task finished in 257.908 seconds\n",
      "\n",
      "\n",
      "\n",
      "Matching data files and location files\n",
      "\n",
      "10 of 245418 genes matched\n",
      "\n",
      "3128903 of 3128903 SNPs matched\n",
      "\n",
      "\n",
      "Task finished in 1.151 seconds\n",
      "\n",
      "Reordering genes\n",
      "\n",
      "Task finished in 7.105 seconds\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.008 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 2.315 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Task finished in 0.04 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "82.57% done, 377 cis-eQTLs\n",
      "\n",
      "84.15% done, 1,393 cis-eQTLs\n",
      "\n",
      "85.74% done, 1,393 cis-eQTLs\n",
      "\n",
      "87.33% done, 1,393 cis-eQTLs\n",
      "\n",
      "88.92% done, 1,393 cis-eQTLs\n",
      "\n",
      "90.50% done, 1,393 cis-eQTLs\n",
      "\n",
      "92.09% done, 1,393 cis-eQTLs\n",
      "\n",
      "93.68% done, 1,393 cis-eQTLs\n",
      "\n",
      "95.26% done, 1,393 cis-eQTLs\n",
      "\n",
      "96.85% done, 1,393 cis-eQTLs\n",
      "\n",
      "98.44% done, 1,393 cis-eQTLs\n",
      "\n",
      "Task finished in 248.225 seconds\n",
      "\n",
      "\n",
      "\n",
      "Matching data files and location files\n",
      "\n",
      "14 of 210614 genes matched\n",
      "\n",
      "3128903 of 3128903 SNPs matched\n",
      "\n",
      "\n",
      "Task finished in 1.106 seconds\n",
      "\n",
      "Reordering genes\n",
      "\n",
      "Task finished in 5.933 seconds\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.006 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 1.702 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Task finished in 0.029 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "82.57% done, 737 cis-eQTLs\n",
      "\n",
      "84.16% done, 2,159 cis-eQTLs\n",
      "\n",
      "85.75% done, 2,159 cis-eQTLs\n",
      "\n",
      "87.33% done, 2,159 cis-eQTLs\n",
      "\n",
      "88.92% done, 2,159 cis-eQTLs\n",
      "\n",
      "90.51% done, 2,159 cis-eQTLs\n",
      "\n",
      "92.10% done, 2,159 cis-eQTLs\n",
      "\n",
      "93.68% done, 2,159 cis-eQTLs\n",
      "\n",
      "95.27% done, 2,159 cis-eQTLs\n",
      "\n",
      "96.86% done, 2,159 cis-eQTLs\n",
      "\n",
      "98.44% done, 2,159 cis-eQTLs\n",
      "\n",
      "Task finished in 235.819 seconds\n",
      "\n",
      "\n",
      "\n",
      "Matching data files and location files\n",
      "\n",
      "3 of 55789 genes matched\n",
      "\n",
      "3128903 of 3128903 SNPs matched\n",
      "\n",
      "\n",
      "Task finished in 0.984 seconds\n",
      "\n",
      "Reordering genes\n",
      "\n",
      "Task finished in 4.103 seconds\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.005 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.419 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Task finished in 0.026 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "82.67% done, 122 cis-eQTLs\n",
      "\n",
      "84.25% done, 397 cis-eQTLs\n",
      "\n",
      "85.84% done, 397 cis-eQTLs\n",
      "\n",
      "87.43% done, 397 cis-eQTLs\n",
      "\n",
      "89.02% done, 397 cis-eQTLs\n",
      "\n",
      "90.60% done, 397 cis-eQTLs\n",
      "\n",
      "92.19% done, 397 cis-eQTLs\n",
      "\n",
      "93.78% done, 397 cis-eQTLs\n",
      "\n",
      "95.37% done, 397 cis-eQTLs\n",
      "\n",
      "96.95% done, 397 cis-eQTLs\n",
      "\n",
      "98.54% done, 397 cis-eQTLs\n",
      "\n",
      "Task finished in 226.659 seconds\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##########################\n",
    "### Run EQTLs ############\n",
    "##########################\n",
    "print(\"Begin eQTL analysis\")\n",
    "\n",
    "meGene = Matrix_eQTL_main(snps=theSnps, gene = geneSlice, \n",
    "    cvrt = covs, output_file_name.cis =  \"cis_eqtls_genes.ctxt\" ,\n",
    "    pvOutputThreshold.cis = 1,  pvOutputThreshold=0,\n",
    "    snpspos = snpspos, genepos = posGene, \n",
    "    useModel = modelLINEAR, cisDist=5e5,\n",
    "    pvalue.hist = 100,min.pv.by.genesnp = TRUE)\t\n",
    "\n",
    "meExon = Matrix_eQTL_main(snps=theSnps, gene = exonSlice, \n",
    "    cvrt = covs, output_file_name.cis =  \"cis_eqtl_exons.ctxt\" ,\n",
    "    pvOutputThreshold.cis = 1,  pvOutputThreshold=0,\n",
    "    snpspos = snpspos, genepos = posExon, \n",
    "    useModel = modelLINEAR, cisDist=5e5,\n",
    "    pvalue.hist = 100,min.pv.by.genesnp = TRUE)\n",
    "\n",
    "meJxn = Matrix_eQTL_main(snps=theSnps, gene = jxnSlice, \n",
    "    cvrt = covs, output_file_name.cis =  \"cis_eqtl_junctions.ctxt\" ,\n",
    "    pvOutputThreshold.cis = 1,  pvOutputThreshold=0,\n",
    "    snpspos = snpspos, genepos = posJxn, \n",
    "    useModel = modelLINEAR, cisDist=5e5,\n",
    "    pvalue.hist = 100,min.pv.by.genesnp = TRUE)\t\n",
    "\n",
    "meTx = Matrix_eQTL_main(snps=theSnps, gene = txSlice, \n",
    "    cvrt = covs, output_file_name.cis =  \"cis_eqtl_trasncripts.ctxt\" ,\n",
    "    pvOutputThreshold.cis = 1,  pvOutputThreshold=0,\n",
    "    snpspos = snpspos, genepos = posTx, \n",
    "    useModel = modelLINEAR, cisDist=5e5,\n",
    "    pvalue.hist = 100,min.pv.by.genesnp = TRUE)\t\n",
    "\n",
    "ifelse(!dir.exists(\"eqtl_tables\"), dir.create(\"eqtl_tables\"), FALSE)\n",
    "save(meGene, meExon, meJxn, meTx,\n",
    "     file=\"eqtl_tables/matrixEqtl_output_gtex_caudate_4features.rda\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
