{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichment and Overlap of PGC2+CLOZUK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import os, errno\n",
    "import functools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from plotnine import *\n",
    "from pandas_plink import read_plink\n",
    "from warnings import filterwarnings\n",
    "from matplotlib.cbook import mplDeprecation\n",
    "from scipy.stats import fisher_exact, binom_test\n",
    "\n",
    "filterwarnings(\"ignore\",category=mplDeprecation)\n",
    "filterwarnings('ignore', category=UserWarning, module='plotnine.*')\n",
    "filterwarnings('ignore', category=DeprecationWarning, module='plotnine.*')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Config and Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = {\n",
    "    'biomart_file': '../_h/biomart.csv',\n",
    "    'phenotype_file': '/ceph/projects/v4_phase3_paper/inputs/phenotypes/_m/merged_phenotypes.csv',\n",
    "    'plink_file_prefix': '/ceph/projects/v4_phase3_paper/inputs/genotypes/_m/LIBD_Brain_TopMed',\n",
    "    'gwas_snp_file': '/ceph/projects/v4_phase3_paper/inputs/sz_gwas/pgc2_clozuk/map_phase3/_m/libd_hg38_pgc2sz_snps.tsv'\n",
    "}\n",
    "\n",
    "config_feature = {\n",
    "    'de_file': '../../differential_expression/_m/genes/diffExpr_szVctl_full.txt',\n",
    "    'residual_expression_file': '../../differential_expression/_m/genes/residualized_expression.tsv',\n",
    "    'fastqtl_output_file': '../../eqtl/caudate/summary_table/_m/Brainseq_LIBD_caudate_4features.signifpairs.txt.gz',\n",
    "}\n",
    "\n",
    "feature = \"genes\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def feature_map(feature):\n",
    "    return {\"genes\": \"Gene\", \"transcripts\": \"Transcript\", \n",
    "            \"exons\": \"Exon\", \"junctions\": \"Junction\"}[feature]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_de_df():\n",
    "    \"\"\"\n",
    "    Load DE analysis\n",
    "    \"\"\"\n",
    "    return pd.read_csv(config_feature['de_file'], sep='\\t', index_col=0)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_eqtl_df():\n",
    "    eqtl_df = pd.read_csv(config_feature['fastqtl_output_file'], sep='\\t')\n",
    "    return eqtl_df[(eqtl_df[\"Type\"] == feature_map(feature))]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gwas_snps():\n",
    "    return pd.read_csv(config['gwas_snp_file'], sep='\\t', index_col=0, low_memory=False)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_integration_df():\n",
    "    return get_gwas_snps().merge(get_eqtl_df(), left_on='our_snp_id', right_on='variant_id', \n",
    "                                 suffixes=['_PGC2', '_eQTL'])\\\n",
    "                          .merge(get_de_df(), left_on='gene_id', right_index=True)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_residual_expression_df():\n",
    "    return pd.read_csv(config_feature['residual_expression_file'], \n",
    "                       sep='\\t', index_col=0).transpose()\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_pheno_df():\n",
    "    return pd.read_csv(config['phenotype_file'], index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def agree_direction(row):\n",
    "    return [-1, 1][row['pgc2_a1_same_as_our_counted']] * np.sign(row['OR'] - 1) * np.sign(row['slope']) * np.sign(row['t'])\n",
    "\n",
    "\n",
    "def letter_snp(number, a0, a1):\n",
    "    '''\n",
    "    Example:\n",
    "    letter_snp(0, 'A', 'G') is 'AA'\n",
    "    letter_snp(1, 'A', 'G') is 'AG'\n",
    "    letter_snp(2, 'A', 'G') is 'GG'\n",
    "    \n",
    "    '''\n",
    "    if np.isnan(number):\n",
    "        return np.nan\n",
    "    if len(a0)==1 and len(a1)==1:\n",
    "        sep = ''\n",
    "    else:\n",
    "        sep = ' '\n",
    "    return sep.join(sorted([a0]*int(number) + [a1]*(2-int(number))))\n",
    "\n",
    "\n",
    "def get_gwas_snp(snp_id):\n",
    "    gwas = get_gwas_snps()\n",
    "    r = gwas[gwas['our_snp_id']==snp_id]\n",
    "    assert len(r) == 1\n",
    "    return r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def get_expression_and_pheno_df():\n",
    "    return pd.merge(get_pheno_df(), get_residual_expression_df(), left_index=True, right_index=True)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_plink_tuple():\n",
    "    '''\n",
    "    Usage: (bim, fam, bed) = get_plink_tuple()\n",
    "    '''\n",
    "    return read_plink(config['plink_file_prefix'])\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def subset_bed():\n",
    "    \"\"\"\n",
    "    This subsets the bed and bim file and returns the new subsetted\n",
    "    data with shared brain_ids.\n",
    "    \n",
    "    This is to speed up accessing the bed file.\n",
    "    \"\"\"\n",
    "    (bim, fam, bed) = get_plink_tuple()\n",
    "    brain_ids = list(set(get_expression_and_pheno_df()['BrNum']).intersection(set(fam['fid'])))\n",
    "    fam_pos = list(fam[(fam[\"fid\"].isin(brain_ids))].drop_duplicates(subset=\"fid\").loc[:, 'i'])\n",
    "    unique_snps = get_eqtl_df().variant_id.unique()\n",
    "    snp_info = bim[(bim[\"snp\"].isin(unique_snps))].copy()\n",
    "    snp_pos = list(snp_info.loc[:, \"i\"])\n",
    "    new_bed = bed[snp_pos].compute()[:,fam_pos]\n",
    "    new_bim = bim[(bim[\"i\"].isin(snp_pos))].reset_index(drop=True)\n",
    "    new_bim['ii'] = new_bim.index\n",
    "    return new_bed, new_bim, brain_ids\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_snp_df(snp_id):\n",
    "    '''\n",
    "    Returns a dataframe containing the genotype on snp snp_id.\n",
    "    The allele count is the same as in the plink files.\n",
    "    \n",
    "    Example: \n",
    "    get_snp_df('rs653953').head(5)\n",
    "    \n",
    "            rs653953_num rs653953_letter rs653953\n",
    "    Br5168             0              GG    0\\nGG\n",
    "    Br2582             1              AG    1\\nAG\n",
    "    Br2378             1              AG    1\\nAG\n",
    "    Br5155             2              AA    2\\nAA\n",
    "    Br5182             2              AA    2\\nAA\n",
    "    '''\n",
    "    bed, bim, brain_ids = subset_bed()\n",
    "    snp_info = bim[bim['snp']==snp_id]\n",
    "    snp_pos = snp_info.iloc[0]['ii']\n",
    "    dfsnp = pd.DataFrame(bed[[snp_pos]], columns=brain_ids, index=[snp_id + '_num']).transpose().dropna()\n",
    "    my_letter_snp = functools.partial(letter_snp, a0=snp_info.iloc[0]['a0'], a1=snp_info.iloc[0]['a1'])\n",
    "    # the 2 - in next line is to workaround a possible bug in pandas_plink? a1 and a0 inverted\n",
    "    dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']].astype('int')\n",
    "    dfsnp[snp_id + '_letter'] = dfsnp[snp_id + '_num'].apply(my_letter_snp)\n",
    "    dfsnp[snp_id] = (dfsnp[snp_id + '_num'].astype('str') + '\\n' + \n",
    "                     dfsnp[snp_id + '_letter'].astype('str')).astype('category')\n",
    "    return dfsnp\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gwas_ordered_snp_df(snp_id):\n",
    "    '''\n",
    "    Returns a dataframe containing the genotype on snp snp_id.\n",
    "    The allele count is the number of risk alleles according to GWAS.\n",
    "    \n",
    "    Example: \n",
    "    get_gwas_ordered_snp_df('rs653953').head(5)\n",
    "    \n",
    "            rs653953_num rs653953_letter rs653953\n",
    "    Br5168             2              GG    2\\nGG\n",
    "    Br2582             1              AG    1\\nAG\n",
    "    Br2378             1              AG    1\\nAG\n",
    "    Br5155             0              AA    0\\nAA\n",
    "    Br5182             0              AA    0\\nAA\n",
    "    '''\n",
    "    pgc = get_gwas_snps()\n",
    "    dfsnp = get_snp_df(snp_id).copy()\n",
    "    gwas_snp = get_gwas_snp(snp_id)\n",
    "    if gwas_snp['pgc2_a1_same_as_our_counted'].iloc[0]:\n",
    "        if gwas_snp['OR'].iloc[0] > 1:\n",
    "            pass\n",
    "        else:\n",
    "            dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']]\n",
    "    else:\n",
    "        if gwas_snp['OR'].iloc[0] > 1:\n",
    "            dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']]\n",
    "        else:\n",
    "            pass\n",
    "    dfsnp[snp_id] = (dfsnp[snp_id + '_num'].astype('str') + '\\n' + \n",
    "                     dfsnp[snp_id + '_letter'].astype('str')).astype('category')\n",
    "    return dfsnp\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_biomart_df():\n",
    "    biomart = pd.read_csv(config['biomart_file'])\n",
    "    biomart['description'] = biomart['description'].str.replace('\\[Source.*$','', regex=True)\n",
    "    return biomart\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_risk_allele(snp_id):\n",
    "    gwas_snp = get_gwas_snp(snp_id)\n",
    "    if gwas_snp['OR'].iloc[0] > 1:\n",
    "        ra = gwas_snp['A1'].iloc[0]\n",
    "    else:\n",
    "        ra = gwas_snp['A2'].iloc[0]\n",
    "    \n",
    "    return ra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_symbol(gene_id, biomart=get_biomart_df()):\n",
    "    ensge = re.sub('\\..+$','', gene_id)\n",
    "    ggg = biomart[biomart['ensembl_gene_id']==ensge]\n",
    "    if ggg.shape[0]==0:\n",
    "        return '', ''\n",
    "    gs = ggg['external_gene_name'].values[0]\n",
    "    de = ggg['description'].values[0]\n",
    "    if type(de)!=str:\n",
    "        de = ''\n",
    "    de = re.sub('\\[Source:.*$','',de)\n",
    "    return gs, de\n",
    "\n",
    "\n",
    "def save_plot(p, fn):\n",
    "    for ext in ['png', 'pdf', 'svg']:\n",
    "        p.save(fn + '.' + ext)\n",
    "        \n",
    "\n",
    "def get_snp_gene_pheno_df(snp_id, gene_id, snp_df_func):\n",
    "    pheno_columns = list(get_pheno_df().columns)\n",
    "    expr_df = get_expression_and_pheno_df()[pheno_columns + [gene_id]]\n",
    "    snp_df =  snp_df_func(snp_id)\n",
    "    return expr_df.merge(snp_df, left_on='BrNum', right_index=True)\n",
    "\n",
    "\n",
    "def simple_snp_expression_pheno_plot_impl(snp_id, gene_id, snp_df_func, pheno_var):\n",
    "    df = get_snp_gene_pheno_df(snp_id, gene_id, snp_df_func)\n",
    "    df['Dx'] = df.Dx.astype('category').cat.rename_categories({'Control': 'CTL', 'Schizo': 'SZ'})\n",
    "    y0 = df[gene_id].quantile(.01) - 0.26\n",
    "    y1 = df[gene_id].quantile(.99) + 0.26\n",
    "    pjd = position_jitterdodge(jitter_width=0.27)\n",
    "    p = ggplot(df, aes(x=snp_id, y=gene_id, fill=pheno_var)) \\\n",
    "    + geom_boxplot(alpha=0.4, outlier_alpha=0) \\\n",
    "    + geom_jitter(position=pjd, stroke=0, alpha=0.6) + ylim(y0, y1) \\\n",
    "    + labs(y='Residualized expression', fill='Diagnosis') \\\n",
    "    + theme_bw(base_size=20)\\\n",
    "    + theme(legend_title=element_text(face='bold'), \n",
    "            panel_grid_major=element_blank(), \n",
    "            panel_grid_minor=element_blank())\n",
    "    return p\n",
    "\n",
    "\n",
    "def simple_gwas_ordered_snp_expression_pheno_plot(snp_id, gene_id, pheno_var):\n",
    "    return simple_snp_expression_pheno_plot_impl(snp_id, gene_id, get_gwas_ordered_snp_df, pheno_var)\n",
    "\n",
    "\n",
    "def gwas_annotation(snp_id):\n",
    "    return 'SZ GWAS pvalue: %.1e' % get_gwas_snp(snp_id).iloc[0]['P']\n",
    "\n",
    "\n",
    "def eqtl_annotation(snp_id, gene_id):\n",
    "    r = get_eqtl_df()[(get_eqtl_df()['variant_id']==snp_id) & \n",
    "                      (get_eqtl_df()['gene_id']==gene_id)]\n",
    "    assert len(r)==1\n",
    "    return 'eQTL nominal p-value: %.1e' % r.iloc[0]['pval_nominal']\n",
    "\n",
    "\n",
    "def de_annotation(gene_id):\n",
    "    g = get_de_df()[(get_de_df()['gencodeID'] == gene_id)]\n",
    "    return 'DE adj.P.Val: %.3f' % g.iloc[0]['adj.P.Val']\n",
    "\n",
    "\n",
    "def risk_allele_annotation(snp_id):\n",
    "    return 'SZ risk allele: %s' % get_risk_allele(snp_id)\n",
    "\n",
    "\n",
    "def gwas_annotated_eqtl_pheno_plot(snp_id, gene_id, pheno_var):\n",
    "    p = simple_gwas_ordered_snp_expression_pheno_plot(snp_id, gene_id, pheno_var)                         \n",
    "    gene_symbol, gene_description = get_gene_symbol(gene_id)\n",
    "    title =\"\\n\".join([gene_symbol, gene_id,\n",
    "                      gwas_annotation(snp_id),\n",
    "                      risk_allele_annotation(snp_id),\n",
    "                      eqtl_annotation(snp_id, gene_id), \n",
    "                      de_annotation(gene_id)])\n",
    "    \n",
    "    p += ggtitle(title)\n",
    "    return p"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    os.makedirs(feature)\n",
    "except OSError as e:\n",
    "    if e.errno != errno.EEXIST:\n",
    "        raise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Enrichment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Integrate DEG with PGC2+CLOZUK SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dft = get_integration_df()\n",
    "dft.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dft['agree_direction'] = dft.apply(agree_direction, axis=1)\n",
    "agreement = {1: 'Yes', -1: 'No', 0: 0}\n",
    "dft.agree_direction = [agreement[item] for item in dft['agree_direction']]\n",
    "## Enrichment test\n",
    "table =  [[np.sum((dft['P']<5e-8) & ((dft['adj.P.Val']<.05))),\n",
    "           np.sum((dft['P']<5e-8) & ((dft['adj.P.Val']>=.05)))],\n",
    "          [np.sum((dft['P']>=5e-8) & ((dft['adj.P.Val']<.05))),\n",
    "           np.sum((dft['P']>=5e-8) & ((dft['adj.P.Val']>=.05)))]]\n",
    "print(table)\n",
    "fisher_exact(table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dft1 = dft[(dft['P']<5e-8) & ((dft['adj.P.Val']<.05))]\n",
    "df = dft1.groupby('agree_direction').size().reset_index()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "binom_test(df[0].iloc[1], df[0].sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dft2 = dft[(dft['P']<=5e-8) & (dft['adj.P.Val'] < 0.05)].copy()\n",
    "dft2.groupby(\"gene_id\").first().reset_index().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dft2['risk_allele'] = dft2['our_snp_id'].apply(get_risk_allele)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "direction = {-1: 'Down', 1: 'Up'}\n",
    "boolean_conv = {True: 1, False: -1}\n",
    "dft2.pgc2_a1_same_as_our_counted = [boolean_conv[item] for item in dft2['pgc2_a1_same_as_our_counted']]\n",
    "dft2['eqtl_gwas_dir'] = [direction[item] for item in np.sign(dft2['pgc2_a1_same_as_our_counted']) * np.sign(dft2['slope']) * np.sign(dft2['OR'] - 1)]\n",
    "dft2['de_dir'] = [direction[item] for item in np.sign(dft2['t'])]\n",
    "dft2['eqtl_slope'] = np.sign(dft2['pgc2_a1_same_as_our_counted']) * np.sign(dft2['OR'] - 1) * dft2['slope']\n",
    "dft3 = dft2[['gene_id', 'Symbol', 'variant_id', 'A1', 'A2', 'risk_allele', 'OR', \n",
    "             'P', 'pval_nominal', 'adj.P.Val', 'logFC', 't', 'eqtl_slope', \n",
    "             'de_dir', 'eqtl_gwas_dir', 'agree_direction']]\n",
    "dft3['Symbol'].fillna(dft3['gene_id'], inplace=True)\n",
    "dft3.to_csv('%s/integration_by_symbol.txt' % feature, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "agree_direction\n",
       "No     13\n",
       "Yes    22\n",
       "dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#dft3 = pd.read_csv(\"../_m/genes/integration_by_symbol.txt\", sep='\\t')\n",
    "#dft3['Symbol'].fillna(dft3['gene_id'], inplace=True)\n",
    "df2 = dft3.groupby(['gene_id']).first().reset_index().sort_values('P')\n",
    "df2.groupby(['agree_direction']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>risk_allele</th>\n",
       "      <th>OR</th>\n",
       "      <th>GWAS_P</th>\n",
       "      <th>eQTL_pvalue</th>\n",
       "      <th>de_FDR</th>\n",
       "      <th>logFC</th>\n",
       "      <th>de_t</th>\n",
       "      <th>eqtl_slope</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ZSCAN9</th>\n",
       "      <td>ENSG00000137185.11</td>\n",
       "      <td>chr6:27731058:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.26220</td>\n",
       "      <td>1.210000e-39</td>\n",
       "      <td>8.946610e-04</td>\n",
       "      <td>2.262693e-02</td>\n",
       "      <td>-0.077130</td>\n",
       "      <td>-3.126621</td>\n",
       "      <td>-0.307218</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCG4</th>\n",
       "      <td>ENSG00000176998.4</td>\n",
       "      <td>chr6:29298706:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.26360</td>\n",
       "      <td>2.500000e-39</td>\n",
       "      <td>4.443110e-05</td>\n",
       "      <td>1.090935e-02</td>\n",
       "      <td>0.231187</td>\n",
       "      <td>3.425733</td>\n",
       "      <td>0.374783</td>\n",
       "      <td>Up</td>\n",
       "      <td>Up</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRD2</th>\n",
       "      <td>ENSG00000204256.12</td>\n",
       "      <td>chr6:32669525:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.21970</td>\n",
       "      <td>6.820000e-30</td>\n",
       "      <td>1.131810e-05</td>\n",
       "      <td>1.718619e-02</td>\n",
       "      <td>-0.051562</td>\n",
       "      <td>-3.239197</td>\n",
       "      <td>-0.404029</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLOT1</th>\n",
       "      <td>ENSG00000137312.14</td>\n",
       "      <td>chr6:30406434:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.23630</td>\n",
       "      <td>2.800000e-27</td>\n",
       "      <td>6.289760e-04</td>\n",
       "      <td>1.347849e-02</td>\n",
       "      <td>-0.047316</td>\n",
       "      <td>-3.337230</td>\n",
       "      <td>-0.193544</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCG11</th>\n",
       "      <td>ENSG00000228223.2</td>\n",
       "      <td>chr6:26466161:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.91432</td>\n",
       "      <td>1.020000e-14</td>\n",
       "      <td>1.374100e-05</td>\n",
       "      <td>1.181958e-03</td>\n",
       "      <td>0.114043</td>\n",
       "      <td>4.193600</td>\n",
       "      <td>0.211170</td>\n",
       "      <td>Up</td>\n",
       "      <td>Up</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BAG6</th>\n",
       "      <td>ENSG00000204463.12</td>\n",
       "      <td>chr6:31222761:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.92265</td>\n",
       "      <td>6.790000e-14</td>\n",
       "      <td>8.199240e-05</td>\n",
       "      <td>1.241493e-02</td>\n",
       "      <td>-0.044002</td>\n",
       "      <td>-3.373864</td>\n",
       "      <td>0.148294</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NELFE</th>\n",
       "      <td>ENSG00000204356.12</td>\n",
       "      <td>chr6:31868665:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.07930</td>\n",
       "      <td>1.660000e-13</td>\n",
       "      <td>7.305820e-04</td>\n",
       "      <td>2.507792e-02</td>\n",
       "      <td>-0.046326</td>\n",
       "      <td>-3.081897</td>\n",
       "      <td>0.090267</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF204P</th>\n",
       "      <td>ENSG00000204789.4</td>\n",
       "      <td>chr6:27343879:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.93306</td>\n",
       "      <td>4.310000e-13</td>\n",
       "      <td>1.058040e-03</td>\n",
       "      <td>1.603175e-02</td>\n",
       "      <td>0.095817</td>\n",
       "      <td>3.267501</td>\n",
       "      <td>0.089913</td>\n",
       "      <td>Up</td>\n",
       "      <td>Up</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NGEF</th>\n",
       "      <td>ENSG00000066248.14</td>\n",
       "      <td>chr2:232843683:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.91758</td>\n",
       "      <td>7.130000e-13</td>\n",
       "      <td>2.415450e-04</td>\n",
       "      <td>9.144005e-03</td>\n",
       "      <td>0.104229</td>\n",
       "      <td>3.493550</td>\n",
       "      <td>-0.144167</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRRC2A</th>\n",
       "      <td>ENSG00000204469.12</td>\n",
       "      <td>chr6:31379292:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.08610</td>\n",
       "      <td>7.930000e-12</td>\n",
       "      <td>4.021460e-04</td>\n",
       "      <td>1.283462e-03</td>\n",
       "      <td>-0.066383</td>\n",
       "      <td>-4.167993</td>\n",
       "      <td>-0.131287</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLCH2</th>\n",
       "      <td>ENSG00000149527.17</td>\n",
       "      <td>chr1:2440958:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.92873</td>\n",
       "      <td>4.630000e-11</td>\n",
       "      <td>2.811440e-09</td>\n",
       "      <td>1.909382e-04</td>\n",
       "      <td>-0.169308</td>\n",
       "      <td>-4.727576</td>\n",
       "      <td>-0.174917</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP6K3</th>\n",
       "      <td>ENSG00000161896.11</td>\n",
       "      <td>chr6:33541236:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.93912</td>\n",
       "      <td>2.350000e-10</td>\n",
       "      <td>1.762100e-04</td>\n",
       "      <td>2.435252e-03</td>\n",
       "      <td>-0.233018</td>\n",
       "      <td>-3.959823</td>\n",
       "      <td>0.131471</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNAJC19</th>\n",
       "      <td>ENSG00000205981.6</td>\n",
       "      <td>chr3:181087982:C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1.07600</td>\n",
       "      <td>7.190000e-10</td>\n",
       "      <td>4.418510e-04</td>\n",
       "      <td>2.188006e-02</td>\n",
       "      <td>0.042839</td>\n",
       "      <td>3.140810</td>\n",
       "      <td>0.098382</td>\n",
       "      <td>Up</td>\n",
       "      <td>Up</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC01470</th>\n",
       "      <td>ENSG00000249484.8</td>\n",
       "      <td>chr5:152727548:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.06910</td>\n",
       "      <td>8.730000e-10</td>\n",
       "      <td>5.630250e-14</td>\n",
       "      <td>1.143030e-02</td>\n",
       "      <td>0.406227</td>\n",
       "      <td>3.407085</td>\n",
       "      <td>0.488819</td>\n",
       "      <td>Up</td>\n",
       "      <td>Up</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNNM2</th>\n",
       "      <td>ENSG00000148842.17</td>\n",
       "      <td>chr10:102825368:C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>1.06040</td>\n",
       "      <td>1.120000e-09</td>\n",
       "      <td>2.413790e-11</td>\n",
       "      <td>3.899296e-03</td>\n",
       "      <td>0.053701</td>\n",
       "      <td>3.799656</td>\n",
       "      <td>-0.169331</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF391</th>\n",
       "      <td>ENSG00000124613.8</td>\n",
       "      <td>chr6:26990749:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.94335</td>\n",
       "      <td>1.550000e-09</td>\n",
       "      <td>1.247880e-03</td>\n",
       "      <td>2.054017e-02</td>\n",
       "      <td>0.082216</td>\n",
       "      <td>3.166580</td>\n",
       "      <td>-0.145940</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLPP5</th>\n",
       "      <td>ENSG00000147535.16</td>\n",
       "      <td>chr8:38263203:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.06970</td>\n",
       "      <td>2.490000e-09</td>\n",
       "      <td>3.233620e-04</td>\n",
       "      <td>2.838093e-02</td>\n",
       "      <td>-0.073766</td>\n",
       "      <td>-3.026111</td>\n",
       "      <td>-0.209961</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTN</th>\n",
       "      <td>ENSG00000105894.11</td>\n",
       "      <td>chr7:137363413:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.94234</td>\n",
       "      <td>2.770000e-09</td>\n",
       "      <td>2.497190e-04</td>\n",
       "      <td>7.972676e-03</td>\n",
       "      <td>-0.086066</td>\n",
       "      <td>-3.545516</td>\n",
       "      <td>0.122412</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRMT61A</th>\n",
       "      <td>ENSG00000166166.12</td>\n",
       "      <td>chr14:103541799:CAGG:C</td>\n",
       "      <td>CAGG</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.93885</td>\n",
       "      <td>4.218000e-09</td>\n",
       "      <td>1.193570e-04</td>\n",
       "      <td>4.958693e-02</td>\n",
       "      <td>-0.058611</td>\n",
       "      <td>-2.775075</td>\n",
       "      <td>-0.127611</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCCB</th>\n",
       "      <td>ENSG00000114054.13</td>\n",
       "      <td>chr3:136088767:G:T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.94493</td>\n",
       "      <td>5.050000e-09</td>\n",
       "      <td>2.185890e-14</td>\n",
       "      <td>3.397384e-02</td>\n",
       "      <td>-0.058710</td>\n",
       "      <td>-2.946922</td>\n",
       "      <td>-0.246348</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPTC7</th>\n",
       "      <td>ENSG00000196850.5</td>\n",
       "      <td>chr12:110126854:A:C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1.06290</td>\n",
       "      <td>5.810000e-09</td>\n",
       "      <td>7.078260e-04</td>\n",
       "      <td>3.514250e-03</td>\n",
       "      <td>0.090141</td>\n",
       "      <td>3.837368</td>\n",
       "      <td>-0.119268</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSNARE1</th>\n",
       "      <td>ENSG00000171045.14</td>\n",
       "      <td>chr8:142261784:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.07400</td>\n",
       "      <td>1.150000e-08</td>\n",
       "      <td>4.723740e-05</td>\n",
       "      <td>4.121874e-03</td>\n",
       "      <td>-0.076791</td>\n",
       "      <td>-3.780711</td>\n",
       "      <td>0.177779</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIRIP3</th>\n",
       "      <td>ENSG00000149929.15</td>\n",
       "      <td>chr16:29959536:C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1.05670</td>\n",
       "      <td>1.190000e-08</td>\n",
       "      <td>6.256180e-04</td>\n",
       "      <td>2.093547e-04</td>\n",
       "      <td>-0.091615</td>\n",
       "      <td>-4.698442</td>\n",
       "      <td>0.139753</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPM1M</th>\n",
       "      <td>ENSG00000164088.17</td>\n",
       "      <td>chr3:52253452:G:T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1.05600</td>\n",
       "      <td>1.350000e-08</td>\n",
       "      <td>6.597780e-04</td>\n",
       "      <td>8.758407e-06</td>\n",
       "      <td>-0.136491</td>\n",
       "      <td>-5.484394</td>\n",
       "      <td>-0.116511</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VRK2</th>\n",
       "      <td>ENSG00000028116.16</td>\n",
       "      <td>chr2:57874972:T:A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.91021</td>\n",
       "      <td>1.370000e-08</td>\n",
       "      <td>6.451330e-04</td>\n",
       "      <td>1.076112e-02</td>\n",
       "      <td>-0.127117</td>\n",
       "      <td>-3.431328</td>\n",
       "      <td>0.150703</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZC3H7B</th>\n",
       "      <td>ENSG00000100403.11</td>\n",
       "      <td>chr22:41357599:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.94102</td>\n",
       "      <td>1.760000e-08</td>\n",
       "      <td>1.177050e-07</td>\n",
       "      <td>1.255326e-02</td>\n",
       "      <td>-0.055238</td>\n",
       "      <td>-3.367649</td>\n",
       "      <td>0.258939</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REEP2</th>\n",
       "      <td>ENSG00000132563.15</td>\n",
       "      <td>chr5:138362610:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.05600</td>\n",
       "      <td>1.860000e-08</td>\n",
       "      <td>3.010360e-07</td>\n",
       "      <td>6.655155e-09</td>\n",
       "      <td>0.137644</td>\n",
       "      <td>6.993625</td>\n",
       "      <td>0.145249</td>\n",
       "      <td>Up</td>\n",
       "      <td>Up</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KDM3B</th>\n",
       "      <td>ENSG00000120733.13</td>\n",
       "      <td>chr5:138362610:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.05600</td>\n",
       "      <td>1.860000e-08</td>\n",
       "      <td>1.455140e-04</td>\n",
       "      <td>1.518040e-02</td>\n",
       "      <td>-0.031591</td>\n",
       "      <td>-3.287837</td>\n",
       "      <td>-0.167676</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SREBF2</th>\n",
       "      <td>ENSG00000198911.11</td>\n",
       "      <td>chr22:41830391:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.05530</td>\n",
       "      <td>2.290000e-08</td>\n",
       "      <td>9.115040e-05</td>\n",
       "      <td>9.947151e-04</td>\n",
       "      <td>-0.075761</td>\n",
       "      <td>-4.243388</td>\n",
       "      <td>-0.177463</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BNIP3L</th>\n",
       "      <td>ENSG00000104765.15</td>\n",
       "      <td>chr8:26368096:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.93463</td>\n",
       "      <td>2.300000e-08</td>\n",
       "      <td>1.403510e-04</td>\n",
       "      <td>1.065324e-02</td>\n",
       "      <td>-0.063722</td>\n",
       "      <td>-3.435641</td>\n",
       "      <td>-0.101083</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228944.1</th>\n",
       "      <td>ENSG00000228944.1</td>\n",
       "      <td>chr7:24695385:G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.91974</td>\n",
       "      <td>2.310000e-08</td>\n",
       "      <td>1.514450e-04</td>\n",
       "      <td>2.256942e-02</td>\n",
       "      <td>0.278474</td>\n",
       "      <td>3.127826</td>\n",
       "      <td>0.412318</td>\n",
       "      <td>Up</td>\n",
       "      <td>Up</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHF1</th>\n",
       "      <td>ENSG00000112511.17</td>\n",
       "      <td>chr6:33343202:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.94764</td>\n",
       "      <td>4.060000e-08</td>\n",
       "      <td>1.541960e-04</td>\n",
       "      <td>2.281934e-02</td>\n",
       "      <td>-0.046170</td>\n",
       "      <td>-3.123441</td>\n",
       "      <td>-0.161077</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C4A</th>\n",
       "      <td>ENSG00000244731.7</td>\n",
       "      <td>chr6:31482137:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.06010</td>\n",
       "      <td>4.120000e-08</td>\n",
       "      <td>1.330480e-04</td>\n",
       "      <td>7.880375e-03</td>\n",
       "      <td>0.358141</td>\n",
       "      <td>3.550690</td>\n",
       "      <td>0.178003</td>\n",
       "      <td>Up</td>\n",
       "      <td>Up</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF14</th>\n",
       "      <td>ENSG00000105708.8</td>\n",
       "      <td>chr19:19623068:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.06130</td>\n",
       "      <td>4.300000e-08</td>\n",
       "      <td>3.436760e-06</td>\n",
       "      <td>1.653178e-02</td>\n",
       "      <td>0.074769</td>\n",
       "      <td>3.254277</td>\n",
       "      <td>0.231031</td>\n",
       "      <td>Up</td>\n",
       "      <td>Up</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000253553.5</th>\n",
       "      <td>ENSG00000253553.5</td>\n",
       "      <td>chr8:88271460:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.06200</td>\n",
       "      <td>4.410000e-08</td>\n",
       "      <td>3.018090e-04</td>\n",
       "      <td>2.909648e-02</td>\n",
       "      <td>0.202146</td>\n",
       "      <td>3.016220</td>\n",
       "      <td>-0.220670</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              gene_id              variant_id    A1 A2  \\\n",
       "Symbol                                                                   \n",
       "ZSCAN9             ENSG00000137185.11       chr6:27731058:C:T     C  T   \n",
       "HCG4                ENSG00000176998.4       chr6:29298706:G:A     G  A   \n",
       "BRD2               ENSG00000204256.12       chr6:32669525:G:A     G  A   \n",
       "FLOT1              ENSG00000137312.14       chr6:30406434:A:G     A  G   \n",
       "HCG11               ENSG00000228223.2       chr6:26466161:G:A     G  A   \n",
       "BAG6               ENSG00000204463.12       chr6:31222761:G:A     G  A   \n",
       "NELFE              ENSG00000204356.12       chr6:31868665:A:G     A  G   \n",
       "ZNF204P             ENSG00000204789.4       chr6:27343879:A:G     A  G   \n",
       "NGEF               ENSG00000066248.14      chr2:232843683:G:A     G  A   \n",
       "PRRC2A             ENSG00000204469.12       chr6:31379292:C:T     C  T   \n",
       "PLCH2              ENSG00000149527.17        chr1:2440958:A:G     A  G   \n",
       "IP6K3              ENSG00000161896.11       chr6:33541236:A:G     A  G   \n",
       "DNAJC19             ENSG00000205981.6      chr3:181087982:C:G     C  G   \n",
       "LINC01470           ENSG00000249484.8      chr5:152727548:A:G     A  G   \n",
       "CNNM2              ENSG00000148842.17     chr10:102825368:C:A     C  A   \n",
       "ZNF391              ENSG00000124613.8       chr6:26990749:T:C     T  C   \n",
       "PLPP5              ENSG00000147535.16       chr8:38263203:T:C     T  C   \n",
       "PTN                ENSG00000105894.11      chr7:137363413:T:C     T  C   \n",
       "TRMT61A            ENSG00000166166.12  chr14:103541799:CAGG:C  CAGG  C   \n",
       "PCCB               ENSG00000114054.13      chr3:136088767:G:T     G  T   \n",
       "PPTC7               ENSG00000196850.5     chr12:110126854:A:C     A  C   \n",
       "TSNARE1            ENSG00000171045.14      chr8:142261784:C:T     C  T   \n",
       "HIRIP3             ENSG00000149929.15      chr16:29959536:C:G     C  G   \n",
       "PPM1M              ENSG00000164088.17       chr3:52253452:G:T     G  T   \n",
       "VRK2               ENSG00000028116.16       chr2:57874972:T:A     T  A   \n",
       "ZC3H7B             ENSG00000100403.11      chr22:41357599:G:A     G  A   \n",
       "REEP2              ENSG00000132563.15      chr5:138362610:C:T     C  T   \n",
       "KDM3B              ENSG00000120733.13      chr5:138362610:C:T     C  T   \n",
       "SREBF2             ENSG00000198911.11      chr22:41830391:C:T     C  T   \n",
       "BNIP3L             ENSG00000104765.15       chr8:26368096:G:A     G  A   \n",
       "ENSG00000228944.1   ENSG00000228944.1       chr7:24695385:G:C     G  C   \n",
       "PHF1               ENSG00000112511.17       chr6:33343202:G:A     G  A   \n",
       "C4A                 ENSG00000244731.7       chr6:31482137:T:C     T  C   \n",
       "ZNF14               ENSG00000105708.8      chr19:19623068:T:C     T  C   \n",
       "ENSG00000253553.5   ENSG00000253553.5       chr8:88271460:A:G     A  G   \n",
       "\n",
       "                  risk_allele       OR        GWAS_P   eQTL_pvalue  \\\n",
       "Symbol                                                               \n",
       "ZSCAN9                      C  1.26220  1.210000e-39  8.946610e-04   \n",
       "HCG4                        G  1.26360  2.500000e-39  4.443110e-05   \n",
       "BRD2                        G  1.21970  6.820000e-30  1.131810e-05   \n",
       "FLOT1                       A  1.23630  2.800000e-27  6.289760e-04   \n",
       "HCG11                       A  0.91432  1.020000e-14  1.374100e-05   \n",
       "BAG6                        A  0.92265  6.790000e-14  8.199240e-05   \n",
       "NELFE                       A  1.07930  1.660000e-13  7.305820e-04   \n",
       "ZNF204P                     G  0.93306  4.310000e-13  1.058040e-03   \n",
       "NGEF                        A  0.91758  7.130000e-13  2.415450e-04   \n",
       "PRRC2A                      C  1.08610  7.930000e-12  4.021460e-04   \n",
       "PLCH2                       G  0.92873  4.630000e-11  2.811440e-09   \n",
       "IP6K3                       G  0.93912  2.350000e-10  1.762100e-04   \n",
       "DNAJC19                     C  1.07600  7.190000e-10  4.418510e-04   \n",
       "LINC01470                   A  1.06910  8.730000e-10  5.630250e-14   \n",
       "CNNM2                       C  1.06040  1.120000e-09  2.413790e-11   \n",
       "ZNF391                      C  0.94335  1.550000e-09  1.247880e-03   \n",
       "PLPP5                       T  1.06970  2.490000e-09  3.233620e-04   \n",
       "PTN                         C  0.94234  2.770000e-09  2.497190e-04   \n",
       "TRMT61A                     C  0.93885  4.218000e-09  1.193570e-04   \n",
       "PCCB                        T  0.94493  5.050000e-09  2.185890e-14   \n",
       "PPTC7                       A  1.06290  5.810000e-09  7.078260e-04   \n",
       "TSNARE1                     C  1.07400  1.150000e-08  4.723740e-05   \n",
       "HIRIP3                      C  1.05670  1.190000e-08  6.256180e-04   \n",
       "PPM1M                       G  1.05600  1.350000e-08  6.597780e-04   \n",
       "VRK2                        A  0.91021  1.370000e-08  6.451330e-04   \n",
       "ZC3H7B                      A  0.94102  1.760000e-08  1.177050e-07   \n",
       "REEP2                       C  1.05600  1.860000e-08  3.010360e-07   \n",
       "KDM3B                       C  1.05600  1.860000e-08  1.455140e-04   \n",
       "SREBF2                      C  1.05530  2.290000e-08  9.115040e-05   \n",
       "BNIP3L                      A  0.93463  2.300000e-08  1.403510e-04   \n",
       "ENSG00000228944.1           C  0.91974  2.310000e-08  1.514450e-04   \n",
       "PHF1                        A  0.94764  4.060000e-08  1.541960e-04   \n",
       "C4A                         T  1.06010  4.120000e-08  1.330480e-04   \n",
       "ZNF14                       T  1.06130  4.300000e-08  3.436760e-06   \n",
       "ENSG00000253553.5           A  1.06200  4.410000e-08  3.018090e-04   \n",
       "\n",
       "                         de_FDR     logFC      de_t  eqtl_slope de_dir  \\\n",
       "Symbol                                                                   \n",
       "ZSCAN9             2.262693e-02 -0.077130 -3.126621   -0.307218   Down   \n",
       "HCG4               1.090935e-02  0.231187  3.425733    0.374783     Up   \n",
       "BRD2               1.718619e-02 -0.051562 -3.239197   -0.404029   Down   \n",
       "FLOT1              1.347849e-02 -0.047316 -3.337230   -0.193544   Down   \n",
       "HCG11              1.181958e-03  0.114043  4.193600    0.211170     Up   \n",
       "BAG6               1.241493e-02 -0.044002 -3.373864    0.148294   Down   \n",
       "NELFE              2.507792e-02 -0.046326 -3.081897    0.090267   Down   \n",
       "ZNF204P            1.603175e-02  0.095817  3.267501    0.089913     Up   \n",
       "NGEF               9.144005e-03  0.104229  3.493550   -0.144167     Up   \n",
       "PRRC2A             1.283462e-03 -0.066383 -4.167993   -0.131287   Down   \n",
       "PLCH2              1.909382e-04 -0.169308 -4.727576   -0.174917   Down   \n",
       "IP6K3              2.435252e-03 -0.233018 -3.959823    0.131471   Down   \n",
       "DNAJC19            2.188006e-02  0.042839  3.140810    0.098382     Up   \n",
       "LINC01470          1.143030e-02  0.406227  3.407085    0.488819     Up   \n",
       "CNNM2              3.899296e-03  0.053701  3.799656   -0.169331     Up   \n",
       "ZNF391             2.054017e-02  0.082216  3.166580   -0.145940     Up   \n",
       "PLPP5              2.838093e-02 -0.073766 -3.026111   -0.209961   Down   \n",
       "PTN                7.972676e-03 -0.086066 -3.545516    0.122412   Down   \n",
       "TRMT61A            4.958693e-02 -0.058611 -2.775075   -0.127611   Down   \n",
       "PCCB               3.397384e-02 -0.058710 -2.946922   -0.246348   Down   \n",
       "PPTC7              3.514250e-03  0.090141  3.837368   -0.119268     Up   \n",
       "TSNARE1            4.121874e-03 -0.076791 -3.780711    0.177779   Down   \n",
       "HIRIP3             2.093547e-04 -0.091615 -4.698442    0.139753   Down   \n",
       "PPM1M              8.758407e-06 -0.136491 -5.484394   -0.116511   Down   \n",
       "VRK2               1.076112e-02 -0.127117 -3.431328    0.150703   Down   \n",
       "ZC3H7B             1.255326e-02 -0.055238 -3.367649    0.258939   Down   \n",
       "REEP2              6.655155e-09  0.137644  6.993625    0.145249     Up   \n",
       "KDM3B              1.518040e-02 -0.031591 -3.287837   -0.167676   Down   \n",
       "SREBF2             9.947151e-04 -0.075761 -4.243388   -0.177463   Down   \n",
       "BNIP3L             1.065324e-02 -0.063722 -3.435641   -0.101083   Down   \n",
       "ENSG00000228944.1  2.256942e-02  0.278474  3.127826    0.412318     Up   \n",
       "PHF1               2.281934e-02 -0.046170 -3.123441   -0.161077   Down   \n",
       "C4A                7.880375e-03  0.358141  3.550690    0.178003     Up   \n",
       "ZNF14              1.653178e-02  0.074769  3.254277    0.231031     Up   \n",
       "ENSG00000253553.5  2.909648e-02  0.202146  3.016220   -0.220670     Up   \n",
       "\n",
       "                  eqtl_gwas_dir agree_direction  \n",
       "Symbol                                           \n",
       "ZSCAN9                     Down             Yes  \n",
       "HCG4                         Up             Yes  \n",
       "BRD2                       Down             Yes  \n",
       "FLOT1                      Down             Yes  \n",
       "HCG11                        Up             Yes  \n",
       "BAG6                         Up              No  \n",
       "NELFE                        Up              No  \n",
       "ZNF204P                      Up             Yes  \n",
       "NGEF                       Down              No  \n",
       "PRRC2A                     Down             Yes  \n",
       "PLCH2                      Down             Yes  \n",
       "IP6K3                        Up              No  \n",
       "DNAJC19                      Up             Yes  \n",
       "LINC01470                    Up             Yes  \n",
       "CNNM2                      Down              No  \n",
       "ZNF391                     Down              No  \n",
       "PLPP5                      Down             Yes  \n",
       "PTN                          Up              No  \n",
       "TRMT61A                    Down             Yes  \n",
       "PCCB                       Down             Yes  \n",
       "PPTC7                      Down              No  \n",
       "TSNARE1                      Up              No  \n",
       "HIRIP3                       Up              No  \n",
       "PPM1M                      Down             Yes  \n",
       "VRK2                         Up              No  \n",
       "ZC3H7B                       Up              No  \n",
       "REEP2                        Up             Yes  \n",
       "KDM3B                      Down             Yes  \n",
       "SREBF2                     Down             Yes  \n",
       "BNIP3L                     Down             Yes  \n",
       "ENSG00000228944.1            Up             Yes  \n",
       "PHF1                       Down             Yes  \n",
       "C4A                          Up             Yes  \n",
       "ZNF14                        Up             Yes  \n",
       "ENSG00000253553.5          Down              No  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.set_index('Symbol').rename(columns={'t': 'de_t', 'P': 'GWAS_P', 'pval_nominal': 'eQTL_pvalue', \n",
    "                                        'adj.P.Val': 'de_FDR'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot with PGC2 risk allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for xx in range(df2.shape[0]):\n",
    "    gg = gwas_annotated_eqtl_pheno_plot(df2.iloc[xx, :].variant_id, df2.iloc[xx, :].gene_id, 'Dx')\n",
    "    print(gg)\n",
    "    label = '%s/eqtl_gwas_%s' % (feature, df2.iloc[xx, :].Symbol)\n",
    "    save_plot(gg, label)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
