{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate supplemental data of all significant integration analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>risk_allele</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_slope</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000148842.17</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102825368:C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0604</td>\n",
       "      <td>1.120000e-09</td>\n",
       "      <td>2.413790e-11</td>\n",
       "      <td>0.003899</td>\n",
       "      <td>0.053701</td>\n",
       "      <td>3.799656</td>\n",
       "      <td>-0.169331</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000148842.17</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102825648:A:T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0635</td>\n",
       "      <td>2.300000e-10</td>\n",
       "      <td>1.944530e-05</td>\n",
       "      <td>0.003899</td>\n",
       "      <td>0.053701</td>\n",
       "      <td>3.799656</td>\n",
       "      <td>-0.119926</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000148842.17</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102832918:G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0644</td>\n",
       "      <td>1.330000e-10</td>\n",
       "      <td>3.352800e-06</td>\n",
       "      <td>0.003899</td>\n",
       "      <td>0.053701</td>\n",
       "      <td>3.799656</td>\n",
       "      <td>-0.127135</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000148842.17</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102835663:T:G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0617</td>\n",
       "      <td>4.970000e-10</td>\n",
       "      <td>2.310390e-10</td>\n",
       "      <td>0.003899</td>\n",
       "      <td>0.053701</td>\n",
       "      <td>3.799656</td>\n",
       "      <td>-0.163240</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000148842.17</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102835962:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0633</td>\n",
       "      <td>2.590000e-10</td>\n",
       "      <td>3.624190e-09</td>\n",
       "      <td>0.003899</td>\n",
       "      <td>0.053701</td>\n",
       "      <td>3.799656</td>\n",
       "      <td>-0.152542</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id Symbol           variant_id A1 A2 risk_allele      OR  \\\n",
       "0  ENSG00000148842.17  CNNM2  chr10:102825368:C:A  C  A           C  1.0604   \n",
       "1  ENSG00000148842.17  CNNM2  chr10:102825648:A:T  A  T           A  1.0635   \n",
       "2  ENSG00000148842.17  CNNM2  chr10:102832918:G:C  G  C           G  1.0644   \n",
       "3  ENSG00000148842.17  CNNM2  chr10:102835663:T:G  T  G           T  1.0617   \n",
       "4  ENSG00000148842.17  CNNM2  chr10:102835962:C:T  C  T           C  1.0633   \n",
       "\n",
       "              P  pval_nominal  adj.P.Val     logFC         t  eqtl_slope  \\\n",
       "0  1.120000e-09  2.413790e-11   0.003899  0.053701  3.799656   -0.169331   \n",
       "1  2.300000e-10  1.944530e-05   0.003899  0.053701  3.799656   -0.119926   \n",
       "2  1.330000e-10  3.352800e-06   0.003899  0.053701  3.799656   -0.127135   \n",
       "3  4.970000e-10  2.310390e-10   0.003899  0.053701  3.799656   -0.163240   \n",
       "4  2.590000e-10  3.624190e-09   0.003899  0.053701  3.799656   -0.152542   \n",
       "\n",
       "  de_dir eqtl_gwas_dir agree_direction Feature_Type  \n",
       "0     Up          Down              No         Gene  \n",
       "1     Up          Down              No         Gene  \n",
       "2     Up          Down              No         Gene  \n",
       "3     Up          Down              No         Gene  \n",
       "4     Up          Down              No         Gene  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_csv('../../_m/genes/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID'})\n",
    "genes['Feature_Type'] = 'Gene'\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>risk_allele</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_slope</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000369878.8</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102825368:C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0604</td>\n",
       "      <td>1.120000e-09</td>\n",
       "      <td>4.928250e-11</td>\n",
       "      <td>0.000852</td>\n",
       "      <td>0.084962</td>\n",
       "      <td>4.753266</td>\n",
       "      <td>-0.179745</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000369878.8</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102825648:A:T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0635</td>\n",
       "      <td>2.300000e-10</td>\n",
       "      <td>7.617130e-05</td>\n",
       "      <td>0.000852</td>\n",
       "      <td>0.084962</td>\n",
       "      <td>4.753266</td>\n",
       "      <td>-0.119358</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000369878.8</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102832918:G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0644</td>\n",
       "      <td>1.330000e-10</td>\n",
       "      <td>1.354810e-05</td>\n",
       "      <td>0.000852</td>\n",
       "      <td>0.084962</td>\n",
       "      <td>4.753266</td>\n",
       "      <td>-0.128029</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000369878.8</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102835663:T:G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0617</td>\n",
       "      <td>4.970000e-10</td>\n",
       "      <td>9.108410e-11</td>\n",
       "      <td>0.000852</td>\n",
       "      <td>0.084962</td>\n",
       "      <td>4.753266</td>\n",
       "      <td>-0.179316</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000369878.8</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102835962:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0633</td>\n",
       "      <td>2.590000e-10</td>\n",
       "      <td>7.879660e-09</td>\n",
       "      <td>0.000852</td>\n",
       "      <td>0.084962</td>\n",
       "      <td>4.753266</td>\n",
       "      <td>-0.160431</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id Symbol           variant_id A1 A2 risk_allele      OR  \\\n",
       "0  ENST00000369878.8  CNNM2  chr10:102825368:C:A  C  A           C  1.0604   \n",
       "1  ENST00000369878.8  CNNM2  chr10:102825648:A:T  A  T           A  1.0635   \n",
       "2  ENST00000369878.8  CNNM2  chr10:102832918:G:C  G  C           G  1.0644   \n",
       "3  ENST00000369878.8  CNNM2  chr10:102835663:T:G  T  G           T  1.0617   \n",
       "4  ENST00000369878.8  CNNM2  chr10:102835962:C:T  C  T           C  1.0633   \n",
       "\n",
       "              P  pval_nominal  adj.P.Val     logFC         t  eqtl_slope  \\\n",
       "0  1.120000e-09  4.928250e-11   0.000852  0.084962  4.753266   -0.179745   \n",
       "1  2.300000e-10  7.617130e-05   0.000852  0.084962  4.753266   -0.119358   \n",
       "2  1.330000e-10  1.354810e-05   0.000852  0.084962  4.753266   -0.128029   \n",
       "3  4.970000e-10  9.108410e-11   0.000852  0.084962  4.753266   -0.179316   \n",
       "4  2.590000e-10  7.879660e-09   0.000852  0.084962  4.753266   -0.160431   \n",
       "\n",
       "  de_dir eqtl_gwas_dir agree_direction Feature_Type  \n",
       "0     Up          Down              No   Transcript  \n",
       "1     Up          Down              No   Transcript  \n",
       "2     Up          Down              No   Transcript  \n",
       "3     Up          Down              No   Transcript  \n",
       "4     Up          Down              No   Transcript  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans = pd.read_csv('../../_m/transcripts/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID', 'gene_name': 'Symbol'})\n",
    "trans['Feature_Type'] = 'Transcript'\n",
    "trans.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>risk_allele</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_slope</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>e598887</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102825368:C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0604</td>\n",
       "      <td>1.120000e-09</td>\n",
       "      <td>1.275560e-10</td>\n",
       "      <td>0.001824</td>\n",
       "      <td>0.08217</td>\n",
       "      <td>4.275115</td>\n",
       "      <td>-0.172984</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>e598887</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102825648:A:T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1.0635</td>\n",
       "      <td>2.300000e-10</td>\n",
       "      <td>8.685790e-05</td>\n",
       "      <td>0.001824</td>\n",
       "      <td>0.08217</td>\n",
       "      <td>4.275115</td>\n",
       "      <td>-0.117049</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>e598887</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102832918:G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1.0644</td>\n",
       "      <td>1.330000e-10</td>\n",
       "      <td>5.547170e-05</td>\n",
       "      <td>0.001824</td>\n",
       "      <td>0.08217</td>\n",
       "      <td>4.275115</td>\n",
       "      <td>-0.117347</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>e598887</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102835663:T:G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0617</td>\n",
       "      <td>4.970000e-10</td>\n",
       "      <td>5.187010e-10</td>\n",
       "      <td>0.001824</td>\n",
       "      <td>0.08217</td>\n",
       "      <td>4.275115</td>\n",
       "      <td>-0.169565</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>e598887</td>\n",
       "      <td>CNNM2</td>\n",
       "      <td>chr10:102835962:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0633</td>\n",
       "      <td>2.590000e-10</td>\n",
       "      <td>9.561270e-09</td>\n",
       "      <td>0.001824</td>\n",
       "      <td>0.08217</td>\n",
       "      <td>4.275115</td>\n",
       "      <td>-0.157295</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id Symbol           variant_id A1 A2 risk_allele      OR  \\\n",
       "0  e598887  CNNM2  chr10:102825368:C:A  C  A           C  1.0604   \n",
       "1  e598887  CNNM2  chr10:102825648:A:T  A  T           A  1.0635   \n",
       "2  e598887  CNNM2  chr10:102832918:G:C  G  C           G  1.0644   \n",
       "3  e598887  CNNM2  chr10:102835663:T:G  T  G           T  1.0617   \n",
       "4  e598887  CNNM2  chr10:102835962:C:T  C  T           C  1.0633   \n",
       "\n",
       "              P  pval_nominal  adj.P.Val    logFC         t  eqtl_slope  \\\n",
       "0  1.120000e-09  1.275560e-10   0.001824  0.08217  4.275115   -0.172984   \n",
       "1  2.300000e-10  8.685790e-05   0.001824  0.08217  4.275115   -0.117049   \n",
       "2  1.330000e-10  5.547170e-05   0.001824  0.08217  4.275115   -0.117347   \n",
       "3  4.970000e-10  5.187010e-10   0.001824  0.08217  4.275115   -0.169565   \n",
       "4  2.590000e-10  9.561270e-09   0.001824  0.08217  4.275115   -0.157295   \n",
       "\n",
       "  de_dir eqtl_gwas_dir agree_direction Feature_Type  \n",
       "0     Up          Down              No         Exon  \n",
       "1     Up          Down              No         Exon  \n",
       "2     Up          Down              No         Exon  \n",
       "3     Up          Down              No         Exon  \n",
       "4     Up          Down              No         Exon  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons = pd.read_csv('../../_m/exons/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID'})\n",
    "exons['Feature_Type'] = 'Exon'\n",
    "exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>risk_allele</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_slope</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:47180691-47182402(-)</td>\n",
       "      <td>PACSIN3</td>\n",
       "      <td>chr11:46729945:T:TAGG</td>\n",
       "      <td>T</td>\n",
       "      <td>TAGG</td>\n",
       "      <td>TAGG</td>\n",
       "      <td>0.925705</td>\n",
       "      <td>3.875000e-09</td>\n",
       "      <td>0.000221</td>\n",
       "      <td>0.022512</td>\n",
       "      <td>-0.186281</td>\n",
       "      <td>-3.419199</td>\n",
       "      <td>0.208793</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11:47180691-47182402(-)</td>\n",
       "      <td>PACSIN3</td>\n",
       "      <td>chr11:46945082:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.896840</td>\n",
       "      <td>3.720000e-08</td>\n",
       "      <td>0.000115</td>\n",
       "      <td>0.022512</td>\n",
       "      <td>-0.186281</td>\n",
       "      <td>-3.419199</td>\n",
       "      <td>0.421745</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11:113412884-113415420(-)</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>chr11:113447023:G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1.066000</td>\n",
       "      <td>4.300000e-10</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.004208</td>\n",
       "      <td>0.302443</td>\n",
       "      <td>4.079084</td>\n",
       "      <td>-0.260710</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:113412884-113415420(-)</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>chr11:113457959:A:C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1.065800</td>\n",
       "      <td>3.690000e-10</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.004208</td>\n",
       "      <td>0.302443</td>\n",
       "      <td>4.079084</td>\n",
       "      <td>-0.252745</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11:113412884-113415420(-)</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>chr11:113463378:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.064600</td>\n",
       "      <td>8.050000e-10</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.004208</td>\n",
       "      <td>0.302443</td>\n",
       "      <td>4.079084</td>\n",
       "      <td>-0.272205</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_id   Symbol             variant_id A1    A2  \\\n",
       "0    chr11:47180691-47182402(-)  PACSIN3  chr11:46729945:T:TAGG  T  TAGG   \n",
       "1    chr11:47180691-47182402(-)  PACSIN3     chr11:46945082:T:C  T     C   \n",
       "2  chr11:113412884-113415420(-)     DRD2    chr11:113447023:G:C  G     C   \n",
       "3  chr11:113412884-113415420(-)     DRD2    chr11:113457959:A:C  A     C   \n",
       "4  chr11:113412884-113415420(-)     DRD2    chr11:113463378:C:T  C     T   \n",
       "\n",
       "  risk_allele        OR             P  pval_nominal  adj.P.Val     logFC  \\\n",
       "0        TAGG  0.925705  3.875000e-09      0.000221   0.022512 -0.186281   \n",
       "1           C  0.896840  3.720000e-08      0.000115   0.022512 -0.186281   \n",
       "2           G  1.066000  4.300000e-10      0.000011   0.004208  0.302443   \n",
       "3           A  1.065800  3.690000e-10      0.000024   0.004208  0.302443   \n",
       "4           C  1.064600  8.050000e-10      0.000005   0.004208  0.302443   \n",
       "\n",
       "          t  eqtl_slope de_dir eqtl_gwas_dir agree_direction Feature_Type  \n",
       "0 -3.419199    0.208793   Down            Up              No     Junction  \n",
       "1 -3.419199    0.421745   Down            Up              No     Junction  \n",
       "2  4.079084   -0.260710     Up          Down              No     Junction  \n",
       "3  4.079084   -0.252745     Up          Down              No     Junction  \n",
       "4  4.079084   -0.272205     Up          Down              No     Junction  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "juncs = pd.read_csv('../../_m/junctions/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID'})\n",
    "juncs['Feature_Type'] = 'Junction'\n",
    "juncs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24722, 17)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature_ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>SNP_eqtl</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>risk_allele</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_t</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000143420.17</td>\n",
       "      <td>ENSA</td>\n",
       "      <td>rs12124898:150154352:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.92697</td>\n",
       "      <td>1.430000e-08</td>\n",
       "      <td>0.003994</td>\n",
       "      <td>0.017038</td>\n",
       "      <td>0.045351</td>\n",
       "      <td>3.249546</td>\n",
       "      <td>-3.926103</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000066248.14</td>\n",
       "      <td>NGEF</td>\n",
       "      <td>rs778353:233788350:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.93531</td>\n",
       "      <td>2.680000e-12</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.015120</td>\n",
       "      <td>0.109010</td>\n",
       "      <td>3.294629</td>\n",
       "      <td>-5.096546</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Feature_ID Symbol                  SNP_eqtl A1 A2 risk_allele  \\\n",
       "0  ENSG00000143420.17   ENSA  rs12124898:150154352:C:T  C  T           T   \n",
       "1  ENSG00000066248.14   NGEF    rs778353:233788350:C:T  C  T           T   \n",
       "\n",
       "        OR             P       FDR  adj.P.Val     logFC         t    eqtl_t  \\\n",
       "0  0.92697  1.430000e-08  0.003994   0.017038  0.045351  3.249546 -3.926103   \n",
       "1  0.93531  2.680000e-12  0.000037   0.015120  0.109010  3.294629 -5.096546   \n",
       "\n",
       "  de_dir eqtl_gwas_dir agree_direction Feature_Type  \n",
       "0     Up          Down              No         Gene  \n",
       "1     Up          Down              No         Gene  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([genes, trans, exons, juncs], axis=0, sort=False)\n",
    "print(df.shape)\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('BrainSeq_Phase3_Caudate_GWAS_DE_eQTL_Integration.txt.gz', index=False, header=True, sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
