{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate supplemental data of all significant integration analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature_ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>SNP_eqtl</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>risk_allele</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_t</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000143420.17</td>\n",
       "      <td>ENSA</td>\n",
       "      <td>rs12124898:150154352:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.92697</td>\n",
       "      <td>1.430000e-08</td>\n",
       "      <td>0.003994</td>\n",
       "      <td>0.017038</td>\n",
       "      <td>0.045351</td>\n",
       "      <td>3.249546</td>\n",
       "      <td>-3.926103</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000066248.14</td>\n",
       "      <td>NGEF</td>\n",
       "      <td>rs778353:233788350:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.93531</td>\n",
       "      <td>2.680000e-12</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.015120</td>\n",
       "      <td>0.109010</td>\n",
       "      <td>3.294629</td>\n",
       "      <td>-5.096546</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000066248.14</td>\n",
       "      <td>NGEF</td>\n",
       "      <td>rs1083521:233788734:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.93639</td>\n",
       "      <td>6.320000e-12</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.015120</td>\n",
       "      <td>0.109010</td>\n",
       "      <td>3.294629</td>\n",
       "      <td>-5.100574</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000066248.14</td>\n",
       "      <td>NGEF</td>\n",
       "      <td>rs1083520:233788862:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.93574</td>\n",
       "      <td>4.980000e-12</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.015120</td>\n",
       "      <td>0.109010</td>\n",
       "      <td>3.294629</td>\n",
       "      <td>-5.100574</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000066248.14</td>\n",
       "      <td>NGEF</td>\n",
       "      <td>rs1083519:233788869:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.93615</td>\n",
       "      <td>5.210000e-12</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.015120</td>\n",
       "      <td>0.109010</td>\n",
       "      <td>3.294629</td>\n",
       "      <td>-5.100574</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Feature_ID Symbol                  SNP_eqtl A1 A2 risk_allele  \\\n",
       "0  ENSG00000143420.17   ENSA  rs12124898:150154352:C:T  C  T           T   \n",
       "1  ENSG00000066248.14   NGEF    rs778353:233788350:C:T  C  T           T   \n",
       "2  ENSG00000066248.14   NGEF   rs1083521:233788734:A:G  A  G           G   \n",
       "3  ENSG00000066248.14   NGEF   rs1083520:233788862:A:G  A  G           G   \n",
       "4  ENSG00000066248.14   NGEF   rs1083519:233788869:C:T  C  T           T   \n",
       "\n",
       "        OR             P       FDR  adj.P.Val     logFC         t    eqtl_t  \\\n",
       "0  0.92697  1.430000e-08  0.003994   0.017038  0.045351  3.249546 -3.926103   \n",
       "1  0.93531  2.680000e-12  0.000037   0.015120  0.109010  3.294629 -5.096546   \n",
       "2  0.93639  6.320000e-12  0.000036   0.015120  0.109010  3.294629 -5.100574   \n",
       "3  0.93574  4.980000e-12  0.000036   0.015120  0.109010  3.294629 -5.100574   \n",
       "4  0.93615  5.210000e-12  0.000036   0.015120  0.109010  3.294629 -5.100574   \n",
       "\n",
       "  de_dir eqtl_gwas_dir agree_direction Feature_Type  \n",
       "0     Up          Down              No         Gene  \n",
       "1     Up          Down              No         Gene  \n",
       "2     Up          Down              No         Gene  \n",
       "3     Up          Down              No         Gene  \n",
       "4     Up          Down              No         Gene  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_csv('../../_m/genes/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID'})\n",
    "genes['Feature_Type'] = 'Gene'\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature_ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>SNP_eqtl</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>risk_allele</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_t</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000617168.4</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>rs414205:28526419:G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1.07930</td>\n",
       "      <td>1.070000e-14</td>\n",
       "      <td>0.002266</td>\n",
       "      <td>0.030366</td>\n",
       "      <td>-0.396302</td>\n",
       "      <td>-3.560797</td>\n",
       "      <td>-4.272345</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000617168.4</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>rs436827:28557133:T:A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1.08020</td>\n",
       "      <td>7.450000e-15</td>\n",
       "      <td>0.001155</td>\n",
       "      <td>0.030366</td>\n",
       "      <td>-0.396302</td>\n",
       "      <td>-3.560797</td>\n",
       "      <td>-4.448464</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000617168.4</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>rs7741058:28596420:T:A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.90788</td>\n",
       "      <td>9.130000e-18</td>\n",
       "      <td>0.004158</td>\n",
       "      <td>0.030366</td>\n",
       "      <td>-0.396302</td>\n",
       "      <td>-3.560797</td>\n",
       "      <td>-4.106542</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000617168.4</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>rs9378185:28606662:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.92754</td>\n",
       "      <td>7.160000e-14</td>\n",
       "      <td>0.002219</td>\n",
       "      <td>0.030366</td>\n",
       "      <td>-0.396302</td>\n",
       "      <td>-3.560797</td>\n",
       "      <td>-4.277918</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000617168.4</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>rs1810619:28609097:G:T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.92814</td>\n",
       "      <td>2.060000e-13</td>\n",
       "      <td>0.002541</td>\n",
       "      <td>0.030366</td>\n",
       "      <td>-0.396302</td>\n",
       "      <td>-3.560797</td>\n",
       "      <td>-4.241781</td>\n",
       "      <td>Down</td>\n",
       "      <td>Down</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Feature_ID   Symbol                SNP_eqtl A1 A2 risk_allele  \\\n",
       "0  ENST00000617168.4  ZSCAN26   rs414205:28526419:G:C  G  C           G   \n",
       "1  ENST00000617168.4  ZSCAN26   rs436827:28557133:T:A  T  A           T   \n",
       "2  ENST00000617168.4  ZSCAN26  rs7741058:28596420:T:A  T  A           A   \n",
       "3  ENST00000617168.4  ZSCAN26  rs9378185:28606662:G:A  G  A           A   \n",
       "4  ENST00000617168.4  ZSCAN26  rs1810619:28609097:G:T  G  T           T   \n",
       "\n",
       "        OR             P       FDR  adj.P.Val     logFC         t    eqtl_t  \\\n",
       "0  1.07930  1.070000e-14  0.002266   0.030366 -0.396302 -3.560797 -4.272345   \n",
       "1  1.08020  7.450000e-15  0.001155   0.030366 -0.396302 -3.560797 -4.448464   \n",
       "2  0.90788  9.130000e-18  0.004158   0.030366 -0.396302 -3.560797 -4.106542   \n",
       "3  0.92754  7.160000e-14  0.002219   0.030366 -0.396302 -3.560797 -4.277918   \n",
       "4  0.92814  2.060000e-13  0.002541   0.030366 -0.396302 -3.560797 -4.241781   \n",
       "\n",
       "  de_dir eqtl_gwas_dir agree_direction Feature_Type  \n",
       "0   Down          Down             Yes   Transcript  \n",
       "1   Down          Down             Yes   Transcript  \n",
       "2   Down          Down             Yes   Transcript  \n",
       "3   Down          Down             Yes   Transcript  \n",
       "4   Down          Down             Yes   Transcript  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans = pd.read_csv('../../_m/transcripts/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID', 'gene_name': 'Symbol'})\n",
    "trans['Feature_Type'] = 'Transcript'\n",
    "trans.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature_ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>SNP_eqtl</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>risk_allele</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_t</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>e127798</td>\n",
       "      <td>VRK2</td>\n",
       "      <td>rs2717068:58094873:A:C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.94673</td>\n",
       "      <td>2.360000e-08</td>\n",
       "      <td>0.005217</td>\n",
       "      <td>0.034502</td>\n",
       "      <td>-0.145416</td>\n",
       "      <td>-3.136916</td>\n",
       "      <td>4.067721</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>e127798</td>\n",
       "      <td>VRK2</td>\n",
       "      <td>rs59159185:58102107:T:A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.91021</td>\n",
       "      <td>1.370000e-08</td>\n",
       "      <td>0.007748</td>\n",
       "      <td>0.034502</td>\n",
       "      <td>-0.145416</td>\n",
       "      <td>-3.136916</td>\n",
       "      <td>3.952051</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>e127798</td>\n",
       "      <td>VRK2</td>\n",
       "      <td>rs2683634</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.94392</td>\n",
       "      <td>2.990000e-09</td>\n",
       "      <td>0.009648</td>\n",
       "      <td>0.034502</td>\n",
       "      <td>-0.145416</td>\n",
       "      <td>-3.136916</td>\n",
       "      <td>3.885998</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>e127798</td>\n",
       "      <td>VRK2</td>\n",
       "      <td>rs1356545:58123798:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.94384</td>\n",
       "      <td>2.850000e-09</td>\n",
       "      <td>0.002981</td>\n",
       "      <td>0.034502</td>\n",
       "      <td>-0.145416</td>\n",
       "      <td>-3.136916</td>\n",
       "      <td>4.225101</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>e127798</td>\n",
       "      <td>VRK2</td>\n",
       "      <td>rs2717056:58128804:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.94400</td>\n",
       "      <td>3.170000e-09</td>\n",
       "      <td>0.009144</td>\n",
       "      <td>0.034502</td>\n",
       "      <td>-0.145416</td>\n",
       "      <td>-3.136916</td>\n",
       "      <td>3.902287</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Feature_ID Symbol                 SNP_eqtl A1 A2 risk_allele       OR  \\\n",
       "0    e127798   VRK2   rs2717068:58094873:A:C  A  C           C  0.94673   \n",
       "1    e127798   VRK2  rs59159185:58102107:T:A  T  A           A  0.91021   \n",
       "2    e127798   VRK2                rs2683634  T  C           C  0.94392   \n",
       "3    e127798   VRK2   rs1356545:58123798:T:C  T  C           C  0.94384   \n",
       "4    e127798   VRK2   rs2717056:58128804:C:T  C  T           T  0.94400   \n",
       "\n",
       "              P       FDR  adj.P.Val     logFC         t    eqtl_t de_dir  \\\n",
       "0  2.360000e-08  0.005217   0.034502 -0.145416 -3.136916  4.067721   Down   \n",
       "1  1.370000e-08  0.007748   0.034502 -0.145416 -3.136916  3.952051   Down   \n",
       "2  2.990000e-09  0.009648   0.034502 -0.145416 -3.136916  3.885998   Down   \n",
       "3  2.850000e-09  0.002981   0.034502 -0.145416 -3.136916  4.225101   Down   \n",
       "4  3.170000e-09  0.009144   0.034502 -0.145416 -3.136916  3.902287   Down   \n",
       "\n",
       "  eqtl_gwas_dir agree_direction Feature_Type  \n",
       "0            Up              No         Exon  \n",
       "1            Up              No         Exon  \n",
       "2            Up              No         Exon  \n",
       "3            Up              No         Exon  \n",
       "4            Up              No         Exon  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons = pd.read_csv('../../_m/exons/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID'})\n",
    "exons['Feature_Type'] = 'Exon'\n",
    "exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature_ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>SNP_eqtl</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>risk_allele</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_t</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:8359987-8360111(-)</td>\n",
       "      <td>RERE</td>\n",
       "      <td>rs13596:8421203:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.05960</td>\n",
       "      <td>1.320000e-08</td>\n",
       "      <td>0.005969</td>\n",
       "      <td>0.009726</td>\n",
       "      <td>-0.227594</td>\n",
       "      <td>-3.772118</td>\n",
       "      <td>4.069341</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:8359987-8360111(-)</td>\n",
       "      <td>RERE</td>\n",
       "      <td>rs2252865</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.05890</td>\n",
       "      <td>1.480000e-08</td>\n",
       "      <td>0.006096</td>\n",
       "      <td>0.009726</td>\n",
       "      <td>-0.227594</td>\n",
       "      <td>-3.772118</td>\n",
       "      <td>4.063307</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:8359987-8360111(-)</td>\n",
       "      <td>RERE</td>\n",
       "      <td>rs10779702</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.05960</td>\n",
       "      <td>1.030000e-08</td>\n",
       "      <td>0.005074</td>\n",
       "      <td>0.009726</td>\n",
       "      <td>-0.227594</td>\n",
       "      <td>-3.772118</td>\n",
       "      <td>4.115450</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:8359987-8360111(-)</td>\n",
       "      <td>RERE</td>\n",
       "      <td>rs894875</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.05790</td>\n",
       "      <td>3.200000e-08</td>\n",
       "      <td>0.003680</td>\n",
       "      <td>0.009726</td>\n",
       "      <td>-0.227594</td>\n",
       "      <td>-3.772118</td>\n",
       "      <td>4.204747</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:8359987-8360111(-)</td>\n",
       "      <td>RERE</td>\n",
       "      <td>rs36085291:8435172:T:TC</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>T</td>\n",
       "      <td>1.05844</td>\n",
       "      <td>2.514000e-08</td>\n",
       "      <td>0.003680</td>\n",
       "      <td>0.009726</td>\n",
       "      <td>-0.227594</td>\n",
       "      <td>-3.772118</td>\n",
       "      <td>4.204747</td>\n",
       "      <td>Down</td>\n",
       "      <td>Up</td>\n",
       "      <td>No</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Feature_ID Symbol                 SNP_eqtl A1  A2 risk_allele  \\\n",
       "0  chr1:8359987-8360111(-)   RERE      rs13596:8421203:T:C  T   C           T   \n",
       "1  chr1:8359987-8360111(-)   RERE                rs2252865  T   C           T   \n",
       "2  chr1:8359987-8360111(-)   RERE               rs10779702  A   G           A   \n",
       "3  chr1:8359987-8360111(-)   RERE                 rs894875  C   T           C   \n",
       "4  chr1:8359987-8360111(-)   RERE  rs36085291:8435172:T:TC  T  TC           T   \n",
       "\n",
       "        OR             P       FDR  adj.P.Val     logFC         t    eqtl_t  \\\n",
       "0  1.05960  1.320000e-08  0.005969   0.009726 -0.227594 -3.772118  4.069341   \n",
       "1  1.05890  1.480000e-08  0.006096   0.009726 -0.227594 -3.772118  4.063307   \n",
       "2  1.05960  1.030000e-08  0.005074   0.009726 -0.227594 -3.772118  4.115450   \n",
       "3  1.05790  3.200000e-08  0.003680   0.009726 -0.227594 -3.772118  4.204747   \n",
       "4  1.05844  2.514000e-08  0.003680   0.009726 -0.227594 -3.772118  4.204747   \n",
       "\n",
       "  de_dir eqtl_gwas_dir agree_direction Feature_Type  \n",
       "0   Down            Up              No     Junction  \n",
       "1   Down            Up              No     Junction  \n",
       "2   Down            Up              No     Junction  \n",
       "3   Down            Up              No     Junction  \n",
       "4   Down            Up              No     Junction  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "juncs = pd.read_csv('../../_m/junctions/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID'})\n",
    "juncs['Feature_Type'] = 'Junction'\n",
    "juncs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24722, 17)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature_ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>SNP_eqtl</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>risk_allele</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_t</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000143420.17</td>\n",
       "      <td>ENSA</td>\n",
       "      <td>rs12124898:150154352:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.92697</td>\n",
       "      <td>1.430000e-08</td>\n",
       "      <td>0.003994</td>\n",
       "      <td>0.017038</td>\n",
       "      <td>0.045351</td>\n",
       "      <td>3.249546</td>\n",
       "      <td>-3.926103</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000066248.14</td>\n",
       "      <td>NGEF</td>\n",
       "      <td>rs778353:233788350:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.93531</td>\n",
       "      <td>2.680000e-12</td>\n",
       "      <td>0.000037</td>\n",
       "      <td>0.015120</td>\n",
       "      <td>0.109010</td>\n",
       "      <td>3.294629</td>\n",
       "      <td>-5.096546</td>\n",
       "      <td>Up</td>\n",
       "      <td>Down</td>\n",
       "      <td>No</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Feature_ID Symbol                  SNP_eqtl A1 A2 risk_allele  \\\n",
       "0  ENSG00000143420.17   ENSA  rs12124898:150154352:C:T  C  T           T   \n",
       "1  ENSG00000066248.14   NGEF    rs778353:233788350:C:T  C  T           T   \n",
       "\n",
       "        OR             P       FDR  adj.P.Val     logFC         t    eqtl_t  \\\n",
       "0  0.92697  1.430000e-08  0.003994   0.017038  0.045351  3.249546 -3.926103   \n",
       "1  0.93531  2.680000e-12  0.000037   0.015120  0.109010  3.294629 -5.096546   \n",
       "\n",
       "  de_dir eqtl_gwas_dir agree_direction Feature_Type  \n",
       "0     Up          Down              No         Gene  \n",
       "1     Up          Down              No         Gene  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([genes, trans, exons, juncs], axis=0, sort=False)\n",
    "print(df.shape)\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('BrainSeq_Phase3_Caudate_GWAS_DE_eQTL_Integration.txt.gz', index=False, header=True, sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
