{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichment in GWAS, TWAS, and DE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import functools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import collections as cx\n",
    "from pybiomart import Dataset\n",
    "from gtfparse import read_gtf\n",
    "from scipy.stats import fisher_exact\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "# GO analysis\n",
    "from goatools.base import download_go_basic_obo\n",
    "from goatools.base import download_ncbi_associations\n",
    "from goatools.obo_parser import GODag\n",
    "from goatools.anno.genetogo_reader import Gene2GoReader\n",
    "from goatools.goea.go_enrichment_ns import GOEnrichmentStudyNS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def get_gtf_genes_df():\n",
    "    gtf_df = read_gtf(\"/ceph/genome/human/gencode25/gtf.CHR/_m/gencode.v25.annotation.gtf\")\n",
    "    return gtf_df[gtf_df[\"feature\"] == \"gene\"][['gene_id', 'gene_name']]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_wgcna_modules():\n",
    "    return pd.read_csv(\"../../_m/modules.csv\", index_col=0)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_database():\n",
    "    dataset = Dataset(name=\"hsapiens_gene_ensembl\", \n",
    "                      host=\"http://www.ensembl.org\",\n",
    "                      use_cache=True)\n",
    "    db = dataset.query(attributes=[\"ensembl_gene_id\", \n",
    "                                   \"external_gene_name\", \n",
    "                                   \"entrezgene_id\"], \n",
    "                       use_attr_names=True).dropna(subset=['entrezgene_id'])\n",
    "    return db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fet(a, b, u):\n",
    "    # a, b, u are sets\n",
    "    # u is the universe\n",
    "    yes_a = u.intersection(a)\n",
    "    yes_b = u.intersection(b)\n",
    "    no_a = u - a\n",
    "    no_b = u - b\n",
    "    m = [[len(yes_a.intersection(yes_b)), len(no_a.intersection(yes_b)) ], \n",
    "         [len(yes_a.intersection(no_b)), len(no_a.intersection(no_b))]]\n",
    "    return fisher_exact(m)\n",
    "\n",
    "\n",
    "def enrichment_rows():\n",
    "    mod = get_wgcna_modules().module.unique()\n",
    "    u = set(get_wgcna_modules().index) \n",
    "    for ii in range(len(mod)): # for each module\n",
    "        a = set(get_wgcna_modules()[(get_wgcna_modules().module) == mod[ii]].index)\n",
    "        yield (mod[ii],\n",
    "               len(a),\n",
    "               *fet(a, gwas_genes, u),\n",
    "               *fet(a, twas_genes, u),\n",
    "               *fet(a, de_genes, u),\n",
    "               )\n",
    "        \n",
    "\n",
    "def enrichment_rows_nomhc():\n",
    "    mod = get_wgcna_modules().module.unique()\n",
    "    u = set(get_wgcna_modules().index) - mhc_genes\n",
    "    for ii in range(len(mod)): # for each module\n",
    "        a = set(get_wgcna_modules()[(get_wgcna_modules().module) == mod[ii]].index) - mhc_genes\n",
    "        yield (mod[ii],\n",
    "               len(a),\n",
    "               *fet(a, gwas_genes - mhc_genes, u),\n",
    "               *fet(a, twas_genes - mhc_genes, u),\n",
    "               *fet(a, de_genes - mhc_genes, u),\n",
    "              )\n",
    "        \n",
    "\n",
    "def convert2entrez(mod):\n",
    "    df = get_wgcna_modules()[(get_wgcna_modules().module) == mod].copy()\n",
    "    df[\"ensemblID\"] = df.index.str.replace(\"\\\\..*\", \"\")\n",
    "    return df.merge(get_database(), left_on='ensemblID', \n",
    "                    right_on='ensembl_gene_id')\n",
    "\n",
    "\n",
    "def obo_annotation(alpha=0.05):\n",
    "    # database annotation\n",
    "    fn_obo = download_go_basic_obo()\n",
    "    fn_gene2go = download_ncbi_associations() # must be gunzip to work\n",
    "    obodag = GODag(fn_obo) # downloads most up-to-date\n",
    "    anno_hs = Gene2GoReader(fn_gene2go, taxids=[9606])\n",
    "    # get associations\n",
    "    ns2assoc = anno_hs.get_ns2assc()\n",
    "    for nspc, id2gos in ns2assoc.items():\n",
    "        print(\"{NS} {N:,} annotated human genes\".format(NS=nspc, N=len(id2gos)))\n",
    "    goeaobj = GOEnrichmentStudyNS(\n",
    "        get_database()['entrezgene_id'], # List of human genes with entrez IDs\n",
    "        ns2assoc, # geneid/GO associations\n",
    "        obodag, # Ontologies\n",
    "        propagate_counts = False,\n",
    "        alpha = alpha, # default significance cut-off\n",
    "        methods = ['fdr_bh'])\n",
    "    return goeaobj\n",
    "\n",
    "\n",
    "def run_goea(mod):\n",
    "    df = convert2entrez(mod)\n",
    "    geneids_study = {z[0]:z[1] for z in zip(df['entrezgene_id'], df['external_gene_name'])}\n",
    "    goeaobj = obo_annotation()\n",
    "    goea_results_all = goeaobj.run_study(geneids_study)\n",
    "    goea_results_sig = [r for r in goea_results_all if r.p_fdr_bh < 0.05]\n",
    "    ctr = cx.Counter([r.NS for r in goea_results_sig])\n",
    "    print('Significant results[{TOTAL}] = {BP} BP + {MF} MF + {CC} CC'.format(\n",
    "        TOTAL=len(goea_results_sig),\n",
    "        BP=ctr['BP'],  # biological_process\n",
    "        MF=ctr['MF'],  # molecular_function\n",
    "        CC=ctr['CC'])) # cellular_component\n",
    "    goeaobj.wr_xlsx(\"GO_analysis_module_%s.xlsx\" % mod, goea_results_sig)\n",
    "    goeaobj.wr_txt(\"GO_analysis_module_%s.txt\" % mod, goea_results_sig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_status', 'gene_name', 'level', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_status', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id gene_name\n",
       "0   ENSG00000223972.5   DDX11L1\n",
       "12  ENSG00000227232.5    WASH7P"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf = get_gtf_genes_df()\n",
    "gtf.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GWAS, TWAS and DE enrichment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load DE, TWAS, and GWAS genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2701"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de_genes = set(pd.read_csv('../../../differential_expression/_m/genes/diffExpr_szVctl_FDR05.txt',\n",
    "                      sep='\\t', usecols=[0], index_col=0).index)\n",
    "len(de_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2000"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_genes = set(pd.read_csv('/ceph/projects/v3_phase3_paper/inputs/gwas/PGC2_CLOZUK/table_s3/hg38/genes/_m/gwas_genes.csv')['gene_id'])\n",
    "len(gwas_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "383"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mhc_genes = set(pd.read_csv('/ceph/projects/v4_phase3_paper/inputs/counts/mhc_region_genes/_m/mhc_genes.csv')['gene_id'])\n",
    "len(mhc_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:Note: NumExpr detected 64 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "684"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot = pd.read_csv(\"/ceph/projects/v4_phase3_paper/inputs/counts/text_files_counts/_m/caudate/gene.bed\", \n",
    "                    sep='\\t', index_col=0)\n",
    "annot[\"Feature\"] = annot.gene_id.str.replace(\"\\\\..*\", \"\", regex=True)\n",
    "twas = pd.read_csv(\"../../../twas/feature_comparison/manuscript_supp_data/_m/\"+\\\n",
    "                   \"BrainSeq_Phase3_Caudate_TWAS_associations_allFeatures.txt.gz\", sep='\\t')\n",
    "twas = twas[(twas[\"FDR\"] < 0.05) & (twas[\"Type\"] == \"Gene\")].merge(annot, on=\"Feature\")\n",
    "twas_genes = set(twas['gene_id'])\n",
    "len(twas_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load WGCNA module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>module</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12.0</th>\n",
       "      <td>grey</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25.0</th>\n",
       "      <td>yellow</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      module            gene_id  gene_name\n",
       "12.0    grey  ENSG00000227232.5     WASH7P\n",
       "25.0  yellow  ENSG00000278267.1  MIR6859-1"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wgcna_df = get_wgcna_modules().merge(gtf, left_index=True, right_on=\"gene_id\", how=\"left\")\n",
    "wgcna_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>module</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>lightgreen</td>\n",
       "      <td>chr11:113412884-113414374(-)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>lightgreen</td>\n",
       "      <td>chr11:113412884-113415420(-)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>lightgreen</td>\n",
       "      <td>chr11:113414462-113415420(-)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         module                       gene_id gene_name\n",
       "NaN  lightgreen  chr11:113412884-113414374(-)       NaN\n",
       "NaN  lightgreen  chr11:113412884-113415420(-)       NaN\n",
       "NaN  lightgreen  chr11:113414462-113415420(-)       NaN"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wgcna_df[(wgcna_df.gene_id.str.startswith(\"chr\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>module</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1914770.0</th>\n",
       "      <td>yellow</td>\n",
       "      <td>ENSG00000099381.16</td>\n",
       "      <td>SETD1A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           module             gene_id gene_name\n",
       "1914770.0  yellow  ENSG00000099381.16    SETD1A"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wgcna_df[(wgcna_df.gene_name == 'SETD1A')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>gwas_or</th>\n",
       "      <th>gwas_p</th>\n",
       "      <th>gwas_fdr_bh</th>\n",
       "      <th>twas_or</th>\n",
       "      <th>twas_p</th>\n",
       "      <th>twas_fdr_bh</th>\n",
       "      <th>de_or</th>\n",
       "      <th>de_p</th>\n",
       "      <th>de_fdr_bh</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>module_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>grey</th>\n",
       "      <td>3940</td>\n",
       "      <td>0.892699</td>\n",
       "      <td>0.214026</td>\n",
       "      <td>0.523174</td>\n",
       "      <td>1.082137</td>\n",
       "      <td>0.439890</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.676330</td>\n",
       "      <td>2.572857e-11</td>\n",
       "      <td>8.086121e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yellow</th>\n",
       "      <td>1587</td>\n",
       "      <td>1.127312</td>\n",
       "      <td>0.308776</td>\n",
       "      <td>0.617552</td>\n",
       "      <td>1.111805</td>\n",
       "      <td>0.444625</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.567600</td>\n",
       "      <td>1.189145e-09</td>\n",
       "      <td>2.906798e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>turquoise</th>\n",
       "      <td>3334</td>\n",
       "      <td>0.864652</td>\n",
       "      <td>0.142167</td>\n",
       "      <td>0.390959</td>\n",
       "      <td>0.877289</td>\n",
       "      <td>0.270678</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.303677</td>\n",
       "      <td>3.095585e-57</td>\n",
       "      <td>3.405144e-56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cyan</th>\n",
       "      <td>375</td>\n",
       "      <td>1.172909</td>\n",
       "      <td>0.522217</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>0.983961</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.867858</td>\n",
       "      <td>7.274883e-06</td>\n",
       "      <td>1.150205e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>green</th>\n",
       "      <td>1568</td>\n",
       "      <td>1.245199</td>\n",
       "      <td>0.063396</td>\n",
       "      <td>0.232452</td>\n",
       "      <td>0.889608</td>\n",
       "      <td>0.537755</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.379828</td>\n",
       "      <td>4.231908e-21</td>\n",
       "      <td>1.862039e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>brown</th>\n",
       "      <td>2253</td>\n",
       "      <td>0.767852</td>\n",
       "      <td>0.025971</td>\n",
       "      <td>0.142843</td>\n",
       "      <td>0.864100</td>\n",
       "      <td>0.327368</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>2.935529</td>\n",
       "      <td>7.804697e-79</td>\n",
       "      <td>1.717033e-77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black</th>\n",
       "      <td>1022</td>\n",
       "      <td>1.108971</td>\n",
       "      <td>0.480575</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>1.127104</td>\n",
       "      <td>0.509117</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>2.301877</td>\n",
       "      <td>1.466545e-23</td>\n",
       "      <td>8.065996e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>midnightblue</th>\n",
       "      <td>370</td>\n",
       "      <td>1.463223</td>\n",
       "      <td>0.093245</td>\n",
       "      <td>0.293057</td>\n",
       "      <td>1.189579</td>\n",
       "      <td>0.535008</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>1.491874</td>\n",
       "      <td>7.027371e-03</td>\n",
       "      <td>8.136956e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lightyellow</th>\n",
       "      <td>226</td>\n",
       "      <td>0.908099</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.584393</td>\n",
       "      <td>0.426782</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>1.826881</td>\n",
       "      <td>8.007519e-04</td>\n",
       "      <td>9.786968e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tan</th>\n",
       "      <td>411</td>\n",
       "      <td>1.003817</td>\n",
       "      <td>0.903400</td>\n",
       "      <td>0.973690</td>\n",
       "      <td>1.415578</td>\n",
       "      <td>0.184201</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>1.956593</td>\n",
       "      <td>3.824870e-07</td>\n",
       "      <td>7.649741e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>royalblue</th>\n",
       "      <td>217</td>\n",
       "      <td>0.728610</td>\n",
       "      <td>0.505137</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>0.609334</td>\n",
       "      <td>0.422714</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.844996</td>\n",
       "      <td>5.253517e-01</td>\n",
       "      <td>5.253517e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>magenta</th>\n",
       "      <td>679</td>\n",
       "      <td>0.623757</td>\n",
       "      <td>0.044510</td>\n",
       "      <td>0.195843</td>\n",
       "      <td>0.528732</td>\n",
       "      <td>0.037884</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>0.449134</td>\n",
       "      <td>1.162519e-07</td>\n",
       "      <td>2.557542e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>darkred</th>\n",
       "      <td>204</td>\n",
       "      <td>1.879159</td>\n",
       "      <td>0.023150</td>\n",
       "      <td>0.142843</td>\n",
       "      <td>1.688871</td>\n",
       "      <td>0.140537</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>2.019585</td>\n",
       "      <td>1.683286e-04</td>\n",
       "      <td>2.178370e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pink</th>\n",
       "      <td>793</td>\n",
       "      <td>1.012345</td>\n",
       "      <td>0.929431</td>\n",
       "      <td>0.973690</td>\n",
       "      <td>1.293422</td>\n",
       "      <td>0.167399</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>3.497723</td>\n",
       "      <td>1.153651e-46</td>\n",
       "      <td>8.460107e-46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>blue</th>\n",
       "      <td>2532</td>\n",
       "      <td>1.351950</td>\n",
       "      <td>0.001631</td>\n",
       "      <td>0.035874</td>\n",
       "      <td>1.165965</td>\n",
       "      <td>0.192978</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>0.729929</td>\n",
       "      <td>7.319488e-06</td>\n",
       "      <td>1.150205e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lightgreen</th>\n",
       "      <td>227</td>\n",
       "      <td>0.193191</td>\n",
       "      <td>0.004850</td>\n",
       "      <td>0.053345</td>\n",
       "      <td>0.731567</td>\n",
       "      <td>0.692431</td>\n",
       "      <td>0.801762</td>\n",
       "      <td>2.029721</td>\n",
       "      <td>4.559218e-05</td>\n",
       "      <td>6.686852e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lightcyan</th>\n",
       "      <td>339</td>\n",
       "      <td>1.014813</td>\n",
       "      <td>0.893464</td>\n",
       "      <td>0.973690</td>\n",
       "      <td>0.784640</td>\n",
       "      <td>0.628696</td>\n",
       "      <td>0.801762</td>\n",
       "      <td>2.727243</td>\n",
       "      <td>1.836597e-13</td>\n",
       "      <td>6.734189e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purple</th>\n",
       "      <td>484</td>\n",
       "      <td>1.146919</td>\n",
       "      <td>0.499134</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>0.611986</td>\n",
       "      <td>0.174787</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>0.437539</td>\n",
       "      <td>4.324439e-06</td>\n",
       "      <td>7.928138e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>red</th>\n",
       "      <td>1260</td>\n",
       "      <td>0.841721</td>\n",
       "      <td>0.287101</td>\n",
       "      <td>0.617552</td>\n",
       "      <td>1.072040</td>\n",
       "      <td>0.669600</td>\n",
       "      <td>0.801762</td>\n",
       "      <td>0.683561</td>\n",
       "      <td>1.292860e-04</td>\n",
       "      <td>1.777682e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>grey60</th>\n",
       "      <td>302</td>\n",
       "      <td>1.229691</td>\n",
       "      <td>0.394382</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>1.234897</td>\n",
       "      <td>0.492128</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.150158</td>\n",
       "      <td>3.957232e-10</td>\n",
       "      <td>1.088239e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>salmon</th>\n",
       "      <td>398</td>\n",
       "      <td>1.162782</td>\n",
       "      <td>0.534400</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>0.837017</td>\n",
       "      <td>0.764880</td>\n",
       "      <td>0.813108</td>\n",
       "      <td>1.393123</td>\n",
       "      <td>2.245696e-02</td>\n",
       "      <td>2.470266e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>greenyellow</th>\n",
       "      <td>440</td>\n",
       "      <td>0.878693</td>\n",
       "      <td>0.723219</td>\n",
       "      <td>0.883934</td>\n",
       "      <td>1.071803</td>\n",
       "      <td>0.776149</td>\n",
       "      <td>0.813108</td>\n",
       "      <td>0.767239</td>\n",
       "      <td>1.164459e-01</td>\n",
       "      <td>1.219909e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              n_genes   gwas_or    gwas_p  gwas_fdr_bh   twas_or    twas_p  \\\n",
       "module_id                                                                    \n",
       "grey             3940  0.892699  0.214026     0.523174  1.082137  0.439890   \n",
       "yellow           1587  1.127312  0.308776     0.617552  1.111805  0.444625   \n",
       "turquoise        3334  0.864652  0.142167     0.390959  0.877289  0.270678   \n",
       "cyan              375  1.172909  0.522217     0.691577  0.983961  1.000000   \n",
       "green            1568  1.245199  0.063396     0.232452  0.889608  0.537755   \n",
       "brown            2253  0.767852  0.025971     0.142843  0.864100  0.327368   \n",
       "black            1022  1.108971  0.480575     0.691577  1.127104  0.509117   \n",
       "midnightblue      370  1.463223  0.093245     0.293057  1.189579  0.535008   \n",
       "lightyellow       226  0.908099  1.000000     1.000000  0.584393  0.426782   \n",
       "tan               411  1.003817  0.903400     0.973690  1.415578  0.184201   \n",
       "royalblue         217  0.728610  0.505137     0.691577  0.609334  0.422714   \n",
       "magenta           679  0.623757  0.044510     0.195843  0.528732  0.037884   \n",
       "darkred           204  1.879159  0.023150     0.142843  1.688871  0.140537   \n",
       "pink              793  1.012345  0.929431     0.973690  1.293422  0.167399   \n",
       "blue             2532  1.351950  0.001631     0.035874  1.165965  0.192978   \n",
       "lightgreen        227  0.193191  0.004850     0.053345  0.731567  0.692431   \n",
       "lightcyan         339  1.014813  0.893464     0.973690  0.784640  0.628696   \n",
       "purple            484  1.146919  0.499134     0.691577  0.611986  0.174787   \n",
       "red              1260  0.841721  0.287101     0.617552  1.072040  0.669600   \n",
       "grey60            302  1.229691  0.394382     0.691577  1.234897  0.492128   \n",
       "salmon            398  1.162782  0.534400     0.691577  0.837017  0.764880   \n",
       "greenyellow       440  0.878693  0.723219     0.883934  1.071803  0.776149   \n",
       "\n",
       "              twas_fdr_bh     de_or          de_p     de_fdr_bh  \n",
       "module_id                                                        \n",
       "grey             0.739413  0.676330  2.572857e-11  8.086121e-11  \n",
       "yellow           0.739413  0.567600  1.189145e-09  2.906798e-09  \n",
       "turquoise        0.739413  0.303677  3.095585e-57  3.405144e-56  \n",
       "cyan             1.000000  1.867858  7.274883e-06  1.150205e-05  \n",
       "green            0.739413  0.379828  4.231908e-21  1.862039e-20  \n",
       "brown            0.739413  2.935529  7.804697e-79  1.717033e-77  \n",
       "black            0.739413  2.301877  1.466545e-23  8.065996e-23  \n",
       "midnightblue     0.739413  1.491874  7.027371e-03  8.136956e-03  \n",
       "lightyellow      0.739413  1.826881  8.007519e-04  9.786968e-04  \n",
       "tan              0.707588  1.956593  3.824870e-07  7.649741e-07  \n",
       "royalblue        0.739413  0.844996  5.253517e-01  5.253517e-01  \n",
       "magenta          0.707588  0.449134  1.162519e-07  2.557542e-07  \n",
       "darkred          0.707588  2.019585  1.683286e-04  2.178370e-04  \n",
       "pink             0.707588  3.497723  1.153651e-46  8.460107e-46  \n",
       "blue             0.707588  0.729929  7.319488e-06  1.150205e-05  \n",
       "lightgreen       0.801762  2.029721  4.559218e-05  6.686852e-05  \n",
       "lightcyan        0.801762  2.727243  1.836597e-13  6.734189e-13  \n",
       "purple           0.707588  0.437539  4.324439e-06  7.928138e-06  \n",
       "red              0.801762  0.683561  1.292860e-04  1.777682e-04  \n",
       "grey60           0.739413  0.150158  3.957232e-10  1.088239e-09  \n",
       "salmon           0.813108  1.393123  2.245696e-02  2.470266e-02  \n",
       "greenyellow      0.813108  0.767239  1.164459e-01  1.219909e-01  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edf1 = pd.DataFrame.from_records(enrichment_rows(), \n",
    "                                 columns=['module_id', 'n_genes', 'gwas_or', 'gwas_p', \n",
    "                                          'twas_or', 'twas_p', 'de_or', 'de_p'],\n",
    "                                 index='module_id')\n",
    "edf1['twas_fdr_bh'] = multipletests(edf1['twas_p'], method='fdr_bh')[1]\n",
    "edf1['gwas_fdr_bh'] = multipletests(edf1['gwas_p'], method='fdr_bh')[1]\n",
    "edf1['de_fdr_bh'] = multipletests(edf1['de_p'], method='fdr_bh')[1]\n",
    "edf1[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', \n",
    "      'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']].to_csv('wgcna_module_enrichment.csv')\n",
    "edf1[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', \n",
    "      'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### No MHC region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edf2 = pd.DataFrame.from_records(enrichment_rows_nomhc(), \n",
    "                                columns=['module_id', 'n_genes', 'gwas_or', 'gwas_p', \n",
    "                                         'twas_or', 'twas_p', 'de_or', 'de_p'],\n",
    "                                index='module_id')\n",
    "edf2['twas_fdr_bh'] = multipletests(edf2['twas_p'], method='fdr_bh')[1]\n",
    "edf2['gwas_fdr_bh'] = multipletests(edf2['gwas_p'], method='fdr_bh')[1]\n",
    "edf2['de_fdr_bh'] = multipletests(edf2['de_p'], method='fdr_bh')[1]\n",
    "edf2[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', \n",
    "      'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']].to_csv('wgcna_module_enrichment_excluding_mhc_region.csv')\n",
    "edf2[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', \n",
    "      'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GO enrichment for each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for mod in get_wgcna_modules().module.unique():\n",
    "    run_goea(mod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
