{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####\n",
    "#\n",
    "# Adapted from:\n",
    "# https://github.com/LieberInstitute/brainseq_phase2.git\n",
    "#\n",
    "#\n",
    "#\n",
    "\n",
    "### libraries\n",
    "library(SummarizedExperiment)\n",
    "library(jaffelab)\n",
    "library(MatrixEQTL)\n",
    "library(sva)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################\n",
    "### load data ####\n",
    "######################\n",
    "\n",
    "load(\"/ceph/projects/v3_phase3_paper/inputs/phase3/_m/count_data/caudate_brainseq_phase3_hg38_rseTx_merged_n464.rda\")\n",
    "load(\"/ceph/projects/v3_phase3_paper/inputs/phase3/_m/count_data/caudate_brainseq_phase3_hg38_rseJxn_merged_n464.rda\")\n",
    "load(\"/ceph/projects/v3_phase3_paper/inputs/phase3/_m/count_data/caudate_brainseq_phase3_hg38_rseExon_merged_n464.rda\")\n",
    "load(\"/ceph/projects/v3_phase3_paper/inputs/phase3/_m/count_data/caudate_brainseq_phase3_hg38_rseGene_merged_n464.rda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#caudate382rnum = read.delim('../_h/382_rnum.txt', header = FALSE, stringsAsFactors = FALSE)$V1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## filter, based on brainseq phase2\n",
    "rse_gene = rse_gene[rowData(rse_gene)$meanExprs > 0.2,]\n",
    "rse_exon = rse_exon[rowData(rse_exon)$meanExprs > 0.2,]\n",
    "\n",
    "rowRanges(rse_jxn)$Length <- 100\n",
    "jRp10m = recount::getRPKM(rse_jxn, 'Length')\n",
    "rse_jxn = rse_jxn[rowMeans(jRp10m) > 0.4,]\n",
    "\n",
    "rse_tx = rse_tx[rowMeans(assays(rse_tx)$tpm) > 0.4,]\n",
    "\n",
    "## keep adult samples & correct region\n",
    "keepInd = which(colData(rse_gene)$Age > 13)\n",
    "rse_gene = rse_gene[,keepInd]\n",
    "rse_exon = rse_exon[,keepInd]\n",
    "rse_jxn = rse_jxn[,keepInd]\n",
    "rse_tx = rse_tx[,keepInd]\n",
    "\n",
    "## extract pd and rpkms\n",
    "pd = colData(rse_gene)\n",
    "geneRpkm = recount::getRPKM(rse_gene, \"Length\")\n",
    "exonRpkm = recount::getRPKM(rse_exon, \"Length\")\n",
    "jxnRp10m = recount::getRPKM(rse_jxn, 'Length')\n",
    "txTpm = assays(rse_tx)$tpm\n",
    "\n",
    "\n",
    "# save expression\n",
    "geneExpression = log2(geneRpkm+1)\n",
    "exonExpression = log2(exonRpkm+1)\n",
    "jxnExpression = log2(jxnRp10m+1)\n",
    "txExpression = log2(txTpm+1)\n",
    "\n",
    "save(geneExpression, exonExpression, jxnExpression, txExpression, file=\"expression.rda\")\n",
    "\n",
    "######################\n",
    "### snp data ####\n",
    "######################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## load SNP data\n",
    "load(\"/ceph/projects/v3_phase3_paper/inputs/phase3/_m/genotype_data/BrainSeq_Phase3_Caudate_RiboZero_Genotypes_n464.rda\")\n",
    "\n",
    "### make mds and snp dimensions equal to N\n",
    "###(repeat rows or columns for BrNum replicates)\n",
    "mds = mds[pd$RNum,]\n",
    "snp = snp[,pd$RNum]\n",
    "\n",
    "## drop SNPs not mapping to hg38\n",
    "keepIndex = which(!is.na(snpMap$chr_hg38))\n",
    "snpMap = snpMap[keepIndex,]\n",
    "snp = snp[keepIndex,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################\n",
    "# statistical model ##\n",
    "######################\n",
    "\n",
    "pd$Dx = factor(pd$Dx,levels = c(\"Control\", \"Schizo\", \"Bipolar\"))\n",
    "\n",
    "mod = model.matrix(~Dx + Sex + as.matrix(mds[,1:5]), data = pd)\n",
    "\n",
    "colnames(mod)[5:9] = colnames(mds)[1:5]\n",
    "\n",
    "######################\n",
    "# create SNP objects #\n",
    "######################\n",
    "\n",
    "theSnps = SlicedData$new(as.matrix(snp))\n",
    "theSnps$ResliceCombined(sliceSize = 50000)\n",
    "\n",
    "snpspos = snpMap[,c(\"SNP\",\"chr_hg38\",\"pos_hg38\")]\n",
    "colnames(snpspos) = c(\"name\",\"chr\",\"pos\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################\n",
    "####### do PCA ########\n",
    "#######################\n",
    "\n",
    "pcaGene = prcomp(t(log2(geneRpkm+1)))\n",
    "kGene = num.sv(log2(geneRpkm+1), mod)\n",
    "genePCs = pcaGene$x[,1:kGene]\n",
    "\n",
    "pcaExon = prcomp(t(log2(exonRpkm+1)))\n",
    "kExon = num.sv(log2(exonRpkm+1), mod, vfilter=50000)\n",
    "exonPCs = pcaExon$x[,1:kExon]\n",
    "\n",
    "pcaJxn = prcomp(t(log2(jxnRp10m+1)))\n",
    "kJxn = num.sv(log2(jxnRp10m+1), mod, vfilter=50000)\n",
    "jxnPCs = pcaJxn$x[,1:kJxn]\n",
    "\n",
    "pcaTx = prcomp(t(log2(txTpm+1)))\n",
    "kTx = num.sv(log2(txTpm+1), mod, vfilter=50000)\n",
    "txPCs = pcaTx$x[,1:kTx]\n",
    "\n",
    "##save(genePCs, file = \"pca_rdas/pcs_caudate_4features_filtered_over13.rda\")\n",
    "\n",
    "dir.create('pca_rdas')\n",
    "save(genePCs, exonPCs, jxnPCs, txPCs, \n",
    "     file=\"pca_rdas/pcs_caudate_4features_filtered_over13.rda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################\n",
    "### Covariates ###\n",
    "##################\n",
    "\n",
    "load(\"pca_rdas/pcs_caudate_4features_filtered_over13.rda\")\n",
    "\n",
    "covsGene0 = t(cbind(mod[,-1],genePCs))\n",
    "covsExon0 = t(cbind(mod[,-1],exonPCs))\n",
    "covsJxn0 = t(cbind(mod[,-1],jxnPCs))\n",
    "covsTx0 = t(cbind(mod[,-1],txPCs))\n",
    "\n",
    "save(covsGene0, covsExon0, covsJxn0, covsTx0, file=\"covariates.rda\")\n",
    "\n",
    "covsGene = SlicedData$new(covsGene0)\n",
    "covsExon = SlicedData$new(covsExon0)\n",
    "covsJxn = SlicedData$new(covsJxn0)\n",
    "covsTx = SlicedData$new(covsTx0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################\n",
    "### feature annotation ###\n",
    "##########################\n",
    "\n",
    "###### gene level\n",
    "posGene = as.data.frame(rowRanges(rse_gene))[,1:3]\n",
    "posGene$name = rownames(posGene)\n",
    "posGene = posGene[,c(4,1:3)]\n",
    "\n",
    "##### exon level \n",
    "posExon = as.data.frame(rowRanges(rse_exon))[,1:3]\n",
    "posExon$name = rownames(posExon)\n",
    "posExon = posExon[,c(4,1:3)]\n",
    "\n",
    "##### junction level \n",
    "posJxn = as.data.frame(rowRanges(rse_jxn))[,1:3]\n",
    "posJxn$name = rownames(posJxn)\n",
    "posJxn = posJxn[,c(4,1:3)]\n",
    "names(posJxn)[2:4] = c(\"Chr\", \"Start\",\"End\")\n",
    "\n",
    "##### transcript level \n",
    "posTx = as.data.frame(rowRanges(rse_tx))[,1:3]\n",
    "posTx$name = rownames(posTx)\n",
    "posTx = posTx[,c(4,1:3)]\n",
    "names(posTx)[2:4] = c(\"Chr\", \"Start\",\"End\")\n",
    "\n",
    "\n",
    "#############################\n",
    "### sliced expression data ##\n",
    "geneSlice = SlicedData$new(geneExpression)\n",
    "exonSlice = SlicedData$new(exonExpression)\n",
    "jxnSlice = SlicedData$new(jxnExpression)\n",
    "txSlice = SlicedData$new(txExpression)\n",
    "\n",
    "geneSlice$ResliceCombined(sliceSize = 5000)\n",
    "exonSlice$ResliceCombined(sliceSize = 5000)\n",
    "jxnSlice$ResliceCombined(sliceSize = 5000)\n",
    "txSlice$ResliceCombined(sliceSize = 5000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################\n",
    "### Run EQTLs ############\n",
    "##########################\n",
    "print(\"Begin eQTL analysis\")\n",
    "\n",
    "meGene = Matrix_eQTL_main(snps=theSnps, gene = geneSlice, \n",
    "    cvrt = covsGene, output_file_name.cis =  \"cis_eqtls_genes.ctxt\" ,\n",
    "    pvOutputThreshold.cis = 0.01,  pvOutputThreshold=0,\n",
    "    snpspos = snpspos, genepos = posGene, \n",
    "    useModel = modelLINEAR, cisDist=5e5,\n",
    "    pvalue.hist = 100,min.pv.by.genesnp = TRUE)\t\n",
    "\n",
    "meExon = Matrix_eQTL_main(snps=theSnps, gene = exonSlice, \n",
    "    cvrt = covsExon, output_file_name.cis =  \"cis_eqtl_exons.ctxt\" ,\n",
    "    pvOutputThreshold.cis = 0.01,  pvOutputThreshold=0,\n",
    "    snpspos = snpspos, genepos = posExon, \n",
    "    useModel = modelLINEAR, cisDist=5e5,\n",
    "    pvalue.hist = 100,min.pv.by.genesnp = TRUE)\n",
    "\n",
    "meJxn = Matrix_eQTL_main(snps=theSnps, gene = jxnSlice, \n",
    "    cvrt = covsJxn, output_file_name.cis =  \"cis_eqtl_junctions.ctxt\" ,\n",
    "    pvOutputThreshold.cis = 0.01,  pvOutputThreshold=0,\n",
    "    snpspos = snpspos, genepos = posJxn, \n",
    "    useModel = modelLINEAR, cisDist=5e5,\n",
    "    pvalue.hist = 100,min.pv.by.genesnp = TRUE)\t\n",
    "\n",
    "meTx = Matrix_eQTL_main(snps=theSnps, gene = txSlice, \n",
    "    cvrt = covsTx, output_file_name.cis =  \"cis_eqtl_trasncripts.ctxt\" ,\n",
    "    pvOutputThreshold.cis = 0.01,  pvOutputThreshold=0,\n",
    "    snpspos = snpspos, genepos = posTx, \n",
    "    useModel = modelLINEAR, cisDist=5e5,\n",
    "    pvalue.hist = 100,min.pv.by.genesnp = TRUE)\t\n",
    "\n",
    "dir.create('eqtl_tables')\n",
    "save(meGene, meExon, meJxn, meTx,\n",
    "     file=\"eqtl_tables/matrixEqtl_output_caudate_4features_p01.rda\")\n",
    "\n",
    "#save(meGene,\n",
    "#    file=\"eqtl_tables/matrixEqtl_output_caudate_4features_p01.rda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################\n",
    "###### annotate ######\n",
    "\n",
    "load(\"eqtl_tables/matrixEqtl_output_caudate_4features_p01.rda\")\n",
    "\n",
    "# extract\n",
    "geneEqtl = meGene$cis$eqtls\n",
    "geneEqtl$gene = as.character(geneEqtl$gene)\n",
    "geneEqtl$snps = as.character(geneEqtl$snps)\n",
    "\n",
    "exonEqtl = meExon$cis$eqtls\n",
    "exonEqtl$gene = as.character(exonEqtl$gene)\n",
    "exonEqtl$snps = as.character(exonEqtl$snps)\n",
    "\n",
    "jxnEqtl = meJxn$cis$eqtls\n",
    "jxnEqtl$gene = as.character(jxnEqtl$gene)\n",
    "jxnEqtl$snps = as.character(jxnEqtl$snps)\n",
    "\n",
    "txEqtl = meTx$cis$eqtls\n",
    "txEqtl$gene = as.character(txEqtl$gene)\n",
    "txEqtl$snps = as.character(txEqtl$snps)\n",
    "\n",
    "################################\n",
    "# add gene annotation info #####\n",
    "################################\n",
    "\n",
    "geneEqtl$Symbol = rowRanges(rse_gene)$Symbol[match(geneEqtl$gene, rownames(rse_gene))]\n",
    "geneEqtl$EnsemblGeneID = rowRanges(rse_gene)$ensemblID[match(geneEqtl$gene, rownames(rse_gene))]\n",
    "geneEqtl$Type = \"Gene\"\n",
    "geneEqtl$Class = \"InGen\"\n",
    "geneEqtl = DataFrame(geneEqtl)\n",
    "# geneEqtl$gene_type = rowRanges(rse_gene)$gene_type[match(geneEqtl$gene, rownames(rse_gene))]\n",
    "\n",
    "exonEqtl$Symbol = rowRanges(rse_exon)$Symbol[match(exonEqtl$gene, rownames(rse_exon))]\n",
    "exonEqtl$EnsemblGeneID = rowRanges(rse_exon)$ensemblID[match(exonEqtl$gene, rownames(rse_exon))]\n",
    "exonEqtl$Type = \"Exon\"\n",
    "exonEqtl$Class = \"InGen\"\n",
    "exonEqtl = DataFrame(exonEqtl)\n",
    "# exonEqtl$gene_type = rowRanges(rse_exon)$gene_type[match(exonEqtl$gene, rownames(rse_exon))]\n",
    "\n",
    "jxnEqtl$Symbol = rowRanges(rse_jxn)$newGeneSymbol[match(jxnEqtl$gene, rownames(rse_jxn))]\n",
    "jxnEqtl$EnsemblGeneID = rowRanges(rse_jxn)$newGeneID[match(jxnEqtl$gene, rownames(rse_jxn))]\n",
    "jxnEqtl$Type = \"Jxn\"\n",
    "jxnEqtl$Class = rowRanges(rse_jxn)$Class[match(jxnEqtl$gene, rownames(rse_jxn))]\n",
    "jxnEqtl = DataFrame(jxnEqtl)\n",
    "# jxnEqtl$gene_type = rowRanges(rse_jxn)$gene_type[match(jxnEqtl$gene, rownames(rse_jxn))]\n",
    "\n",
    "txEqtl$Symbol = rowRanges(rse_tx)$gene_name[match(txEqtl$gene, rownames(rse_tx))]\n",
    "txEqtl$EnsemblGeneID = ss(rowRanges(rse_tx)$gene_id[match(txEqtl$gene, rownames(rse_tx))],\"\\\\.\",1)\n",
    "txEqtl$Type = \"Tx\"\n",
    "txEqtl$Class = \"InGen\"\n",
    "txEqtl = DataFrame(txEqtl)\n",
    "# txEqtl$gene_type = rowRanges(rse_tx)$gene_type[match(txEqtl$gene, rownames(rse_tx))]\n",
    "\n",
    "\n",
    "# merge\n",
    "allEqtl = rbind(geneEqtl, exonEqtl, jxnEqtl, txEqtl)\n",
    "allEqtl$gencodeTx = CharacterList(c(as.list(rowRanges(rse_gene)$gencodeTx[match(geneEqtl$gene, \n",
    "    rownames(rse_gene))]),\n",
    "    as.list(rowRanges(rse_exon)$gencodeTx[match(exonEqtl$gene, rownames(rse_exon))]),\n",
    "    as.list(rowRanges(rse_jxn)$gencodeTx[match(jxnEqtl$gene, rownames(rse_jxn))]),\n",
    "    as.list(txEqtl$gene)))\n",
    "\n",
    "## add snp rs number\n",
    "allEqtl$snpRsNum = snpMap$name[match(allEqtl$snps, snpMap$SNP)]\n",
    "\n",
    "##geneEqtl$snpRsNum = snpMap$name[match(geneEqtl$snps, snpMap$SNP)]\n",
    "\n",
    "\n",
    "## save\n",
    "save(allEqtl, file=\"eqtl_tables/mergedEqtl_output_caudate_4features.rda\",compress=TRUE)\n",
    "\n",
    "## significance filter\n",
    "allEqtl_signif = allEqtl[allEqtl$FDR < 0.05,]\n",
    "save(allEqtl_signif, file=\"eqtl_tables/mergedEqtl_output_caudate_4features_FDR05.rda\",compress=TRUE)\n",
    "\n",
    "## significance filter\n",
    "allEqtl_signif = allEqtl[allEqtl$FDR < 0.01,]\n",
    "save(allEqtl_signif, file=\"eqtl_tables/mergedEqtl_output_caudate_4features_FDR01.rda\",compress=TRUE)\n",
    "\n",
    "##save(geneEqtl, file=\"eqtl_tables/gene_eqtl.rda\")\n",
    "##geneEqtl_signif = geneEqtl[geneEqtl$FDR < 0.05,]\n",
    "##save(geneEqtl_signif, file=\"eqtl_tables/gene_eqtl_fdr05.rda\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
