{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# plotting eQTLs\n",
    "\n",
    "### Kynon Jade Benjamin and Apu√£ Paquola"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import functools\n",
    "import subprocess\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from plotnine import *\n",
    "from pandas_plink import read_plink\n",
    "from warnings import filterwarnings\n",
    "from matplotlib.cbook import mplDeprecation\n",
    "\n",
    "filterwarnings(\"ignore\",category=mplDeprecation)\n",
    "filterwarnings('ignore', category=UserWarning, module='plotnine.*')\n",
    "filterwarnings('ignore', category=DeprecationWarning, module='plotnine.*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = {\n",
    "    #'de_file': '/ceph/projects/v2_phase3_paper/inputs/differentialExpression/_m/genes/diffExpr_szVctl_full.txt',\n",
    "    'biomart_file': '../_h/biomart.csv',\n",
    "    'residual_expression_file': '../../_m/genes_residualized_expression.csv',\n",
    "    'phenotype_file': '/ceph/projects/v3_phase3_paper/inputs/phenotypes/merged/_m/merged_phenotypes.csv',\n",
    "    'plink_file_prefix': '/ceph/projects/v3_phase3_paper/inputs/genotypes/to_brnum/phase3_plink/_m/phase3',\n",
    "    'matrixeqtl_output_file': '../../../_m/cis_eqtls_genes.ctxt',\n",
    "    'gwas_snp_file': '/ceph/projects/v3_phase3_paper/inputs/gwas/PGC2_CLOZUK/map_phase3/_m/libd_hg38_pggc2sz_snps.tsv'\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def get_biomart_df():\n",
    "    biomart = pd.read_csv(config['biomart_file'])\n",
    "    biomart['description'] = biomart['description'].str.replace('\\[Source.*$','')\n",
    "    return biomart\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_residual_expression_df():\n",
    "    residual_expression_df = pd.read_csv(config['residual_expression_file'], index_col=0).transpose()\n",
    "    return residual_expression_df\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_pheno_df():\n",
    "    pheno_df = pd.read_csv(config['phenotype_file'], index_col=0)\n",
    "    return pheno_df\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_expression_and_pheno_df():\n",
    "    return pd.merge(get_pheno_df(), get_residual_expression_df(), left_index=True, right_index=True)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_de_df(filename):\n",
    "    de_df = pd.read_csv(filename, sep='\\t', index_col=0, usecols=[0]+list(range(5,13)))\n",
    "    return de_df\n",
    "\n",
    "\n",
    "def get_gene_symbol(gene_id, biomart=get_biomart_df()):\n",
    "    ensge = re.sub('\\..+$','', gene_id)\n",
    "    ggg = biomart[biomart['ensembl_gene_id']==ensge]\n",
    "    if ggg.shape[0]==0:\n",
    "        return '', ''\n",
    "    gs = ggg['external_gene_name'].values[0]\n",
    "    de = ggg['description'].values[0]\n",
    "    if type(de)!=str:\n",
    "        de = ''\n",
    "    de = re.sub('\\[Source:.*$','',de)\n",
    "    return gs, de\n",
    "\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gene_id_df():\n",
    "    return pd.DataFrame({'gene_id': get_residual_expression_df().columns,\n",
    "                         'ensembl_gene_id': get_residual_expression_df().columns.str.replace('\\..+$','')})\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def gene_info_from_symbol(gene_symbol):\n",
    "    gene_id_df = get_gene_id_df()\n",
    "    biomart=get_biomart_df()\n",
    "    r = biomart[biomart['external_gene_name']==gene_symbol].merge(gene_id_df, on='ensembl_gene_id', how='left')\n",
    "    return r\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def gene_id_from_symbol(gene_symbol):\n",
    "    df = gene_info_from_symbol(gene_symbol)\n",
    "    assert df.shape[0] == 1\n",
    "    return df[['gene_id']].iloc[0].values[0]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_plink_tuple():\n",
    "    '''\n",
    "    Usage: (bim, fam, bed) = get_plink_tuple()\n",
    "    '''\n",
    "    return read_plink(config['plink_file_prefix'])\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_eqtl_df(fdr=0.05):\n",
    "    with subprocess.Popen('''awk ' ($6<%f) || (NR==1) {print}' %s ''' % \n",
    "                          (fdr, config['matrixeqtl_output_file']),\n",
    "                          shell=True, stdout=subprocess.PIPE) as p:\n",
    "        eqtl_df = pd.read_csv(p.stdout, sep='\\t')\n",
    "    return eqtl_df\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gwas_snps():\n",
    "    return pd.read_csv(config['gwas_snp_file'], sep='\\t', index_col=0)\n",
    " \n",
    "    \n",
    "def get_gwas_snp(snp_id):\n",
    "    gwas = get_gwas_snps()\n",
    "    r = gwas[gwas['our_snp_id']==snp_id]\n",
    "    assert len(r) == 1\n",
    "    return r\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def letter_snp(number, a0, a1):\n",
    "    '''\n",
    "    Example:\n",
    "    letter_snp(0, 'A', 'G') is 'AA'\n",
    "    letter_snp(1, 'A', 'G') is 'AG'\n",
    "    letter_snp(2, 'A', 'G') is 'GG'\n",
    "    \n",
    "    '''\n",
    "    if np.isnan(number):\n",
    "        return np.nan\n",
    "    if len(a0)==1 and len(a1)==1:\n",
    "        sep = ''\n",
    "    else:\n",
    "        sep = ' '\n",
    "    return sep.join(sorted([a0]*int(number) + [a1]*(2-int(number))))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def get_snp_df(snp_id):\n",
    "    '''\n",
    "    Returns a dataframe containing the genotype on snp snp_id.\n",
    "    The allele count is the same as in the plink files.\n",
    "    \n",
    "    Example: \n",
    "    get_snp_df('rs653953').head(5)\n",
    "    \n",
    "            rs653953_num rs653953_letter rs653953\n",
    "    Br5168             0              GG    0\\nGG\n",
    "    Br2582             1              AG    1\\nAG\n",
    "    Br2378             1              AG    1\\nAG\n",
    "    Br5155             2              AA    2\\nAA\n",
    "    Br5182             2              AA    2\\nAA\n",
    "    '''\n",
    "    (bim, fam, bed) = get_plink_tuple()\n",
    "    brain_ids = list(set(get_expression_and_pheno_df()['BrNum']).intersection(set(fam['fid'])))\n",
    "    snp_info = bim[bim['snp']==snp_id]\n",
    "    snp_pos = snp_info.iloc[0]['i']\n",
    "    fam_pos = list(fam.set_index('fid').loc[brain_ids]['i'])\n",
    "    dfsnp = (pd.DataFrame(bed[[snp_pos]].compute()[:,fam_pos], \n",
    "                          columns=brain_ids, index=[snp_id + '_num'])\n",
    "             .transpose().dropna())\n",
    "    my_letter_snp = functools.partial(letter_snp, a0=snp_info.iloc[0]['a0'], a1=snp_info.iloc[0]['a1'])\n",
    "    # the 2 - in next line is to workarount a possible bug in pandas_plink? a1 and a0 inverted\n",
    "    dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']].astype('int')\n",
    "    dfsnp[snp_id + '_letter'] = dfsnp[snp_id + '_num'].apply(my_letter_snp)\n",
    "    dfsnp[snp_id] = (dfsnp[snp_id + '_num'].astype('str') + '\\n' + \n",
    "                     dfsnp[snp_id + '_letter'].astype('str')).astype('category')\n",
    "    return dfsnp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def get_gwas_ordered_snp_df(snp_id):\n",
    "    '''\n",
    "    Returns a dataframe containing the genotype on snp snp_id.\n",
    "    The allele count is the number of risk alleles according to GWAS.\n",
    "    \n",
    "    Example: \n",
    "    get_gwas_ordered_snp_df('rs653953').head(5)\n",
    "    \n",
    "            rs653953_num rs653953_letter rs653953\n",
    "    Br5168             2              GG    2\\nGG\n",
    "    Br2582             1              AG    1\\nAG\n",
    "    Br2378             1              AG    1\\nAG\n",
    "    Br5155             0              AA    0\\nAA\n",
    "    Br5182             0              AA    0\\nAA\n",
    "    '''\n",
    "    pgc = get_gwas_snps()\n",
    "    dfsnp = get_snp_df(snp_id).copy()\n",
    "    gwas_snp = get_gwas_snp(snp_id)\n",
    "    if gwas_snp['pgc2_a1_same_as_our_counted'].iloc[0]:\n",
    "        if gwas_snp['OR'].iloc[0] > 1:\n",
    "            pass\n",
    "        else:\n",
    "            dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']]\n",
    "    else:\n",
    "        if gwas_snp['OR'].iloc[0] > 1:\n",
    "            dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']]\n",
    "        else:\n",
    "            pass\n",
    "    dfsnp[snp_id] = (dfsnp[snp_id + '_num'].astype('str') + '\\n' + \n",
    "                     dfsnp[snp_id + '_letter'].astype('str')).astype('category')\n",
    "    \n",
    "    return dfsnp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def get_risk_allele(snp_id):\n",
    "    gwas_snp = get_gwas_snp(snp_id)\n",
    "    if gwas_snp['OR'].iloc[0] > 1:\n",
    "        ra = gwas_snp['A1'].iloc[0]\n",
    "    else:\n",
    "        ra = gwas_snp['A2'].iloc[0]\n",
    "    return ra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_snp_gene_pheno_df(snp_id, gene_id, snp_df_func):\n",
    "    pheno_columns = list(get_pheno_df().columns)\n",
    "    expr_df = get_expression_and_pheno_df()[pheno_columns + [gene_id]]\n",
    "    snp_df =  snp_df_func(snp_id)\n",
    "    return expr_df.merge(snp_df, left_on='BrNum', right_index=True)\n",
    "    \n",
    "\n",
    "def simple_snp_expression_plot_impl(snp_id, gene_id, snp_df_func):\n",
    "    df = get_snp_gene_pheno_df(snp_id, gene_id, snp_df_func)\n",
    "    y0 = df[gene_id].quantile(.01) - 0.26\n",
    "    y1 = df[gene_id].quantile(.99) + 0.26\n",
    "    p = ggplot(df, aes(x=snp_id, y=gene_id)) \\\n",
    "    + geom_boxplot(fill='red', alpha=0.4, outlier_alpha=0) \\\n",
    "    + geom_jitter(width=0.25, stroke=0, alpha=0.6) + ylim(y0, y1) \\\n",
    "    + theme_matplotlib()\\\n",
    "    + theme(axis_text=element_text(size=18), \n",
    "            axis_title=element_text(size=21), \n",
    "            plot_title=element_text(size=22))\n",
    "    return p\n",
    "    \n",
    "\n",
    "def simple_snp_expression_plot(snp_id, gene_id):\n",
    "    return simple_snp_expression_plot_impl(snp_id, gene_id, get_snp_df)\n",
    "\n",
    "\n",
    "def simple_gwas_ordered_snp_expression_plot(snp_id, gene_id):\n",
    "    return simple_snp_expression_plot_impl(snp_id, gene_id, get_gwas_ordered_snp_df)\n",
    "\n",
    "\n",
    "def simple_snp_expression_pheno_plot_impl(snp_id, gene_id, snp_df_func, pheno_var):\n",
    "    df = get_snp_gene_pheno_df(snp_id, gene_id, snp_df_func)\n",
    "    y0 = df[gene_id].quantile(.01) - 0.26\n",
    "    y1 = df[gene_id].quantile(.99) + 0.26\n",
    "    pjd = position_jitterdodge(jitter_width=0.22)\n",
    "    p = ggplot(df, aes(x=snp_id, y=gene_id, fill=pheno_var)) \\\n",
    "    + geom_boxplot(alpha=0.4, outlier_alpha=0) \\\n",
    "    + geom_jitter(position=pjd, stroke=0, alpha=0.6) + ylim(y0, y1) \\\n",
    "    + theme_matplotlib()\\\n",
    "    + theme(axis_text=element_text(size=18), \n",
    "            axis_title=element_text(size=21), \n",
    "            plot_title=element_text(size=22), \n",
    "            legend_text=element_text(size=14))\n",
    "    return p\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gwas_annotation(snp_id):\n",
    "    return 'SZ GWAS pvalue: %.1e' % get_gwas_snp(snp_id).iloc[0]['P']\n",
    "\n",
    "\n",
    "def eqtl_interaction_annotation(snp_id, gene_id):\n",
    "    eqtl_df = get_eqtl_df()\n",
    "    r = eqtl_df[(eqtl_df['SNP']==snp_id) & (eqtl_df['gene']==gene_id)]\n",
    "    assert len(r)==1\n",
    "    return 'genotype:brain region eQTL FDR: %.1e' % r.iloc[0]['FDR']\n",
    "\n",
    "\n",
    "def risk_allele_annotation(snp_id):\n",
    "    return 'SZ risk allele: %s' % get_risk_allele(snp_id)\n",
    "\n",
    "\n",
    "def annotated_genotype_brain_region_interaction_eqtl_plot(snp_id, gene_id):\n",
    "    p = simple_snp_expression_pheno_plot_impl(snp_id, gene_id,\n",
    "                                              get_snp_df, 'Region')\n",
    "    gene_symbol, gene_description = get_gene_symbol(gene_id)\n",
    "    title =\"\\n\".join([gene_symbol,\n",
    "                      eqtl_interaction_annotation(snp_id, gene_id)])\n",
    "    p += ggtitle(title) + ylab('Residualized expression')\n",
    "    return p\n",
    "\n",
    "\n",
    "def gwas_annotated_genotype_brain_region_interaction_eqtl_plot(snp_id, gene_id):\n",
    "    p = simple_snp_expression_pheno_plot_impl(snp_id, gene_id, \n",
    "                                              get_gwas_ordered_snp_df, 'Region')\n",
    "    gene_symbol, gene_description = get_gene_symbol(gene_id)\n",
    "    title =\"\\n\".join([gene_symbol,\n",
    "                      risk_allele_annotation(snp_id),\n",
    "                      gwas_annotation(snp_id),     \n",
    "                      eqtl_interaction_annotation(snp_id, gene_id)])\n",
    "    p += ggtitle(title) + ylab('Residualized expression')\n",
    "    return p\n",
    "\n",
    "\n",
    "def save_plot(p, fn):\n",
    "    for ext in ['png', 'pdf', 'svg']:\n",
    "        p.save(fn + '.' + ext)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Top 10 eQTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_df = get_eqtl_df()\n",
    "eqtl_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_10 = eqtl_df.sort_values('p-value', ascending=True).groupby('gene').first().sort_values('p-value').head(10).reset_index()\n",
    "for x in top_10.itertuples():\n",
    "    filename = \"top_%d_interaction_eqtl_caudate_hippo\" % x.Index\n",
    "    p = annotated_genotype_brain_region_interaction_eqtl_plot(x.SNP, x.gene)\n",
    "    print(filename, x.Index, x.SNP, x.gene)\n",
    "    print(p)\n",
    "    save_plot(p, filename)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Top 10 eQTLs with GWAS significant index SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_eqtl_df = eqtl_df.merge(get_gwas_snps(), left_on = 'SNP', right_on = 'our_snp_id', suffixes=['','_gwas'])\n",
    "print(gwas_eqtl_df.shape)\n",
    "gwas_eqtl_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_gwas_eqtl_df = gwas_eqtl_df[(gwas_eqtl_df['P']<5e-8) & (gwas_eqtl_df['is_index_snp'])].sort_values(['FDR', 'P'])\n",
    "print(top_gwas_eqtl_df.shape)\n",
    "top_gwas_eqtl_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top10 = top_gwas_eqtl_df.sort_values('p-value', ascending=True).groupby('gene').first().sort_values('p-value').head(10).reset_index()\n",
    "for x  in top10.itertuples():\n",
    "    filename = \"top_%d_caudate_hippo_interaction_eqtl_in_gwas_significant_index_snps\" % x.Index\n",
    "    p = gwas_annotated_genotype_brain_region_interaction_eqtl_plot(x.SNP, x.gene)\n",
    "    print(filename, x.Index, x.SNP, x.gene)\n",
    "    print(p)\n",
    "    save_plot(p, filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# eQTL for top GWAS index SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = top_gwas_eqtl_df.sort_values(['P', 'FDR'])\n",
    "p = gwas_annotated_genotype_brain_region_interaction_eqtl_plot(a.iloc[0]['SNP'], a.iloc[0]['gene'])\n",
    "save_plot(p, 'caudate_hippo_interaction_eqtl_for_top_index_snp')\n",
    "p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
